{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d3716dad",
   "metadata": {},
   "source": [
    "# Enrichment Analysis\n",
    "\n",
    "Enrichment analysis (EA) is a technique used to derive biological insight from lists of significantly altered genes. The list of genes can be obtained from Differential Expression (DE) analysis or users' interest. The EA methods rely on the knowledge databases (e.g. KEGG and GO) to identify biological pathways or terms that are enriched in a gene list more than would be expected by chance. The outcome of the EA would be the in-depth and contextualized findings to help understand the mechanisms of disease, genes and proteins associated with the etiology of a specific disease or drug target.\n",
    "\n",
    "Over more than a decade, there are over 50 methods have been developed for EA. In this module, we will focus on pathway analysis using three popular methods including Over Representation Analysis (ORA), Fast Gene Set Enrichment Analysis (FGSEA), and Gene Set Analysis (GSA).\n",
    "\n",
    "## Learning Objectives:\n",
    "1. Data preparation\n",
    "2. Perform enrichment analysis using ORA, FGSEA and GSA\n",
    "3. Visualize and interpret the outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37ff2142-8b48-4250-bb3b-b47175b79f9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-1.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-1.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4731471e",
   "metadata": {},
   "source": [
    "## Gene Mapping\n",
    "In this module, we will use the DE genes with statistic generated in the [**submodule 02** ](./Module02-DE_Analysis.ipynb) using `limma` R package. We can use the following command to load the DE analysis result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5c0da12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>222178_s_at</th><td>222178_s_at</td><td>-0.4473780</td><td>0.3653633</td><td>-73.01106</td><td>4.019112e-55</td><td>2.197450e-50</td><td>81.05538</td></tr>\n",
       "\t<tr><th scope=row>224687_at</th><td>224687_at  </td><td>-4.1431631</td><td>2.2572347</td><td>-48.92036</td><td>5.489424e-46</td><td>1.500671e-41</td><td>74.18354</td></tr>\n",
       "\t<tr><th scope=row>207488_at</th><td>207488_at  </td><td>-0.4017530</td><td>0.6243144</td><td>-39.29933</td><td>4.735684e-41</td><td>8.630785e-37</td><td>68.83493</td></tr>\n",
       "\t<tr><th scope=row>239226_at</th><td>239226_at  </td><td> 0.4483302</td><td>1.3245501</td><td> 28.36499</td><td>7.626558e-34</td><td>1.042455e-29</td><td>58.84908</td></tr>\n",
       "\t<tr><th scope=row>234109_x_at</th><td>234109_x_at</td><td>-0.2289726</td><td>0.7508256</td><td>-27.67010</td><td>2.635337e-33</td><td>2.881741e-29</td><td>58.00438</td></tr>\n",
       "\t<tr><th scope=row>212833_at</th><td>212833_at  </td><td>-2.5976592</td><td>1.7308944</td><td>-24.01172</td><td>2.914912e-30</td><td>2.656213e-26</td><td>52.99567</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & PROBEID & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t222178\\_s\\_at & 222178\\_s\\_at & -0.4473780 & 0.3653633 & -73.01106 & 4.019112e-55 & 2.197450e-50 & 81.05538\\\\\n",
       "\t224687\\_at & 224687\\_at   & -4.1431631 & 2.2572347 & -48.92036 & 5.489424e-46 & 1.500671e-41 & 74.18354\\\\\n",
       "\t207488\\_at & 207488\\_at   & -0.4017530 & 0.6243144 & -39.29933 & 4.735684e-41 & 8.630785e-37 & 68.83493\\\\\n",
       "\t239226\\_at & 239226\\_at   &  0.4483302 & 1.3245501 &  28.36499 & 7.626558e-34 & 1.042455e-29 & 58.84908\\\\\n",
       "\t234109\\_x\\_at & 234109\\_x\\_at & -0.2289726 & 0.7508256 & -27.67010 & 2.635337e-33 & 2.881741e-29 & 58.00438\\\\\n",
       "\t212833\\_at & 212833\\_at   & -2.5976592 & 1.7308944 & -24.01172 & 2.914912e-30 & 2.656213e-26 & 52.99567\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 222178_s_at | 222178_s_at | -0.4473780 | 0.3653633 | -73.01106 | 4.019112e-55 | 2.197450e-50 | 81.05538 |\n",
       "| 224687_at | 224687_at   | -4.1431631 | 2.2572347 | -48.92036 | 5.489424e-46 | 1.500671e-41 | 74.18354 |\n",
       "| 207488_at | 207488_at   | -0.4017530 | 0.6243144 | -39.29933 | 4.735684e-41 | 8.630785e-37 | 68.83493 |\n",
       "| 239226_at | 239226_at   |  0.4483302 | 1.3245501 |  28.36499 | 7.626558e-34 | 1.042455e-29 | 58.84908 |\n",
       "| 234109_x_at | 234109_x_at | -0.2289726 | 0.7508256 | -27.67010 | 2.635337e-33 | 2.881741e-29 | 58.00438 |\n",
       "| 212833_at | 212833_at   | -2.5976592 | 1.7308944 | -24.01172 | 2.914912e-30 | 2.656213e-26 | 52.99567 |\n",
       "\n"
      ],
      "text/plain": [
       "            PROBEID     logFC      AveExpr   t         P.Value     \n",
       "222178_s_at 222178_s_at -0.4473780 0.3653633 -73.01106 4.019112e-55\n",
       "224687_at   224687_at   -4.1431631 2.2572347 -48.92036 5.489424e-46\n",
       "207488_at   207488_at   -0.4017530 0.6243144 -39.29933 4.735684e-41\n",
       "239226_at   239226_at    0.4483302 1.3245501  28.36499 7.626558e-34\n",
       "234109_x_at 234109_x_at -0.2289726 0.7508256 -27.67010 2.635337e-33\n",
       "212833_at   212833_at   -2.5976592 1.7308944 -24.01172 2.914912e-30\n",
       "            adj.P.Val    B       \n",
       "222178_s_at 2.197450e-50 81.05538\n",
       "224687_at   1.500671e-41 74.18354\n",
       "207488_at   8.630785e-37 68.83493\n",
       "239226_at   1.042455e-29 58.84908\n",
       "234109_x_at 2.881741e-29 58.00438\n",
       "212833_at   2.656213e-26 52.99567"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load previously saved data\n",
    "DE.df <- readRDS(\"./data/DE_genes.rds\")\n",
    "rownames(DE.df) <- DE.df$PROBEID\n",
    "head(DE.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722ce8e2",
   "metadata": {},
   "source": [
    "We can see that the genes are saved with probe IDs, we need to convert them into gene symbols so that they can be analyzed using the enrichment analysis method later in this module. We will use the same approach presented in the [**submodule 01** ](./Module01-GEO_Data_Processing.ipynb) with `hgu133plus2.db` and `AnnotationDbi` databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0cb1e8f5-4142-49e5-89cc-1df2da15e1e7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84f28dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the databases\n",
    "suppressMessages({\n",
    "  library(hgu133plus2.db)\n",
    "  library(AnnotationDbi)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51c9e14",
   "metadata": {},
   "source": [
    "Then, we can retrieve vector of probe IDs to perform gene symbol mapping using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f04cf4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>222178_s_at</td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>224687_at  </td><td>ankyrin repeat and IBR domain containing 1</td><td>ANKIB1  </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>207488_at  </td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>239226_at  </td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>234109_x_at</td><td>one cut homeobox 3                        </td><td>ONECUT3 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>212833_at  </td><td>solute carrier family 25 member 46        </td><td>SLC25A46</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & PROBEID & GENENAME & SYMBOL\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 222178\\_s\\_at & NA                                         & NA      \\\\\n",
       "\t2 & 224687\\_at   & ankyrin repeat and IBR domain containing 1 & ANKIB1  \\\\\n",
       "\t3 & 207488\\_at   & NA                                         & NA      \\\\\n",
       "\t4 & 239226\\_at   & NA                                         & NA      \\\\\n",
       "\t5 & 234109\\_x\\_at & one cut homeobox 3                         & ONECUT3 \\\\\n",
       "\t6 & 212833\\_at   & solute carrier family 25 member 46         & SLC25A46\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | 222178_s_at | NA                                         | NA       |\n",
       "| 2 | 224687_at   | ankyrin repeat and IBR domain containing 1 | ANKIB1   |\n",
       "| 3 | 207488_at   | NA                                         | NA       |\n",
       "| 4 | 239226_at   | NA                                         | NA       |\n",
       "| 5 | 234109_x_at | one cut homeobox 3                         | ONECUT3  |\n",
       "| 6 | 212833_at   | solute carrier family 25 member 46         | SLC25A46 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID     GENENAME                                   SYMBOL  \n",
       "1 222178_s_at NA                                         NA      \n",
       "2 224687_at   ankyrin repeat and IBR domain containing 1 ANKIB1  \n",
       "3 207488_at   NA                                         NA      \n",
       "4 239226_at   NA                                         NA      \n",
       "5 234109_x_at one cut homeobox 3                         ONECUT3 \n",
       "6 212833_at   solute carrier family 25 member 46         SLC25A46"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "probeIDs = DE.df$PROBEID\n",
    "# Map probe IDs to their symbols\n",
    "suppressMessages({\n",
    "annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID','GENENAME','SYMBOL'))\n",
    "})\n",
    "# View the first few genes in the mapping table\n",
    "head(annotLookup)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed42b60",
   "metadata": {},
   "source": [
    "Now, we can merge DE data frame with annotation table by the `PROBEID` and remove NA gene symbols using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c5ac1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "DE.df = merge(annotLookup, DE.df, by=\"PROBEID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3bef4b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1007_s_at</td><td>discoidin domain receptor tyrosine kinase 1 </td><td>DDR1  </td><td>-0.23858465</td><td>1.0967772</td><td>-2.08257585</td><td>0.04210211</td><td>0.2531265</td><td>-4.284228</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1053_at  </td><td>replication factor C subunit 2              </td><td>RFC2  </td><td> 0.05830367</td><td>0.9657758</td><td> 1.42311356</td><td>0.16052793</td><td>0.4674726</td><td>-5.359907</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>117_at   </td><td>heat shock protein family A (Hsp70) member 6</td><td>HSPA6 </td><td>-0.01138494</td><td>1.1501582</td><td>-0.08100885</td><td>0.93573821</td><td>0.9861536</td><td>-6.337785</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>121_at   </td><td>paired box 8                                </td><td>PAX8  </td><td> 0.02301084</td><td>1.0581847</td><td> 0.87166663</td><td>0.38729816</td><td>0.7236528</td><td>-5.968512</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1255_g_at</td><td>guanylate cyclase activator 1A              </td><td>GUCA1A</td><td> 0.36999950</td><td>1.4080351</td><td> 1.78615022</td><td>0.07976123</td><td>0.3351480</td><td>-4.812134</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1294_at  </td><td>ubiquitin like modifier activating enzyme 7 </td><td>UBA7  </td><td>-0.11221434</td><td>1.0859983</td><td>-1.61593670</td><td>0.11200995</td><td>0.3925252</td><td>-5.082923</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & PROBEID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1007\\_s\\_at & discoidin domain receptor tyrosine kinase 1  & DDR1   & -0.23858465 & 1.0967772 & -2.08257585 & 0.04210211 & 0.2531265 & -4.284228\\\\\n",
       "\t2 & 1053\\_at   & replication factor C subunit 2               & RFC2   &  0.05830367 & 0.9657758 &  1.42311356 & 0.16052793 & 0.4674726 & -5.359907\\\\\n",
       "\t3 & 117\\_at    & heat shock protein family A (Hsp70) member 6 & HSPA6  & -0.01138494 & 1.1501582 & -0.08100885 & 0.93573821 & 0.9861536 & -6.337785\\\\\n",
       "\t4 & 121\\_at    & paired box 8                                 & PAX8   &  0.02301084 & 1.0581847 &  0.87166663 & 0.38729816 & 0.7236528 & -5.968512\\\\\n",
       "\t5 & 1255\\_g\\_at & guanylate cyclase activator 1A               & GUCA1A &  0.36999950 & 1.4080351 &  1.78615022 & 0.07976123 & 0.3351480 & -4.812134\\\\\n",
       "\t6 & 1294\\_at   & ubiquitin like modifier activating enzyme 7  & UBA7   & -0.11221434 & 1.0859983 & -1.61593670 & 0.11200995 & 0.3925252 & -5.082923\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1007_s_at | discoidin domain receptor tyrosine kinase 1  | DDR1   | -0.23858465 | 1.0967772 | -2.08257585 | 0.04210211 | 0.2531265 | -4.284228 |\n",
       "| 2 | 1053_at   | replication factor C subunit 2               | RFC2   |  0.05830367 | 0.9657758 |  1.42311356 | 0.16052793 | 0.4674726 | -5.359907 |\n",
       "| 3 | 117_at    | heat shock protein family A (Hsp70) member 6 | HSPA6  | -0.01138494 | 1.1501582 | -0.08100885 | 0.93573821 | 0.9861536 | -6.337785 |\n",
       "| 4 | 121_at    | paired box 8                                 | PAX8   |  0.02301084 | 1.0581847 |  0.87166663 | 0.38729816 | 0.7236528 | -5.968512 |\n",
       "| 5 | 1255_g_at | guanylate cyclase activator 1A               | GUCA1A |  0.36999950 | 1.4080351 |  1.78615022 | 0.07976123 | 0.3351480 | -4.812134 |\n",
       "| 6 | 1294_at   | ubiquitin like modifier activating enzyme 7  | UBA7   | -0.11221434 | 1.0859983 | -1.61593670 | 0.11200995 | 0.3925252 | -5.082923 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID   GENENAME                                     SYMBOL logFC      \n",
       "1 1007_s_at discoidin domain receptor tyrosine kinase 1  DDR1   -0.23858465\n",
       "2 1053_at   replication factor C subunit 2               RFC2    0.05830367\n",
       "3 117_at    heat shock protein family A (Hsp70) member 6 HSPA6  -0.01138494\n",
       "4 121_at    paired box 8                                 PAX8    0.02301084\n",
       "5 1255_g_at guanylate cyclase activator 1A               GUCA1A  0.36999950\n",
       "6 1294_at   ubiquitin like modifier activating enzyme 7  UBA7   -0.11221434\n",
       "  AveExpr   t           P.Value    adj.P.Val B        \n",
       "1 1.0967772 -2.08257585 0.04210211 0.2531265 -4.284228\n",
       "2 0.9657758  1.42311356 0.16052793 0.4674726 -5.359907\n",
       "3 1.1501582 -0.08100885 0.93573821 0.9861536 -6.337785\n",
       "4 1.0581847  0.87166663 0.38729816 0.7236528 -5.968512\n",
       "5 1.4080351  1.78615022 0.07976123 0.3351480 -4.812134\n",
       "6 1.0859983 -1.61593670 0.11200995 0.3925252 -5.082923"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove NA gene symbol\n",
    "DE.df = DE.df[!is.na(DE.df$SYMBOL), ]\n",
    "# Remove duplicated genes\n",
    "DE.df = DE.df[!duplicated(DE.df$SYMBOL, fromLast=FALSE),]\n",
    "head(DE.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf678d9",
   "metadata": {},
   "source": [
    "As a result, we obtained the new DE table with two more columns containing gene names and gene symbols. The use of this DE table can be varied based on the selected enrichment analysis tools."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8538d33d",
   "metadata": {},
   "source": [
    "## Enrichment Analysis Using Over-Representation Analysis\n",
    "Over-representation analysis (ORA) is a statistical method that determines whether genes from a pre-defined gene set of a specific GO term or KEGG pathway are presented more than would be expected (over-represented) in a subset of your data. In our learning module, this subset refers to the list of DE genes generated from the `limma` method. For each gene set, an enrichment *p-value* is calculated using the Binomial distribution, Hypergeometric distribution, the Fisher exact test, or the Chi-square test.  The hypergeometric distribution is a popular approach used to calculate enrichment p-value. The formula can be presented as follows:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eca231f2",
   "metadata": {},
   "source": [
    "$$\n",
    "     P(X\\geq x) = 1 - P(X \\leq x-1) = 1 - \\sum\\limits_{i=0}^{x-1}\\frac{\\hphantom{}{M \\choose i }{N - M \\choose n-i}}{N \\choose n}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c18d887",
   "metadata": {},
   "source": [
    "where *N* is the number of background genes (all genes presented in the expression matrix), *n* is the number of “interesting” genes (DE genes), *M* is the number of genes that are annotated to a particular gene set *S* (list of genes in a specific KEGG pathway or GO term), and *x* is the number of “interesting” genes that are annotated to *S* (genes presented in DE genes list and a specific KEGG pathway or GO term)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccade46c",
   "metadata": {},
   "source": [
    "For example, suppose we have an expression matrix with 20,000 genes, of which 500 are differently expressed. Also, suppose that 100 of the 20,000 genes are annotated to a particular gene set S. Of these 100 genes, 20 are members of the DE genes list. The probability that 20 or more (up to 100) genes annotated to S are in the DE genes list by chance is given by"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed50cd53",
   "metadata": {},
   "source": [
    "$$\n",
    "P(X\\geq 20) = 1 - P(X \\leq 19) = 1-\\sum \\limits_{i=0}^{19}\\frac{\\hphantom{}{100 \\choose i}{20000 - 100 \\choose 500-i}}{20000 \\choose 500} = 5.26 \\times 10^{-13}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2e2d72e",
   "metadata": {},
   "source": [
    "The *p-value* indicates that it is very rare to observe 20 of the 100 genes from this set are in the DE genes list by chance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b0be768-1749-4752-b31f-bf5818b33be5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-2.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-2.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e332d07b",
   "metadata": {},
   "source": [
    "### Data preparation\n",
    "To conduct enrichment analysis using ORA, there are several input data that we need to prepare. First, we need to select a set of genes that\n",
    "are significantly altered (*p-value < 0.05*) in the DE genes generated from the `limma` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b1b28353",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selecting a list of significant DE genes\n",
    "DEGenes <- DE.df[DE.df$adj.P.Val <= 0.05, ]\n",
    "# Select genes with symbol\n",
    "DEGenes <- DEGenes$SYMBOL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0d9f75f",
   "metadata": {},
   "source": [
    "Next, we need to define a list of background genes. In this analysis, they are all the genes generated from the DE analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "303d3696",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining background genes\n",
    "backgroundSet <- DE.df$SYMBOL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f19e685",
   "metadata": {},
   "source": [
    "Then, we need to obtain a list of geneset from knowledge databases such as GO and KEGG. In this learning module, the geneset will be retrieved from\n",
    "the *.gmt* files that were processed from the [**submodule 03**](./Module03-Gene_Set_and_Pathway.ipynb). To load the geneset, we will use the `gmt2geneset` function below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d05545e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into 'C:/Users/ASUS/AppData/Local/R/win-library/4.3'\n",
      "(as 'lib' is unspecified)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package 'tidyverse' successfully unpacked and MD5 sums checked\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\tC:\\Users\\ASUS\\AppData\\Local\\Temp\\RtmpC4cJMm\\downloaded_packages\n"
     ]
    }
   ],
   "source": [
    "# install tidyverse\n",
    "suppressWarnings({\n",
    "  install.packages(\"tidyverse\")\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "79cc3d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading tidyverse library that provides a function to read the .gmt file\n",
    "suppressMessages({\n",
    "  library(tidyverse)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0042697a",
   "metadata": {},
   "source": [
    "Here, we also write a function to perform over-representation analysis based on the hyper-geometric testing formula presented above. The `ORA` method will perform\n",
    "hyper-geometric testing for each geneset obtained from GO or KEGG using the function `phyper` available for the `stats` R base package. The output of the `ORA` function is a table that contains a column of terms or pathway names and a column of *p-value*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "38686455",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A customized function to parse a gmt file to a list of genesets\n",
    "gmt2geneset <- function(path) {\n",
    "  # Read the GMT file and create a list of gene sets\n",
    "  genesets <- read_tsv(path, col_names = F) %>% apply(MARGIN = 1, function(r) {\n",
    "    # Extract unique genes from the current row (excluding the first two elements)\n",
    "    genes = unique(r[-(1:2)]) \n",
    "    # Create a list for each gene set with ID, description, and non-NA genes\n",
    "    list(\n",
    "      id = r[1],\n",
    "      description = r[2],\n",
    "      genes = genes[!is.na(genes)]\n",
    "    )\n",
    "  })\n",
    "  # Extract the gene sets from the list and convert gene vectors to character\n",
    "  gs <- lapply(genesets, function(g) g$genes %>% as.character())\n",
    "  # Set names for the gene sets based on their descriptions\n",
    "  names(gs) <- lapply(genesets, function(g) g$description)\n",
    "  # Return the resulting list of gene sets\n",
    "  gs\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c891d7f-4d97-48e3-aab8-7068c9b6c448",
   "metadata": {},
   "source": [
    "**Arguments** of the `ORA` function:\n",
    "\n",
    "- `geneset` (*list*): A list of gene sets to be analyzed.\n",
    "- `DEGenes` (*vector*): A vector of differentially expressed genes.\n",
    "- `backgroundSet` (*vector*): A set of genes representing the background.\n",
    "- `DE.df` (*data.frame*): A data frame containing information about differentially expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d809f4b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Over-Representation Analysis (ORA) Function\n",
    "ORA <- function(geneset, DEGenes, backgroundSet, DE.df) {\n",
    "  # Initialize an empty vector to store the results of the analysis\n",
    "  res <- sapply(geneset, function(gs) {\n",
    "    # Count the number of differentially expressed genes in the current gene set (gs)\n",
    "    wBallDraw <- intersect(gs, DEGenes) %>% length() - 1\n",
    "    # Check for a potential issue with the intersection count being negative\n",
    "    if (wBallDraw < 0) return(1)\n",
    "    # Total number of differentially expressed genes\n",
    "    wBall <- length(DEGenes)\n",
    "    # Remaining genes in the background set\n",
    "    bBall <- nrow(DE.df) - length(DEGenes)\n",
    "    # Number of genes in the current gene set that are in the background set\n",
    "    ballDraw <- length(intersect(gs, backgroundSet))\n",
    "    # Calculate the hypergeometric probability using the phyper function\n",
    "    hypergeo_prob <- phyper(wBallDraw, wBall, bBall, ballDraw)\n",
    "    # Calculate the complementary probability (1 - hypergeometric probability)\n",
    "    1 - hypergeo_prob\n",
    "  })\n",
    "  # Return the vector of results (p-values)\n",
    "  res\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dedfa2a3",
   "metadata": {},
   "source": [
    "### Enrichment analysis using ORA and GO terms\n",
    "In this section, we will perform ORA using genesets obtained from the GO database and the function `ORA` defined above. The detailed code is presented below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "975e07c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GOterms</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>mitochondrial electron transport, cytochrome c to oxygen</th><td>mitochondrial electron transport, cytochrome c to oxygen</td><td>0.0001704211</td></tr>\n",
       "\t<tr><th scope=row>aerobic electron transport chain</th><td>aerobic electron transport chain                        </td><td>0.0004551632</td></tr>\n",
       "\t<tr><th scope=row>respiratory electron transport chain</th><td>respiratory electron transport chain                    </td><td>0.0004551632</td></tr>\n",
       "\t<tr><th scope=row>ATP synthesis coupled electron transport</th><td>ATP synthesis coupled electron transport                </td><td>0.0004551632</td></tr>\n",
       "\t<tr><th scope=row>mitochondrial ATP synthesis coupled electron transport</th><td>mitochondrial ATP synthesis coupled electron transport  </td><td>0.0004551632</td></tr>\n",
       "\t<tr><th scope=row>inner dynein arm assembly</th><td>inner dynein arm assembly                               </td><td>0.0010416338</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & GOterms & pvalue\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tmitochondrial electron transport, cytochrome c to oxygen & mitochondrial electron transport, cytochrome c to oxygen & 0.0001704211\\\\\n",
       "\taerobic electron transport chain & aerobic electron transport chain                         & 0.0004551632\\\\\n",
       "\trespiratory electron transport chain & respiratory electron transport chain                     & 0.0004551632\\\\\n",
       "\tATP synthesis coupled electron transport & ATP synthesis coupled electron transport                 & 0.0004551632\\\\\n",
       "\tmitochondrial ATP synthesis coupled electron transport & mitochondrial ATP synthesis coupled electron transport   & 0.0004551632\\\\\n",
       "\tinner dynein arm assembly & inner dynein arm assembly                                & 0.0010416338\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | GOterms &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| mitochondrial electron transport, cytochrome c to oxygen | mitochondrial electron transport, cytochrome c to oxygen | 0.0001704211 |\n",
       "| aerobic electron transport chain | aerobic electron transport chain                         | 0.0004551632 |\n",
       "| respiratory electron transport chain | respiratory electron transport chain                     | 0.0004551632 |\n",
       "| ATP synthesis coupled electron transport | ATP synthesis coupled electron transport                 | 0.0004551632 |\n",
       "| mitochondrial ATP synthesis coupled electron transport | mitochondrial ATP synthesis coupled electron transport   | 0.0004551632 |\n",
       "| inner dynein arm assembly | inner dynein arm assembly                                | 0.0010416338 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                         GOterms                                                 \n",
       "mitochondrial electron transport, cytochrome c to oxygen mitochondrial electron transport, cytochrome c to oxygen\n",
       "aerobic electron transport chain                         aerobic electron transport chain                        \n",
       "respiratory electron transport chain                     respiratory electron transport chain                    \n",
       "ATP synthesis coupled electron transport                 ATP synthesis coupled electron transport                \n",
       "mitochondrial ATP synthesis coupled electron transport   mitochondrial ATP synthesis coupled electron transport  \n",
       "inner dynein arm assembly                                inner dynein arm assembly                               \n",
       "                                                         pvalue      \n",
       "mitochondrial electron transport, cytochrome c to oxygen 0.0001704211\n",
       "aerobic electron transport chain                         0.0004551632\n",
       "respiratory electron transport chain                     0.0004551632\n",
       "ATP synthesis coupled electron transport                 0.0004551632\n",
       "mitochondrial ATP synthesis coupled electron transport   0.0004551632\n",
       "inner dynein arm assembly                                0.0010416338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading geneset from GO database\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/GO_terms.gmt\")})})\n",
    "# Perform ORA\n",
    "res.ORA <- ORA(geneset = geneset, DEGenes = DEGenes, backgroundSet = backgroundSet, DE.df = DE.df)\n",
    "# Save the result to a table where the first column is GO term name and the second column is the p-value\n",
    "res.df <- data.frame(\n",
    "  GOterms = names(res.ORA),\n",
    "  pvalue = res.ORA\n",
    ")\n",
    "# Order the table based on the p-value\n",
    "res.df <- res.df[order(res.df$pvalue),]\n",
    "# View the most significant GO terms\n",
    "head(res.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0d8278",
   "metadata": {},
   "source": [
    "From the result table, we might conclude that the pathways shown are key components of the disease in study."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa3dbd7",
   "metadata": {},
   "source": [
    "### Enrichment analysis using ORA and KEGG pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69dc9176-9f21-4198-8873-c391fd58b1d2",
   "metadata": {},
   "source": [
    "We will perform the same steps as we did above for the GO terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c31dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the KEGG pathways\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/KEGG_pathways.gmt\")})})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ee2d6ff9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathways</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Purine metabolism - Homo sapiens (human)</th><td>Purine metabolism - Homo sapiens (human)                  </td><td>0.03779064</td></tr>\n",
       "\t<tr><th scope=row>AMPK signaling pathway - Homo sapiens (human)</th><td>AMPK signaling pathway - Homo sapiens (human)             </td><td>0.03926858</td></tr>\n",
       "\t<tr><th scope=row>Complement and coagulation cascades - Homo sapiens (human)</th><td>Complement and coagulation cascades - Homo sapiens (human)</td><td>0.03926858</td></tr>\n",
       "\t<tr><th scope=row>Primary immunodeficiency - Homo sapiens (human)</th><td>Primary immunodeficiency - Homo sapiens (human)           </td><td>0.06601342</td></tr>\n",
       "\t<tr><th scope=row>Proteasome - Homo sapiens (human)</th><td>Proteasome - Homo sapiens (human)                         </td><td>0.09258568</td></tr>\n",
       "\t<tr><th scope=row>Virion - Herpesvirus - Homo sapiens (human)</th><td>Virion - Herpesvirus - Homo sapiens (human)               </td><td>0.09772411</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathways & pvalue\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tPurine metabolism - Homo sapiens (human) & Purine metabolism - Homo sapiens (human)                   & 0.03779064\\\\\n",
       "\tAMPK signaling pathway - Homo sapiens (human) & AMPK signaling pathway - Homo sapiens (human)              & 0.03926858\\\\\n",
       "\tComplement and coagulation cascades - Homo sapiens (human) & Complement and coagulation cascades - Homo sapiens (human) & 0.03926858\\\\\n",
       "\tPrimary immunodeficiency - Homo sapiens (human) & Primary immunodeficiency - Homo sapiens (human)            & 0.06601342\\\\\n",
       "\tProteasome - Homo sapiens (human) & Proteasome - Homo sapiens (human)                          & 0.09258568\\\\\n",
       "\tVirion - Herpesvirus - Homo sapiens (human) & Virion - Herpesvirus - Homo sapiens (human)                & 0.09772411\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathways &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| Purine metabolism - Homo sapiens (human) | Purine metabolism - Homo sapiens (human)                   | 0.03779064 |\n",
       "| AMPK signaling pathway - Homo sapiens (human) | AMPK signaling pathway - Homo sapiens (human)              | 0.03926858 |\n",
       "| Complement and coagulation cascades - Homo sapiens (human) | Complement and coagulation cascades - Homo sapiens (human) | 0.03926858 |\n",
       "| Primary immunodeficiency - Homo sapiens (human) | Primary immunodeficiency - Homo sapiens (human)            | 0.06601342 |\n",
       "| Proteasome - Homo sapiens (human) | Proteasome - Homo sapiens (human)                          | 0.09258568 |\n",
       "| Virion - Herpesvirus - Homo sapiens (human) | Virion - Herpesvirus - Homo sapiens (human)                | 0.09772411 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                           pathways                                                  \n",
       "Purine metabolism - Homo sapiens (human)                   Purine metabolism - Homo sapiens (human)                  \n",
       "AMPK signaling pathway - Homo sapiens (human)              AMPK signaling pathway - Homo sapiens (human)             \n",
       "Complement and coagulation cascades - Homo sapiens (human) Complement and coagulation cascades - Homo sapiens (human)\n",
       "Primary immunodeficiency - Homo sapiens (human)            Primary immunodeficiency - Homo sapiens (human)           \n",
       "Proteasome - Homo sapiens (human)                          Proteasome - Homo sapiens (human)                         \n",
       "Virion - Herpesvirus - Homo sapiens (human)                Virion - Herpesvirus - Homo sapiens (human)               \n",
       "                                                           pvalue    \n",
       "Purine metabolism - Homo sapiens (human)                   0.03779064\n",
       "AMPK signaling pathway - Homo sapiens (human)              0.03926858\n",
       "Complement and coagulation cascades - Homo sapiens (human) 0.03926858\n",
       "Primary immunodeficiency - Homo sapiens (human)            0.06601342\n",
       "Proteasome - Homo sapiens (human)                          0.09258568\n",
       "Virion - Herpesvirus - Homo sapiens (human)                0.09772411"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Perform Hyper-geometric testing\n",
    "res.ORA <- ORA(geneset = geneset, DEGenes = DEGenes, backgroundSet = backgroundSet, DE.df = DE.df)\n",
    "# Save the results to a data frame\n",
    "res.df <- data.frame(\n",
    "  pathways = names(res.ORA),\n",
    "  pvalue = res.ORA\n",
    ")\n",
    "\n",
    "res.df <- res.df[order(res.df$pvalue),]\n",
    "head(res.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa0e1c7",
   "metadata": {},
   "source": [
    "## Enrichment Analysis using FGSEA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b46d66",
   "metadata": {},
   "source": [
    "This submodule describes `FGSEA`, one of the methods for evaluating pathway enrichment in transcriptional data and it stands for Fast preranked Gene Set Enrichment Analysis (GSEA). `FGSEA` quickly calculates arbitrarily low GSEA P-values for a collection of gene sets and is based on an algorithm that allows  it to make more permutations and get accurate *p-values*.\n",
    "It extends the GSEA algorithm which calculates the enrichment score (the degree to which a set S is over-represented at the top or bottom of the ranked list L), estimates the p-value of the Enrichment Score using a permutation test and adjusts the estimated significance level to account for multiple hypotheses in addition to calculating the q-values for FDR control. Typically, GSEA requires inputs of a list of gene sets (GO term or pathway with a set of genes), a vector DE genes with statistic. In addition, users can limit the number of terms or pathways of interest by adjusting the `minSize` and `maxSize` parameters. The code to install and load the FGSEA R package is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b9702a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To perform enrichment analysis using FGSEA, we first install the fgsea package\n",
    "suppressMessages({if (!require(\"BiocManager\", quietly = TRUE))\n",
    "        install.packages(\"BiocManager\")\n",
    "  suppressWarnings(BiocManager::install(\"fgsea\", update = F))\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "52f7aae1-0bd8-4ff4-86f2-041eb6e41db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the package\n",
    "suppressPackageStartupMessages({\n",
    "  library(\"fgsea\")\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cdb41931-b689-4fb9-9c5d-0e1c443bfe00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-3.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Run the following command to take the quiz\n",
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-3.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb6f9407",
   "metadata": {},
   "source": [
    "Next, we need to prepare a vector of DE genes with its statistic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "38d25313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>DDR1</dt><dd>0.25312654919025</dd><dt>RFC2</dt><dd>0.467472580609517</dd><dt>HSPA6</dt><dd>0.986153555292704</dd><dt>PAX8</dt><dd>0.723652752171903</dd><dt>GUCA1A</dt><dd>0.335147977832318</dd><dt>UBA7</dt><dd>0.392525231374342</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[DDR1] 0.25312654919025\n",
       "\\item[RFC2] 0.467472580609517\n",
       "\\item[HSPA6] 0.986153555292704\n",
       "\\item[PAX8] 0.723652752171903\n",
       "\\item[GUCA1A] 0.335147977832318\n",
       "\\item[UBA7] 0.392525231374342\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "DDR1\n",
       ":   0.25312654919025RFC2\n",
       ":   0.467472580609517HSPA6\n",
       ":   0.986153555292704PAX8\n",
       ":   0.723652752171903GUCA1A\n",
       ":   0.335147977832318UBA7\n",
       ":   0.392525231374342\n",
       "\n"
      ],
      "text/plain": [
       "     DDR1      RFC2     HSPA6      PAX8    GUCA1A      UBA7 \n",
       "0.2531265 0.4674726 0.9861536 0.7236528 0.3351480 0.3925252 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get the gene list and their statistic from DE results\n",
    "stats = DE.df$adj.P.Val\n",
    "names(stats) = DE.df$SYMBOL\n",
    "head(stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9950a6f6",
   "metadata": {},
   "source": [
    "### Enrichment analysis using FGSEA and GO terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5093c0c",
   "metadata": {},
   "source": [
    "The package `fgsea` has its own function named `gmtPathways` to load the genesets from the `gmt` file. It is recommended to use this function for compatibility. We can load the GO terms geneset and view the first five GO terms with associated genes using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b2d3d20a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`GO:0000002`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AKT3'</li><li>'DNA2'</li><li>'DNAJA3'</li><li>'ENDOG'</li><li>'FLCN'</li><li>'LIG3'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000003`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'A1CF'</li><li>'A2M'</li><li>'AAAS'</li><li>'ABAT'</li><li>'ABCC8'</li><li>'ABHD2'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000012`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'APLF'</li><li>'APTX'</li><li>'ERCC6'</li><li>'ERCC8'</li><li>'LIG4'</li><li>'PARP1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000018`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABL1'</li><li>'ACTB'</li><li>'ACTL6A'</li><li>'ACTR2'</li><li>'ALYREF'</li><li>'ANKLE1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000022`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AURKB'</li><li>'AURKC'</li><li>'BIRC5'</li><li>'CDCA8'</li><li>'INCENP'</li><li>'KIF23'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`GO:0000002`] \\begin{enumerate*}\n",
       "\\item 'AKT3'\n",
       "\\item 'DNA2'\n",
       "\\item 'DNAJA3'\n",
       "\\item 'ENDOG'\n",
       "\\item 'FLCN'\n",
       "\\item 'LIG3'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000003`] \\begin{enumerate*}\n",
       "\\item 'A1CF'\n",
       "\\item 'A2M'\n",
       "\\item 'AAAS'\n",
       "\\item 'ABAT'\n",
       "\\item 'ABCC8'\n",
       "\\item 'ABHD2'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000012`] \\begin{enumerate*}\n",
       "\\item 'APLF'\n",
       "\\item 'APTX'\n",
       "\\item 'ERCC6'\n",
       "\\item 'ERCC8'\n",
       "\\item 'LIG4'\n",
       "\\item 'PARP1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000018`] \\begin{enumerate*}\n",
       "\\item 'ABL1'\n",
       "\\item 'ACTB'\n",
       "\\item 'ACTL6A'\n",
       "\\item 'ACTR2'\n",
       "\\item 'ALYREF'\n",
       "\\item 'ANKLE1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000022`] \\begin{enumerate*}\n",
       "\\item 'AURKB'\n",
       "\\item 'AURKC'\n",
       "\\item 'BIRC5'\n",
       "\\item 'CDCA8'\n",
       "\\item 'INCENP'\n",
       "\\item 'KIF23'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`GO:0000002`\n",
       ":   1. 'AKT3'\n",
       "2. 'DNA2'\n",
       "3. 'DNAJA3'\n",
       "4. 'ENDOG'\n",
       "5. 'FLCN'\n",
       "6. 'LIG3'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000003`\n",
       ":   1. 'A1CF'\n",
       "2. 'A2M'\n",
       "3. 'AAAS'\n",
       "4. 'ABAT'\n",
       "5. 'ABCC8'\n",
       "6. 'ABHD2'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000012`\n",
       ":   1. 'APLF'\n",
       "2. 'APTX'\n",
       "3. 'ERCC6'\n",
       "4. 'ERCC8'\n",
       "5. 'LIG4'\n",
       "6. 'PARP1'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000018`\n",
       ":   1. 'ABL1'\n",
       "2. 'ACTB'\n",
       "3. 'ACTL6A'\n",
       "4. 'ACTR2'\n",
       "5. 'ALYREF'\n",
       "6. 'ANKLE1'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000022`\n",
       ":   1. 'AURKB'\n",
       "2. 'AURKC'\n",
       "3. 'BIRC5'\n",
       "4. 'CDCA8'\n",
       "5. 'INCENP'\n",
       "6. 'KIF23'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`GO:0000002`\n",
       "[1] \"AKT3\"   \"DNA2\"   \"DNAJA3\" \"ENDOG\"  \"FLCN\"   \"LIG3\"  \n",
       "\n",
       "$`GO:0000003`\n",
       "[1] \"A1CF\"  \"A2M\"   \"AAAS\"  \"ABAT\"  \"ABCC8\" \"ABHD2\"\n",
       "\n",
       "$`GO:0000012`\n",
       "[1] \"APLF\"  \"APTX\"  \"ERCC6\" \"ERCC8\" \"LIG4\"  \"PARP1\"\n",
       "\n",
       "$`GO:0000018`\n",
       "[1] \"ABL1\"   \"ACTB\"   \"ACTL6A\" \"ACTR2\"  \"ALYREF\" \"ANKLE1\"\n",
       "\n",
       "$`GO:0000022`\n",
       "[1] \"AURKB\"  \"AURKC\"  \"BIRC5\"  \"CDCA8\"  \"INCENP\" \"KIF23\" \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "GO_term_hallmark <- gmtPathways(\"./data/GO_terms.gmt\")\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(GO_term_hallmark, head)\n",
    "tmp[1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d06027d9",
   "metadata": {},
   "source": [
    "Running the FGSEA can be done by calling the `fgsea` with two required inputs that we created from the two previous steps.\n",
    "\n",
    "Used **arguments** of the `fgsea` function:\n",
    "- `pathways` (*list*): List of gene sets to check.\n",
    "- `stats` (*numeric vector*): Named vector of gene-level stats. Names should be the same as in 'pathways'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6efd9f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running fgsea\n",
    "suppressWarnings(fgseaRes <- fgsea(pathways = GO_term_hallmark,\n",
    "                  stats = stats\n",
    "                  ))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad26fa54",
   "metadata": {},
   "source": [
    "The output of the FGSEA method is a table where rows are GO terms IDs or pathway IDs. The columns are *p-values*, *adjusted p-values*, *Enrichment Score*, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f90ac106",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GO:0006122</td><td>8.953719e-06</td><td>0.04648262</td><td>0.5933255</td><td>-0.7106070</td><td>-2.576761</td><td> 11</td></tr>\n",
       "\t<tr><td>GO:0050907</td><td>1.940800e-05</td><td>0.04648262</td><td>0.5756103</td><td> 0.3788597</td><td> 1.572075</td><td>132</td></tr>\n",
       "\t<tr><td>GO:0060390</td><td>2.028628e-05</td><td>0.04648262</td><td>0.5756103</td><td> 0.4911948</td><td> 1.855808</td><td> 46</td></tr>\n",
       "\t<tr><td>GO:0007606</td><td>8.178477e-05</td><td>0.12802215</td><td>0.5384341</td><td> 0.3450003</td><td> 1.465727</td><td>182</td></tr>\n",
       "\t<tr><td>GO:0050911</td><td>1.167654e-04</td><td>0.12802215</td><td>0.5384341</td><td> 0.3929034</td><td> 1.584887</td><td> 94</td></tr>\n",
       "\t<tr><td>GO:0030850</td><td>1.235671e-04</td><td>0.12802215</td><td>0.5188481</td><td> 0.4547892</td><td> 1.738423</td><td> 51</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t GO:0006122 & 8.953719e-06 & 0.04648262 & 0.5933255 & -0.7106070 & -2.576761 &  11\\\\\n",
       "\t GO:0050907 & 1.940800e-05 & 0.04648262 & 0.5756103 &  0.3788597 &  1.572075 & 132\\\\\n",
       "\t GO:0060390 & 2.028628e-05 & 0.04648262 & 0.5756103 &  0.4911948 &  1.855808 &  46\\\\\n",
       "\t GO:0007606 & 8.178477e-05 & 0.12802215 & 0.5384341 &  0.3450003 &  1.465727 & 182\\\\\n",
       "\t GO:0050911 & 1.167654e-04 & 0.12802215 & 0.5384341 &  0.3929034 &  1.584887 &  94\\\\\n",
       "\t GO:0030850 & 1.235671e-04 & 0.12802215 & 0.5188481 &  0.4547892 &  1.738423 &  51\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| GO:0006122 | 8.953719e-06 | 0.04648262 | 0.5933255 | -0.7106070 | -2.576761 |  11 |\n",
       "| GO:0050907 | 1.940800e-05 | 0.04648262 | 0.5756103 |  0.3788597 |  1.572075 | 132 |\n",
       "| GO:0060390 | 2.028628e-05 | 0.04648262 | 0.5756103 |  0.4911948 |  1.855808 |  46 |\n",
       "| GO:0007606 | 8.178477e-05 | 0.12802215 | 0.5384341 |  0.3450003 |  1.465727 | 182 |\n",
       "| GO:0050911 | 1.167654e-04 | 0.12802215 | 0.5384341 |  0.3929034 |  1.584887 |  94 |\n",
       "| GO:0030850 | 1.235671e-04 | 0.12802215 | 0.5188481 |  0.4547892 |  1.738423 |  51 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway    pval         padj       log2err   ES         NES       size\n",
       "1 GO:0006122 8.953719e-06 0.04648262 0.5933255 -0.7106070 -2.576761  11 \n",
       "2 GO:0050907 1.940800e-05 0.04648262 0.5756103  0.3788597  1.572075 132 \n",
       "3 GO:0060390 2.028628e-05 0.04648262 0.5756103  0.4911948  1.855808  46 \n",
       "4 GO:0007606 8.178477e-05 0.12802215 0.5384341  0.3450003  1.465727 182 \n",
       "5 GO:0050911 1.167654e-04 0.12802215 0.5384341  0.3929034  1.584887  94 \n",
       "6 GO:0030850 1.235671e-04 0.12802215 0.5188481  0.4547892  1.738423  51 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][, -8])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7866448",
   "metadata": {},
   "source": [
    "From the result table, we can select top five up regulated GO terms and top five down regulated GO terms. Then we can plot them using the built-in function `plotGseaTable`\n",
    "\n",
    "Used **arguments** of the `plotGseaTable` function:\n",
    "\n",
    "- `pathways` (*list*): Pathways to plot table.\n",
    "- `stats` (*data.frame*): Gene-level stats.\n",
    "- `fgseaRes` (*data.frame*): Table with fgsea results.\n",
    "- `gseaParam` (*numeric*): GSEA-like parameter. Adjusts displayed statistic values, values closer to 0 flatten plots. Default = 1, value of 0.5 is a good choice too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f1b4ed14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2di7qrKLNFae+J8fL+T9sCXgAxMWtXYhLHON9ZQcSKmcVE\nRP/eqgeAf0YdfQIAvwBGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICR\nAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABG\nAhAAIwEIgJE2UYakaIP62u484Ixcuks+nF1+EQr3yt+jVO18iZrRVV2ZKpVJ/YqX/Q4d9n7o\no/vDBzPlO/GdlCq785BzmqmT+Nn9ldcaKXG+xDNSN/6MpBP7LqFA67AY6Y+MQ2amykj1wUaq\nlSpuw+ctnzrpP/JaI6lq+RLvmwqVDQNBG2r8L98lFOjZsBhpk1G8LuiqH2GkZJotDV1RZF70\nWiOlqp2/xPsmpcylqBP7eoz0eUzimc96uCNJyn6cmZjKUiVVf1OFblPbjl2o29yyU6k5fPqU\n5KryqdiW+srUX1KVXOzZtrk+Ma9y+iFdqg+cf4vT2P7YUl866my4a6l7OZRq7PnGjPR8MC28\no286eNH9RX8/yzFw7yo0aJIMl0umdv+Ae0Wq7FS+dIyk7/WHq0FiWhV2bjI0XVrmynTx6ziv\nkWQM7VZoMnPW5raj8iqnH5TrE1vO0GlsfmypG1/sbrn7fx27MCe8NlKpVms5D4NV46/KzGWu\nHYreL/r7WU6B3XjDtyijG0b6M1Y4O39X6qotsXSEQfFu6HPpIPq1n+6ntWeWlrW9WBVKZjlg\nfWoLtT6b4W6udk7MrZyOyro++C1TY71pfDSMDI3eLXgZHWLby4e/2GD2Df00LW93Dw+DJU3f\nJMNvsCNUNfy+VXb+dpZTYDfedazFSH9nXrXrlpp+MdI4xOoRcZjg5br7ZbNnTKPU3AC8YGY3\n5XTukbn5pk5PoOYTcyuno25BiKWxmdhkdofktG76qou+xEWM1NeFlnj/V9rTq81P1cqmatnT\n/5uR5sBOPHvtrzHSP+A/R2rrKvOMNP3Nhv5aDncBlfWU0/Kih8zbC2Z2KyM5nXM5saDHOj0h\n8luGa9Pos3KYyTSN+NnqUWWjq9+qJJyqPghmP/S1vrUTsOAX/fks53lIEA8j/QOecNncJ0Mj\n1YNTkrRPp1ne0nK8u5Kf2Xn3SM8bKfZb9JAxXjsrfd8k9Hhq+V69KrPZ1Zv9l22nZ+sxqjRK\nhL/oz2c5fqziYaR/wBWuUOmlbqNGGmYYt2FYLPWSWOe1HOpqbTB5rvb2az5N51Rdb8R/UPS3\nDJejxtwaaOoyFb5H6rX5m1VXn8v7+7/b34fxK9HnufpFfz/L8VIXxsNI/8Aq4xtGKlUxGKYe\n/hZ+y2GgzZpXzOzc50idvR1yVhSmv3l4s+P2wbWR9MUzCduKMM6WVLrq6vm4OBg+q7sbzN60\naK1LZWbP61/0t7OcAzvxrIg3jPQP+Ea69c00a257z0iDyvpSpOzdqtNS3xgkr5jZmbvf3LzZ\nUJr1QrO2NNyT5e6JOZXeD/J+y1Q/3sdU+s9VftVOU60nX4N0l0G6W7Z/tX1aXNNat1b59S/6\n21nOgZ14Nat2/4wrXDnebtx0R9Nd170u2XlQZkdVp6Xp76+Y2Q3cpnft7CTPTur1jY1zYkul\n94OcMwyM1Oh+eXXOX4ZJyGTs6s6923Qu2b3jg2D2yY7ZSO2Bq1/0t7OcA7s5NA/jCoz0D3jC\nDVJmN7M2ektDI1Vm5aga3xdbWpp510tmdpprPnTMrBptchn8XQSXyqXS/0HLGYY3VJV58UG/\n2SDoo/l767WR+qbQv+L6VLB8uIWxG9dx9hr+oj+e5RzYzWHFmw3HU6vXzOxOjOTN23OBMdJx\nPDH3h30cZqQWIx3FU3N/2MdRRiqV88ghdrjoyYBLsiyYgRRHGSk3/8upO4eLngzAScFIAAJg\nJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiA\nkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIx2I+Q9Kj3+m/wy3mv/j2L162X/DDcQhVQey\nx0hzzbS3/8d/vgReAvk4kH8z0nLNsvXk8kgQ/0CkjeS2g/eC4gfyQiOZC5T9Q4rfASofyPuM\npLhMvRjEPZB3G4l7qdeBqAdyoJG4PAmDngdytJFU39MDhEDGA/kYI9EL/hkkPJBPM5KiO/wZ\nlDuQjzSSYjniLyDZgXyykegYz4FeB4KRfgf0OpAvMtKX3j5NirwksF/xim+BfXyZkV7TJV/K\nq854Vu7l3wQ7+EojfdO072UnipE+Coz0Yl42s7PRNzfgvXy9kT6996wvHOLRoxvwXjDSO3jp\nfdLLv+Wz6S750Knyy7xdpkqlZec1KhOVjFVLUc29NWyQ1cFhSrmNY/yekTZ/68Ufvr1TfTQB\n8491tV43mdR3Qt4LffN2NoVSRXs/8hx4JdAJqZMxe4lV7Tpl081PZmpSv9g4aV81qPzaKWiy\neR7nMVLjOcVR8bFIwbGu1usmTk7m07nTxbvE3VnbM/GG063Iq6hnNNIgWHEbPm+5zd+wXQ6O\nakvXSTeVNH2TqJtXbFQeaXBRWdd3hWq82unLbv0WpzHSIIhvpNzff0+k4FhH63WTUH3nt0XJ\nvdDJcHCXq3JvZI8zGimZ/VKoYXbXzf6pneGoNJVXPfQ5xcsyFDq1mdG31Slwag1dEvQZl7MY\naej8gRn8C8o9kcJjHa3XTUL13RlehKu392pidu7FcTPyMi+Zv2n7a36V6zIetuWQlWpJSqmN\nZTtMrvS0z4ydTvGi5vsqp3aaNWde7djKv/Hy+FUjraY9ZR8Y6eLtvydSeKyj9bpJqP59Wt+j\n/lXu2cgnNFIezCPyRcCbEUrNPW3sVksxV3Ux3HP2mw2coqbxJgohZzFSswhicFR8KFJ4bCCw\n3ySy8w6Zat2WqeqrRBWup5+IfEIjhWK420t5y0iGzG+QmuHqFjHS3QvSaYwUiOyq+Fik4FhH\n63WTp4xUqWuQe3NewbIHRtpk1mTsd88YSQ3i952ZATq1lcq7vsnWRmpUcfdMTmokR8XHIgXH\nOlrHsxUesIU7KR+P0osNhbkNej4yRnrKSJZOL4O6tWY5PV+3LVX4wMM/k5MayVHxsUjhsYvW\n/2akNOnC3Ospfjs+ung28gmN5N4jaWGy5R7JuZlMFuWSlYhh7TCOJVWkbXJf3nMbaadI4bGL\n1puh9hipMPaNDKJ/jHxCI12dqYSdmDmrdvPCqV2maZdVu9ZZsdGHrWobPZZ5tY+WjzDSDpGi\nxzbOo1GnyTpTd4LOOo/k94z0MPIJjeQ8R+q0TPHnSJWprLXJnGJi7ouNnqvaS1C7XuYNOauR\nHBUfixQ99hL06LGJp/6joKGR7MGtu7D+ROQzGqke5tjmzYbSrtE4bzZc50bxNxtKrWRnZvVe\n7XCNu6X6aO8JeL5+MnFWPDM4Ku4RKTh21nrdZPVeyTOnNdwdmXcm/hb5jEbqb9O7duMkr542\nrYZWu9TUZH6xs0eW0do8OGwo31/XPRFTj7W3lI6KO0SKHZvHm7jqP3Fa9qOKHbw38imNNNwn\n5UM+smp607er9Nvf1ZhRq11nXvftI8X0Eta2xZDaOqzdt3p0DjwzuCruEMk/1tF63cRV/4nT\nGj/qbH3w3sikGkAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEI\ngJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQg\nAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIx2M/ffp7f+bv2rMifsXPh4S\ndTB7jOQ06XHYZ0I+DubvRlLTwfABkIiDwUi/AYk4mJcaiey+DaQ+mDcYiRy/AUQ+mDcbiXy/\nCIQ9mIOMRN6FQdCDwUi/AYIeDEb6DRD0YD7KSIr+8FcQ7mA+z0h0ib+AagfzyUaic+wHrQ4G\nI/0GaHUw32Ck7+0kSr3oLapV3O/V6EfASK/kVWfuJOLF3wQ7+TojfVOPedm5YqSP43uN9AVd\n52UzOxt9cwPeD0Z6IesLh3j06Aa8n58wkgobfRIvvU96+bd8Nt0lH/pgfpm3y1SptOy8RmWi\nkrHKKTaFUkUb1Ha6spkPy+pVhE1+x0jqXm+6eLu6uKDxI9NAxJiqc/hppzOnuyfRzdsZO5Uw\n8hw4CHtKI9WJsiRWteu4qWqnUWZq0qBY2+M6v9bGa5a2VXDYNj9mpI3zabwTbZNFfUfQKGW4\nP6bqHD7Y+Whq1yXuztipbEVeRT2jkQbBitvwectVMm6XQ07b0nXSTSVN3yTq5hX7ZCh2uSq9\n2lIV+k/e6wEs6/QFqvEP2+YURhpUcHcUWj+rmiNo/EhVdFrVYqqIqTqHD3c+mnLm3mlFTuVu\nZI8zGimZ/VKoYXbXzf6pneGoNJVXfW1xilcjc6cN6NQmSh9mJM+M1K1u5jS4wxmMNIwu3olO\nkyPlCRolV+4RfVTVJfxq5/1Vu6u3N3Iqm5HVaj3whEa6mkuHoS2Hbl8tY1CpjWWzlis98Wh0\nW6dYqGZq69RaTAamPpJFGkT5RSOtpj1l7w/943klnqD3WA6PqLqE3yf5ROv7O3IqT0Q+oZHy\nYLKVLwLejFBq6hfjh1NMVV8lZrbh1hocE64Ou8MZjNQEKlTj1K7yBL1Dpwem8avWqjZh7c5n\nR5lq3ZaRU3ki8gmNFIrhbi/luCOUys0NaR9KO8wSzHUtNSPXDSOFkb26i15tSOzAMwt6h8ty\n8xpX9S9GqtQ1yH3sVDDSJrMm4zz3OSPpG9JCj6W+tJc8MfPnSuVd32QYKYzs1VXz0qYj6DZt\nssyoXFWj2Vp92QZmmubnfjmV5yNjpCeNpKeBrV4GXUlrVi7sSniOkcLIbt1FX707o5cj6CZd\nki0bYkZKky7M/XIqGGkH7j2SFiZb7pGcm8lkUc4pqmitxa73DP0jqeINopzRSKlZ5eyiw1GE\nzHVZXNVxc5/khsLMFiOD6B8jn9BI1+WhhBHGW7Wbpxh2maZdVu3G4nyYU+sEszS6j6wbxDij\nkZSr4nq3T5tm7ssGcVXH4/dJPh6h5jnJFDl2Knsjn9BIznOkTssUf45Umcpam2xVbPUiklNr\nnyOZSYEtXrTgToM7nNFIdng3V3BH0Dh1sCuu6hh+n+TjEaGR4qeyN/IZjVQPNzHmzYbSrtE4\nbzZc50bxNxsGs5gXF67rNxu6XM/5TfGWBg3AM1Kp9Otwpe6VjqBRVhaLqzqGf1py97Tip7I3\n8hmN1N+md+3GSV49bVoNrXapqcmCYhWtTeZiZ4t50OD0TD3WfmZRQWMUznVjlZd1+Gcl9+6L\noqeyN/IpjTTcJ+VDh8+qafLdVfrt72qc11nt7BvKfVDs6yxWOxRT+y55O6Q+r1cNzo5vpD4q\naPy40EhRVafwz0ruLzDETmVv5JMaCUAWjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlA\nAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQA\nATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIx2IUr35\n11HVVJhK9l9NVYrsfA2k6kD2GGmpGcu98w8Tw8dAPg7k34w07piOIpeHgvgHIm0ktWyT1zeD\n4AfyUiNNVyk89RZQ+UDeZ6R5F7wIxD2QdxuJe6nXgagHcpiR1HQISIGYB3K0kcYyCICMB/Ix\nRuL26Z9BwQP5JCPNR8CfQLkD+UgjKfz0F5DsQD7ZSHSM50CvA/l4IykctRdkOpCvMdJ0el/H\npMgrQt/dhHfyZUb6wuvT60439OeXCfNbfKWRvslNL/QRV6QP4nuN9CXd5oUzO4z0QXy9kVT/\n2T1IzX9eFHtzE97JTxnpY3sSRnod3SUfekZ+mbfLVKm07LxGZaKSsSpWXPrmQFMoVbRhW6c2\nyu8ZabM7XVZ7buM3OCrG6Bw9x1BpUOGGn1o7Ie90cf+sOp2uZqvJGHkOjJH6vk7G7CW2l1+n\nbNZOo8zUpJvF6ZhEB7Slzm/r1Mb5RSPFO1SzckqXeEZKNiRqEzdRmjKi6hzeUd85py2Cs7Jf\n1cSbBJExUm86eHEbPm+5zd+wXQ6JakvXSTeVNH2TqNtWcYqli8lQ2+Wq9BsstRucxkiDIGF9\n7tbUi6ABhZGvVMUcShWdvk4UTqM5fJge9/wenpX5klLl0SarxGMkPfJMfinUMLvrZv/UzkBX\nmsqrqraKhi7JzbbOdqdt6TRwajc4i5EuKguNdFVOjVVxSyLno9cGDCrc8GF63Bnew7NKVBcE\nvhcZIw1qzGlry2GEqZZrRqmNZbXMlZ5NNLptvGjIjfjFMh9wGhQqnG+H/KqRwk41CBz059br\nxFbFKMnom3A0cuMt4cP03GV9VqtveiLyCY2UB/OIfOnwNyPU1FP78SNe1DTWg6nqq8RMOdwG\nTu0GZzFSEwz0+o6jXWqaO7PfapzaVX51p7JY+NX16x7rs+qnofQPkU9opFAMb3BTQem+kcah\nVCm9CGjGMq/tXLt9Jqcw0kr0Sl292dqdweailwCSS1jrrQv9zUiRhsOEM/Q0Rtpk1mTunqt9\nvafcppGa8ZZX6TvRrtDDptd2rt0+k1MaycyQ5prGWzgIqcxpBxq29qYqmq3wy+4RNrzkif2q\n5yNjpH8wUjmOi3bNtNWLo17buXb7TE5ppDTpnJoyuLp4XPQloiu8CVffJVkQVMhI/bj+hJH2\n4N4jaWGy5R7JuZlMFuXixaWNq7LT4HFWT2qkwjhnrkm2BdI3mnra1/mjUbYam/yE/IuRwlXW\nvZFPaKSrM5XQwnirdvMcwi7TtMtSXVh0bOcsyXpt59oNTmoktZxN/2iNLTIatWm2eltk3O+l\nZwex3AR1eyOf0EjOc6ROyxR/jlSZylqbLF7U046L27bVK0mrtq27vhTww0YK+9W2kS4qXEnw\nc2XztFwm6pigY3gvPTvwTGOfI4VT8b2Rz2ikWqncvNlQzi/4TG82XOdGO95smNfNB/E7PZG/\neg2cWliP/VNNfvdxW6n0m3Pl0oHjA9MYbP3+wRNnZd5s6PLA13sjn9FI/W16126c5NXTpu3z\nVrvU1GTbxXH+rqmiDZza0zN12aXrTqX03uL39I5bNh9RuFezVfj0Ocn9s0piB++NfEojDfdJ\n+aBaVk1z7a7Sb39XY0atdva143676I5ndRZrsNSenm0jPVoZWPQMJqmx8K76z5/VcHAazjP3\nRj6pkQBkwUgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACM\nBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEw\nEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMNKBKPNPotp/p9T+W6VK/5+us4WjTxB2\nQ64OZI+RzIZa9o4pw2QfBvk4kH8zkt3RT23I5aEg/oGIG2lsqLhevR0UP5DXGomr1DtB6AN5\nn5HmL4AXgboH8m4jcS/1OhD1QA40EnkXBkEP5GgjLX/gX0HGA8FIvwMyHsinGUnRHf4Myh3I\nZxqJ5Yi/gGQH8slGomM8B3odCEb6HdDrQL7ISN/bT1515j8j0A/wZUZS39hZXnXSYdxv1OZn\n+Eojfde071ULkau4X6TJ74GRXs0LTxYjfQ4Y6dVgpFOAkV7M/EtfFHt78yR0l3zoVPll3i5T\npdKy8xqViUrGqqWoLJEGWT0deFlEvdzV90RG8oVwVOwcFXcd21/S9QFzkynaKGJ//2T9yF2h\nVNFsNRkjz9nHSH1fJ2Mmk9ZsX8dNVTuNMlOT+sXG6QKrBpU9sFkWdJxijPMYyRfCUbFN3ETs\nOLYv7QFdtImTk94f9XZEtqfSxJv4kXuMNFAPA89t+LzlKhm3yyGRbek66aaSpm8SdfOKjcoj\nDS4q6/R4ZnIwVE2iOsUopzFSIISjYqHKXpuj2H9s0WnFi1gTJyfeme+KbE6iXE7tQWSMNIw8\nk18KNczuutk/tTPQlabyqi8zTvEyXXa82szo25o+MZhqSo9TjHMWI4VCOCpO86TNkOGx+fqA\npYmTE+/Md0VOVBeeyb3IGOm6DDptOTigMgYwlNpYVstc6dmGGTud4kXN91VO7dQbMv2nnHPh\nFOP8qpHC/aEQjorJKF2yGSsuolu3NHFy8ph4ZO9Mnoh8QiPl3pVfb8/T4psRSi29xunj5iNX\ndTHcc/abDfpmSXLTR/vAwlmMFArhqFiNU7uq3yAuYmcGrVWTh9e3h5HLxePPRT6hkUIxvMFN\nBaWVkQyZ3yA1w9VtvhKpSLzomZzESKEQjor9Rd/iJ5f1ERvHWi7eutDfjBRpeFXL7OTJyOc1\n0ryk84yRlLrqFVs9bDm1lcq7vpln3BgpUqW8jVnFQTtnwXMj3Cpem+T+DiEjXfLEnsrzkTHS\nU0aydHoZ1K01C6c5RnIah1XrOqPiRV8CukLduyStju2SLNghZKR+XH/CSHtw75G0MNlyj+Tc\nTCaLcslKxLB26AlJNTfASJGqjbrULJV17gOax8dmq9Zjk3WmHpxozN5JrMnDyCc00tV5BmEn\nZs6q3TzHsMs07bJq1zorNvqwVW0z9QaMFKnaqNtzDQl2tmm2enw7Nlln6sGJPj7TvZFPaCTn\nOVKnZYo/R6pMZa1N5hTtswaj56r2MqmMkSJVbp2joh3pw8vAnWMHxbPNJk5OdhE9qzTW5GHk\nMxqpHu5nzJsNpTIJdN5suM6N4m82lGZKb57OebXDNe6WTkfvNtKJ8IRwVByK3Vix79g25qOp\nSezNhifOashilwe3a3sjnzLRt+ldu3GSV0+b1glWu3RZol2KnT2yjNZOl32MtMa7f3RVzBaZ\ndx1bTLmKNkkfR9uMPL5rl8WbPIp80kRf80G1rJrm2l2l3/6uxnmd1c6+ltxHiuklrG2H9Obz\nkw2MtMbvso6K/aLirmPVXSN1j6Ntn1XvnNXTkUk0gAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgA\nAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMB\nCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIx0\nHGj/Q5DM41CT+vZD/2ulylaPH2TneyBVx7HTSGZ7ZSoS91mQj+P4VyOpJX2k8WjIwHG80Eik\n9d2g+HFgpB8CxY8DI/0QKH4cbzcSyX4daHscRxqJvAuDoMeBkX4IBD2OzzESj37/GQQ8js8y\nEj3hn0C+4/hQI6nN42EbFDuOTzcSfeMJEOs4vsJIdJB9oNNxfJGRvribvOrU1/mAg/hGI31d\nf1EvOuMw7tcJ80N8pZFe1TFfxKtWI/mftXwQX2yk2Bl9Iup1PRwjfQxfb6QXdlMhMNIZwEiv\nZv4pL4q9vXkSuks+9Ij8Mm+XqVJp2XmNykQlY9VS7AqliiZs0C1FZQkjxDiTkS5uQ6Vclfr+\ndj/Kxd99Sdeqzk0myU105wfsiezmdjPyfNoYqe/rZExk0prt65TY2mmUmZo0KNojG7+2TeZo\njdNFnMOi/KKRNvpT461RTHIndrNL7vZC/9i+tId20Sa+5L5dH0Z2cvsoco+RBuph4LkNn7fc\npnLYLgcPtKXrpJtKmr5J1M0rlqrQf3K/QaHKftzXmH1hhDjnMdKgwnpHPSmT310JDI5tVNHp\n60QRaxKT/I6P/MhObvdExkjDyDP5pVDD7K6b/VM7A11pKq+q8oqJ0g2MuE7tdLFXOsNVJEKc\nXzXS6ssvKlt7pUvGLnu9u6QeHpsvYq2bxCTfjB1GdnK7JzJGui6DTlsOI0xlrieGUhvLapkr\nPe0zFxinaDEXMqd2nJro6ouab7xWh4WcxkiDwGuv5MoOWm3MZQ+O9fv70uSh5I8jT/PNZyOf\n0Eh5MNnKl2nxzQhl+7HTu4NB0LGb/ajGqV2lg9XFcFPa9+vDVpzGSE1EhWYavjLV3jNS7Fg9\ni8hiTR5K/jByuQyEz0U+oZFWL3eoWHnTSMNMZOWTi75PTXQKcnt/m/W7tD+JkfqICtMFqVLX\nRx0/svvirQv9zUiRhlNu/xD5vEaal3SeNNIlT8xM2a2tTChbe9Wr4Xpcw0hufVDZjKsFZrL0\ntJFae38VzVb0gJ2Rp9z+ITJGetZI/bhG4a0wlOYpxPJcSq+TYiS3Pqgsx0tKmnTPG6lLsmCH\nkJH6MbcYaQ/uPZIWJlvukZybyWRRLglF7PQdqVObmmlK5zxliB4WcmYjjeIUxk/PGilbPZsb\nmzyU/FHkMbd/iHxCI12dZxDLUoGhXJY37TJNu6zatc6KjT7MqV0PV0GDOCc20jRize843OuH\nwc42zdqNJg8lvx85Vrc38gmN5DxH6rRM8edIlamstcmcon3W0Oprj1drgyVLg9xrEOeXjbSq\n9yumpwR/MFLtLtgFTR5Kfi+yk9s/RD6jkWqlcvNmQ2nfUnHebLjOje682dDluht4tfr1rlKr\nbP505g7g4ZsNZyJwSu69ifPM1K6N+Whq8qzkXmQnt3+IfEYj9bfpXbtxkldPm9ZHVrt0Xsd2\ni0m0NpuLnW1QBg1Oz9Rlx89UdZGde44topewafNJyf2zSmIH7418SiMN90n5oFpWTXPtrtJv\nf1djcq129o3uPijql4DTS6zWaRtpcHYCI/k+eMZI8bngtPmk5MFZLal7PvJJjQQgC0YCEAAj\nAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACM\nBG38pHUAABxWSURBVCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiA\nkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABjpdyG3bwSx\nfxcVKcGLQOLfBSO9EST+XTDSG0Hi3+WhkUi+HGj5uzxjJPrBP4KAvwtGeiMI+LtgpDeCgL8L\nRnojCPi7YKQ3goC/C0bq1cCrQt/dhB8CI6n5j3xojHQaXmOkb+oxrzOS4op0Hl5ppC/qNy86\nVYx0Gt5hpJfdggjBPRL8M2800id3I65Ir6O75MNQlV/m7TJVKi27oNllUqdMVFbPxWRpeFPj\n8VOtmvAbHMLbjfRg9L94uztHyaZQqmijB3V6V+PWxBq7mTJBnRRs/fggU72X5s3I8xEYqe/r\nZNQwsfm4Tpr6KjaTyJnZWS3FdGzRJaZFm8zRpkiJ1+AYDjDSvTCNt9fRrK9tMRzHDLaZ46RY\n4yBTU3ruLjYEmeq9ND+MjJF6k4riNnzecqvisF0O+WxL30lNMop4UVmnR8ZGX2CSRu+42Sa5\nbVGocvhbqmL5Br/BMRxjpK1Is5wWV7NkELXLTUWIaVCqfKmJNJ5DB+nZsWo3Z8pL8+PIGKnX\ng9zkl0INs7tu9k/tDnSDrqOImZGv1bkrTdPrOGxdxyF46k2TmF0yJv56f4x+MR9lpEVO/0Cl\nVdKu6Jxrw0Kiur5354yRxktoPz394weyc6Z6L807ImOkQY1ZvLYctKsW6UptrDFtQ61vEJXp\nK4yeijQ2QjvJPM7f5uTmyvqxDTrPmznMSLFgi5wWRzP/IhD9zsU2kcZLaC89e5gyZc/efmR/\ninxCI+XO9dluz6m5GaGsdk14pTG9xK3JVGsL1ThNGYerZnLm3OAYPspIi5wWR7NU9VWiiugt\nksGOb5ZI42imdtG4l5/IwU9EPqGRVi93qK1941ZqRqNbaKRKXafti74lTqZsT8Oc0+AQPspI\nfSjuoplSuX/X7zPMj73uHmv8RyO5FyQ3zX+IfF4jzYufD41UqbzrmywwkrnMzy2c9Z5mXHRw\nGxzCZxtp0UzpG/nhLr+KH3bJE/fWxGm8BPybkZpleaj30vyHyBjpsZHsEmweGClNumn7oofM\nrhgnIOW4duE0OIaPNpKjmV3dbt215YBimdu5jf/VSKX/vGNJM0bag3uPpIXJlnuk4GZykm3I\nd1KZrWTWszBJsC1SM0Poxp4wtnEbHMMnGMl95ulp4Wj2uPs7i3TxxuNm8jCSR/CQb0nzHyKf\n0EhX54KuhfFW7bzphSdbo3NuF2/awW/zg/Hgzmk2o9vgGD7aSI5m+ePuv+yLN/bW1tqdq3bR\nNbgmuDDujXxCIznPkTot08ZzpN4ZjXTtRYtYmaa1WRZdfGKHq3HYvIzTEIy0OtKtdzSzorbu\nurPTrOu9aV+88XyvOqVnDxdnMbD30vyHyGc0Uj3MhM2bDaVd/XHebLh6DUcRzdP1W6p3rh6d\nK9tAv95VWpVz70HHkVO7T8PTwtFssIl5peAaOcZo3+VLl483HkOv3z+4h58pN81/iHzKPN+m\nd+3GSV49bVoNVTDcdra5GalSU5yHw7FF5tSm7pIqRnLwp3iOZpUvqkey7LPHRRtPodM7kVYs\nmTKHu2l+PvJJ83zNB9WyanqHuKv029/VqGtopL4tBn3t9M++szzHmVo4tfGFPwiN5GpWZ66o\nPkOz9OIeF2s8Lwv56dl3PlPJSfPzkckzgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEw\nEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATA\nSAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAA\nIwEIgJEABMBIAAJgpN+F3L4RxP5d1OYGiIO+v0toJEW6XwfK/i4Y6Y2g7O+Ckd4Iyv4uGOmN\noOzv8ryR6A1/Bul+F4z0RpDud8FIbwTpfhcpI9FHdoBIvwtG6tXAq0Lf3YQfAiOp+Y98aIx0\nGl5lpO/pM68zkuKKdB4wkuFFp4uRTsM7jPTp/Yd7JPhn3mWkD+9DXJFeR3fJh6Eqv8zbZapU\nWnZeo0uqkrGqTGLFoYlatVUWU9sUShXtK3/IXT7NSBevYbco2Xmi+nRaw8YLk64bz6GnSM5y\n3cbpKeWmKv5VkcjzERip7+tk1DCxvfw6aVo7jUrbQmcsM8W094uDUaYsLG0bJzv1EuEQjjDS\nnS9ovFlWm8wZcIpr7D6ne5cRVefQXnruLzZMOU+2v+pOZIzUmw5e3IbPW25VHLZLnc/SdVKj\nik6PR8XQTiVN3yTq5hV7/anCto3K5xDJ0LbLVfnO3+ZykJE2vmNWy1IYXUqtmVNcYWpLR1RH\n63VoLz39fSNZ6rnt+qvuRcZIvR55Jr8UapjddbN/amegy8cLuNLi6v1XVXnFIZnZqLLT9mJ2\nGa6mf3TOkPdmPspIi1pjw0Uzp7giUZ2/K183XkK76Rmb3V9s6BJn3Au/6m5kjDSoMYvXlsMI\nUy3XjFIby9NSl3OlJx3mWuMU++EwP0vWSPONV6HC+fab+SgjhWoloyESr7h1XqtdXo7m0G56\ndpGrcOLtfdUTkU9opNy5PtvtucPfrEOc+0+VbQ6eTTB8mba5qovhplRvp6qvEjMROYiPMlKo\nVjXO5yqvGKdcRqcRo/U69L1rW/SkVvNu/6ueiHxCI61e7lDb+y76ir49CwnmAbU2kiEzO3P/\nZvbdfJSR+lDci763Ty5BMcJVrW8zL9660J+NFF6QIl+FkTaZNRmX1+4YqU3yfq+RprZXvZyr\nxzWl70+7YnugfTWfbaTKyF8FxQiXPAn3Wa2dgH8zUhOub8xf9XxkjHTHSF2SeQfcM9LYdtzQ\n66R2JbV1l0zfy0cb6aLH/s6s9jjFDYJ9k9b/aqQyuK4tX4WRduDeI2lhsuUeKbiZzKwHkkXE\nZNNImeeXR4tRb+ETjOQ+8/SUSM2syow4TnGDYOkzW7UcQyfPSZ5E2oWrrHsjn9BIV+eCroXx\nVu2cOUSbZvYRoV2xaZdVu3by26zr3NaJG1mofS8fbaSdy99+a0OotbM/SM8D4mtwwWnsjXxC\nIznPkTotU/w50rAxzdUqs7/WfnOKmkn0pa19GGEEt21bd33pvXyCkbzj3D12iO+W5e/4A7dJ\nz+UaVMcEHUMH6XnAxZ8wrr/qmchnNFKtVG7ebCjtmprzZsN1buQYYOvNhlllp21p5vtm8j2k\npNNT/yXmyfGMNAjVjXI5xRXmdYMuX7p8fGDafP/gHrn/oG/1VU9FPqOR+tv0rt04yaunTdvn\njXaF805jOi9pu8V+Vtlp29nQpk9UXlvwp3jZok52R6hk2bfKyzp0+ozk6bz4bQ9PYgfvjXxK\nIw33SfmgWlZNc+2u0m9/V6OuRjvlJMy+nWxbLsW+X2b3ftt0HNTqzGkLwb2So6QnasCi5yov\n69DdvUhb5+OeURouHe6NfFIjAciCkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEA\nBMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYC\nEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgpM+G/HwJJOqz\nIT9fAon6bMjPl0CiPhvy8yWQqM+G/HwJJOqzIT9fAon6bMjPl0CiPhvy8yWQqM+G/HwJJOqz\neU9+6AX/DBJ+Nm8zkjJfRXf4Kyj32bzdSOp93/pTINlng5G+BCT7bD7ASHSRPaDSZ4OR/gk1\n8KrQdzfhw/goI31dZ3nhPDU06NdpczI+zkhf1WFeZyTFFem7wEj/zIvOGCN9FRjp3+AeCQxf\nY6TP7UhckV5Hd8mHoSq/zNtlqlRadl6jMlHJWDUX1UTQoO8vo5JdoVTR9EHtn/klIwViDPJl\n9bx12w7gH9d5sodNpp3Oct1m4EanqnWjpI8jz4ExUt/XyWiHxMp4nexRO40yU5P6xalh4jcY\nkjLlzYZugto/821GuvODAzGsfNW41SXb/d07rk3czIVNvJw8WGyobaDFOGVYcScyRuqNgsVt\n+LzlxhB6uxzy0pauk24qafomUTevOAUIa4dPq2SpCv0n773av/M7RgrEuKis05fvccjJNw8M\njitU2Y8yr5uEmbprpGRo2+UmnI2iik6f167IGKnXF43JL4UaZnfd7J/aGY5KU3nVQ6ZTNHRJ\n7jfQvWIUPFF2Bth7tX/nZ4wUipGZPtmO/fi6uSgQHjdNrFSsSZipew9kr+arOzuWavKnImOk\nQY18KrblkM1qGZRKbSwrZa705KHRbZ2iITdmcWuHCH62TXbC2r/wM0ZaSTR220z/bbeHnPC4\ncQq49H+3SZipe8wXw/ALnW97IvIJjZQ712e7PSt6s7ZQvTfyBYNgY43n1ja+/taPYe2f+Bkj\nhWJ4omaq3TowPK4ap3ZVrEnkcrVJqvoqMbM5j85a++nIJzTS6uUOFStvG8lekMLa5chhmlIG\nUf7hZP/x+N3fImqkeCNPjNQM8TdTV6nrPaX8XRe92pBcok2eMZJS+bRs5IX3Vpww0h1mTcZ1\n7CeN1Ix3o5tGuuTJPF6e1UjRVsGVJe/6xszozIRpt5GqZbUvmq3Nr19F1YsNhXtp6/WiYP63\nyBjpWSOV45C1aaR+XMRY1f7pZP/x+N3f8m4j2ecEZrEuTbr9Rrroy31XLDezbpPnjKRn9K27\noq3XkbIgAkbaxL1H0sJkyz2SczOZLMolnojTAw+/1pd4Xgs6vZGWx9ehGIMZkkrXFWZk2muk\n1EysO7//B7nZaaRI2yzdaPYw8gmNdHWeFGhhvFW7+UJvl2naZdWutSabvebVhhLH+84f+F0j\nGZrBEPPbIpuh1XojaOytrbW7Vu0iq91tmrVhs72RT2gk5zlSp2WKP0eqTGWtTeYU9dTismqg\nmUcuHWGeMJzeSF4Tt9YKdRl65nNGspeGLlgkGJsEObmLbds6i3S1u2D3bOQzGqke5ubmzYbS\nLto4bzZc50ZbbzbMi+XBs+75Dmq43HW52D3SL+GJYYS6pbPie6d2pdJvw5VBjx6brN8/2GYY\n7MyrFdelIuKj3ZFPmejb9K7dOMmrp00rqtUuNTVZUBzn6GHtku0kWgt9MN/trFB5uPPRceNL\nb1m8SRrbuUG1tDWHF9EL497IJ030NR/ymFXTjLir9Nvf1WgRq519y7gPik7C3VqnfqhNL6ta\nCA3RDh03r1c7Hx7Xe7IHTbrYzi3qbGprDo/PMPdGJtEAAmAkAAEwEoAAGAlAAIwEIABGAhAA\nIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAA\njAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAAB\nMBKAABgJQACMBCAARgIQACMBCICRfhmy+zaQ+pchu28DqX8Zsvs2kPqXIbtvA6l/mfvZJfeC\nIOYvs9dI9IJ/Bgl/maeNRHf4Kyj3y2Ckt4FyvwxGehso98tgpLeBcr+MlJHUF3eUV524ursJ\nP4Wskb6yr6gXnXUY9yvFgZ28wEhf1mHUi054FffLdIGneK2RvqDvvPA6ipFOxKuNpNTLpk4y\nYCSQ4E1GUg+/6iBeeWIYaaC75Eqp/DJvl6lSadl5jcpEJWOVU+z7y6RZUyhVtEGDzml7SZ3D\njuCTjBQRY1bS0zf4aotT48i+Hckcc/+mrtNxmgenuIo8nwxG6vs6GdOT2Hxcx01VO40yU5MG\nxSGPU1ZrG6LzGrTJErhcGhzE+420+Y0RMWYlPX19mpWR6oiqG5HWJnSxmXKcFMvX9jliJJ2K\n4jZ83nKVjNvl0PHb0nXSTSVN3yTq5hV7/Tlqlgy1Xa5Kr0Ght4eUFDoHRadHtOLtP3DmECNF\nY0TEmJX09F0dlwc1i+z7Im1pYHJUOuFj+boTGSMNQ9Hkl0INs7tu9k/tDEelqbyqyisOOmej\ntFeTy0570WkwXfeHj3wpHsXnGGktxqKkq2/IJax1ZN8XafuC1PknFMnXvcgY6bqMQm05jDDV\nMrqV2lhWylzp2ZkZEJ1iP7QdpS2WWYHTIBmzseQZI7mBnO+clXT1Dbmoi19RqCZssy/Sxgkl\nqxr3F+6PfEIj5cEcIl9Sc7NmUX3vXlqcYt/Mm6nqq8TMBtwG1Ti1m0euTmWv+ymP+DQjuWI0\nvrAbI06u6mK4yV8qHNmfixSlDH3q5+uJyCc00urlDhUrbxjJ3ZObe1O/bX/R97DJkp6Lt4Lx\nZj7NSIEYu4xkyJxjZtmfixThqrxbrdgpYqRtZk3GJZ0/G0nf9RbBjdFwSdLMF6Q22T/NkOfD\njBSKsaOTKnXVjxSWC4cjezxbW5EiXPIkvAUbT/H5yBjp70bSM8JWr4g6DS56jOuKKfFdcuDE\n7tOMtBJjd/fvlnVnR/Z/NlI/LjetTxEj7cC9R9LCZMs9knMzmSzKJb6Ia2mdBqlZC5oTn0Uf\njryN443kPslZiTHuSB53/62O/XykgM6fJK7ztTfyCY10dZ4ULOsDBmeRwC7TtMuqXTuZbF7H\nmbecBl6e2zQLnsC/mU8yUkQMb0WsvbfWtnTg+EOF/ZG2I8fztTfyCY3kPEfqtEzx50iVqay1\nyZyiZpTW1rb6PthpYEcuO87VRy7YGY430kxMDE/Jen3j309Pe5wO7Mj+XKStyMs1KJqvvZHP\naKRaqdy82VCqscPPbzZc50ZbbzbM0g4p6PTd0NVrUCr9slapBQ+TfWqiYoxK3nuzwSjZlctS\nmiP7c5HWkYeJSZcv90jxfO2NfEYj9bfpXbtxkldPmzY7Vrt0WXd1iv0yGaiiDbK5WExR3/fD\nPhdXjFmRqeDr69HZVJVL8yrWeEekFcnS1hwez9feyCfN8jUfZMyqaUbcVfrt72qc11nt7Gvc\nfVDsnX5QZ7EGc1FFE3NSXDFWRvL19dH70ovbfJHdDf840oogcjxfeyOTZQABMBKAABgJQACM\nBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEw\nEoAAGAlAAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATA\nSAACYCQAATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAA\nIwEIgJEABMBIEIWO8RzoBVFUtAhbIBJE+REjKfWas1dh4G8WCV7I2kjf2FVU/5rzVvMfrwYg\n5CeM9LJTxkiwk4dG+oae89pzxEjwmB8x0qvukWz0rQ2AiSeM9Ll9SL3qHqlfxf1cEV5Jd8mH\noSq/zNtlqlRadl6jMlHJWBUvNoVSRTu2vkxKXtK5gVP8OsSNFIqhLLrYOaKuGPZldVCxajyr\nP+20ode3MgE3b6eX0I3I83LdF16gxamTMYuJVe06bio3YZmpSbeLtQ0xmmqaQpRLbek2+Db+\naKTNmVQoRrMYqU3cZARYwauwIvUazeoHOx8ZqUvcnXUkX1uRV1HPaKRBsOI2fN5ylYzb5ZDE\ntnSddFNJ0zeJum0V+2QodrkqdeuhyirZqKLTw1jhFb+Qvxsp2qVWYjQqn4qF0bCMCnVRWdd3\nhWqmCjcDc6xJ/XDnIyPlnu/dhO6K7HJGIyWzXwo1zO662T+1MxyVpvKqx8J48WoU74wXh3SP\ngufjdV95xS/kH4wU+8ErMS7LVUbdESozXbddereTgSXUpP5q5+q5qcfV2+sm9FHkZVo6f9P2\n1/wq12UsbMshTZWbJX3bZBTKlZ5pmHEzXnRGyX6IEMiqYsVv4l+MtP2LXSPNd6jj/MrtweEB\nKpsqnAzMbWb1Izvv0M42MbgJfT7yd2b5n8i9aYHeduYNWii71rMMk/FiqvoqMROWQeDAL92S\neKf4VbzESJ1riLoY7t51sRqndlXkiNXVKnL5asLanWNXplq3pZvQ5yOf0EihGO52NGFbxdzc\nm8aiXpabrYu3gvE9vMRIjhi5nR0ZY130akNyiR2RmivB7a6RVrX7jFSpa5B7L6FPRj6vkeZV\noz8bSd+bFsuseYnSJnmk+F28wkiuGGroxn1n59JVuDK3UKm865ssbqQte+0ykpmm+blfEvp8\nZIz0dyPpGWE7rYg6UbokixS/jBcYKSJGp+W76KldV6joJcksjecvMFKadGHul4RipB2490ha\nmGy5R3JuJpNFuXgxkNaROEtjxS9DyEju8lZMDHu3qe9LOhXVajBYUjnqJtEePW7Gd0YpzCwz\nMoj+MfIJjXR1HlhoYbxVu3l+YZdp2mWpbl2cI7iffZtm7br4dYgbKS5GZEiK0CwmczIQhNnc\nGUXNLJFjp7E38gmN5DxH6rRM8edIlamstcnuFdtpHWrSv14WpuovXbAzSE/tQjEScxkyPdOO\n9l10+ds2uywd2MmA/7XbO6OsjRQk9MnIZzRSPUy6zZsNpV2jcd5suM6NHr7ZMEymzTP38ZBR\ncCcPYUpOzUqM0twYmeecQ7EbK1aY9x1u6ZKY2JsNs/rxnXfwloj8hD4Z+YxG6m/Tu3bjJK+e\nNq2GVrt0WZ6NF6ulOB80TL3ncc4pgiuG+dPZHBjzZJ6SHmMzc0Fa5WVhEjm68w7erLL6l8gn\nzfI1HxKUVdOkvav029/VOK+z2tlXkvvtYl9nczFcC7RTf4w044qx6JuOC3WOqCHtYMG8HmNM\nx60aTyJHd949LffDS+iTkckygAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlA\nAIwEIABGAhAAIwEIgJEABMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQA\nATASgAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEA\nBMBIAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASMfx3+uPeMNX/OGIXwQjHQdG+iEw\n0nFgpB8CIx0HRvohMNJxYKQfAiMBCICRAATASAACYCQAATASgAAY6SiSgX2tEqd1+Bk272ON\n7h20HLHvax7F3vezfg+MdBDJ/Odxs6V1+Bm2Tu41jh009vrdX/Mo9s6f9XtgpIN4hZESp35f\nJ0+mK9Ler8FIG2Ckg9jX4xL3c0d/ffqK5FfuvfBhpBUY6SB2Gmm6dxlbv8pIz30NRlqDkQ7i\niSvSE/31X65Iu7/mnuUwEryXJ3rc64303NdgpAgY6SC+10h/uIadAIx0EF87tUv+dGI/D0Y6\niP1GujeT2o76tJF2fk3iH4aRRjDSUex/s+HeZ9j86YOePCJ59A4EbzYAwJ/BSAACYCQAATAS\ngAAYCUAAjAQgAEYCEAAjAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIABGAhAAIwEIgJEABMBI\nAAJgJAABMBKAABgJQACMBCAARgIQACMBCICRAATASAACYCQAATASgAAYCUAAjAQgAEYCEAAj\nAQiAkQAEwEgAAmAkAAEwEoAAGAlAAIwEIMD/T8ovx1Dn6hYAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "topGOUp <- fgseaRes[ES > 0][head(order(pval), n=5), pathway]\n",
    "topGODown <- fgseaRes[ES < 0][head(order(pval), n=5), pathway]\n",
    "topGO <- c(topGOUp, rev(topGODown))\n",
    "# create the plot\n",
    "plotGseaTable(pathways = GO_term_hallmark[topGO], stats = stats, fgseaRes = fgseaRes,\n",
    "              gseaParam = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef50627",
   "metadata": {},
   "source": [
    "### Enrichment analysis using FGSEA and KEGG pathways\n",
    "We can perform enrichment analysis using FGSEA with KEGG pathway using the same procedure mentioned above. The only thing we need to change is the\n",
    "list of genesets that are available in the KEGG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "93b11008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$hsa00010</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'HK3'</li><li>'HK1'</li><li>'HK2'</li><li>'HKDC1'</li><li>'GCK'</li><li>'GPI'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00020</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CS'</li><li>'ACLY'</li><li>'ACO2'</li><li>'ACO1'</li><li>'IDH1'</li><li>'IDH2'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00030</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GPI'</li><li>'G6PD'</li><li>'PGLS'</li><li>'H6PD'</li><li>'PGD'</li><li>'RPE'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00040</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GUSB'</li><li>'KL'</li><li>'UGT2A1'</li><li>'UGT2A3'</li><li>'UGT2B17'</li><li>'UGT2B11'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00051</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MPI'</li><li>'PMM2'</li><li>'PMM1'</li><li>'GMPPB'</li><li>'GMPPA'</li><li>'GMDS'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$hsa00010] \\begin{enumerate*}\n",
       "\\item 'HK3'\n",
       "\\item 'HK1'\n",
       "\\item 'HK2'\n",
       "\\item 'HKDC1'\n",
       "\\item 'GCK'\n",
       "\\item 'GPI'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00020] \\begin{enumerate*}\n",
       "\\item 'CS'\n",
       "\\item 'ACLY'\n",
       "\\item 'ACO2'\n",
       "\\item 'ACO1'\n",
       "\\item 'IDH1'\n",
       "\\item 'IDH2'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00030] \\begin{enumerate*}\n",
       "\\item 'GPI'\n",
       "\\item 'G6PD'\n",
       "\\item 'PGLS'\n",
       "\\item 'H6PD'\n",
       "\\item 'PGD'\n",
       "\\item 'RPE'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00040] \\begin{enumerate*}\n",
       "\\item 'GUSB'\n",
       "\\item 'KL'\n",
       "\\item 'UGT2A1'\n",
       "\\item 'UGT2A3'\n",
       "\\item 'UGT2B17'\n",
       "\\item 'UGT2B11'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00051] \\begin{enumerate*}\n",
       "\\item 'MPI'\n",
       "\\item 'PMM2'\n",
       "\\item 'PMM1'\n",
       "\\item 'GMPPB'\n",
       "\\item 'GMPPA'\n",
       "\\item 'GMDS'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$hsa00010\n",
       ":   1. 'HK3'\n",
       "2. 'HK1'\n",
       "3. 'HK2'\n",
       "4. 'HKDC1'\n",
       "5. 'GCK'\n",
       "6. 'GPI'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00020\n",
       ":   1. 'CS'\n",
       "2. 'ACLY'\n",
       "3. 'ACO2'\n",
       "4. 'ACO1'\n",
       "5. 'IDH1'\n",
       "6. 'IDH2'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00030\n",
       ":   1. 'GPI'\n",
       "2. 'G6PD'\n",
       "3. 'PGLS'\n",
       "4. 'H6PD'\n",
       "5. 'PGD'\n",
       "6. 'RPE'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00040\n",
       ":   1. 'GUSB'\n",
       "2. 'KL'\n",
       "3. 'UGT2A1'\n",
       "4. 'UGT2A3'\n",
       "5. 'UGT2B17'\n",
       "6. 'UGT2B11'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00051\n",
       ":   1. 'MPI'\n",
       "2. 'PMM2'\n",
       "3. 'PMM1'\n",
       "4. 'GMPPB'\n",
       "5. 'GMPPA'\n",
       "6. 'GMDS'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$hsa00010\n",
       "[1] \"HK3\"   \"HK1\"   \"HK2\"   \"HKDC1\" \"GCK\"   \"GPI\"  \n",
       "\n",
       "$hsa00020\n",
       "[1] \"CS\"   \"ACLY\" \"ACO2\" \"ACO1\" \"IDH1\" \"IDH2\"\n",
       "\n",
       "$hsa00030\n",
       "[1] \"GPI\"  \"G6PD\" \"PGLS\" \"H6PD\" \"PGD\"  \"RPE\" \n",
       "\n",
       "$hsa00040\n",
       "[1] \"GUSB\"    \"KL\"      \"UGT2A1\"  \"UGT2A3\"  \"UGT2B17\" \"UGT2B11\"\n",
       "\n",
       "$hsa00051\n",
       "[1] \"MPI\"   \"PMM2\"  \"PMM1\"  \"GMPPB\" \"GMPPA\" \"GMDS\" \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "KEGG_hallmark <- gmtPathways(\"./data/KEGG_pathways.gmt\")\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(KEGG_hallmark, head)\n",
    "tmp[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9eb1ede2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running fgsea\n",
    "suppressWarnings(fgseaRes <- fgsea(pathways = KEGG_hallmark,\n",
    "                  stats    = stats))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cbe55bb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>hsa04060</td><td>3.181473e-06</td><td>0.001081701</td><td>0.6272567</td><td>0.3356103</td><td>1.457547</td><td>293</td></tr>\n",
       "\t<tr><td>hsa04740</td><td>1.058477e-03</td><td>0.179941158</td><td>0.4550599</td><td>0.3409007</td><td>1.414354</td><td>132</td></tr>\n",
       "\t<tr><td>hsa04061</td><td>2.497384e-03</td><td>0.283036864</td><td>0.4317077</td><td>0.3528680</td><td>1.439385</td><td> 98</td></tr>\n",
       "\t<tr><td>hsa04080</td><td>7.547702e-03</td><td>0.639029354</td><td>0.4070179</td><td>0.2778361</td><td>1.214380</td><td>358</td></tr>\n",
       "\t<tr><td>hsa00500</td><td>9.397490e-03</td><td>0.639029354</td><td>0.3807304</td><td>0.4184500</td><td>1.539228</td><td> 36</td></tr>\n",
       "\t<tr><td>hsa05340</td><td>1.195356e-02</td><td>0.677368355</td><td>0.3807304</td><td>0.4011639</td><td>1.480033</td><td> 37</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t hsa04060 & 3.181473e-06 & 0.001081701 & 0.6272567 & 0.3356103 & 1.457547 & 293\\\\\n",
       "\t hsa04740 & 1.058477e-03 & 0.179941158 & 0.4550599 & 0.3409007 & 1.414354 & 132\\\\\n",
       "\t hsa04061 & 2.497384e-03 & 0.283036864 & 0.4317077 & 0.3528680 & 1.439385 &  98\\\\\n",
       "\t hsa04080 & 7.547702e-03 & 0.639029354 & 0.4070179 & 0.2778361 & 1.214380 & 358\\\\\n",
       "\t hsa00500 & 9.397490e-03 & 0.639029354 & 0.3807304 & 0.4184500 & 1.539228 &  36\\\\\n",
       "\t hsa05340 & 1.195356e-02 & 0.677368355 & 0.3807304 & 0.4011639 & 1.480033 &  37\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| hsa04060 | 3.181473e-06 | 0.001081701 | 0.6272567 | 0.3356103 | 1.457547 | 293 |\n",
       "| hsa04740 | 1.058477e-03 | 0.179941158 | 0.4550599 | 0.3409007 | 1.414354 | 132 |\n",
       "| hsa04061 | 2.497384e-03 | 0.283036864 | 0.4317077 | 0.3528680 | 1.439385 |  98 |\n",
       "| hsa04080 | 7.547702e-03 | 0.639029354 | 0.4070179 | 0.2778361 | 1.214380 | 358 |\n",
       "| hsa00500 | 9.397490e-03 | 0.639029354 | 0.3807304 | 0.4184500 | 1.539228 |  36 |\n",
       "| hsa05340 | 1.195356e-02 | 0.677368355 | 0.3807304 | 0.4011639 | 1.480033 |  37 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway  pval         padj        log2err   ES        NES      size\n",
       "1 hsa04060 3.181473e-06 0.001081701 0.6272567 0.3356103 1.457547 293 \n",
       "2 hsa04740 1.058477e-03 0.179941158 0.4550599 0.3409007 1.414354 132 \n",
       "3 hsa04061 2.497384e-03 0.283036864 0.4317077 0.3528680 1.439385  98 \n",
       "4 hsa04080 7.547702e-03 0.639029354 0.4070179 0.2778361 1.214380 358 \n",
       "5 hsa00500 9.397490e-03 0.639029354 0.3807304 0.4184500 1.539228  36 \n",
       "6 hsa05340 1.195356e-02 0.677368355 0.3807304 0.4011639 1.480033  37 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][, -8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a63408eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2di7qyKtdA2R4rM73/q90CHgDxUGv2ajXG/+wWIU5pylCk\n/u9VLQD8GXV0BwC+AUQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARApEWUISkeQX1lNx7QI5fmmne9y69C4d75eZSqnIOoEV3VlKlSmdSneNvn0GHXQx89Hk7M\ncL4T36RU2Y2H9GmkSuK9e5X3ipQ4B/FEavqPkTRixxIKNA+LSC/SXzIzVUaqDxapUqq4d3/v\n+TBI/8h7RVKX6SDekQqVdReCR5jjvxxLKNCzYRFpkT55TTBUTyFSMsyWuqEoMi96r0ipeowH\n8Y6klLkVNWKHR6TzMSTP/K26J5KkbPuZiaksVXJp76rQbSo7sAt1H1s2KjW7D38lual8KD5K\nfWdqr6lKrra3j1x3zKscPkiT6h3Hz+I0th+21LeOKuueWqpWDqVq29+YSM8H04l38pt2Lrqf\n6PVe9oFbN0NdTpLudsnU7g+4d6SLncqXjkj6Wb+7GySmVWHnJl3TqWWuzBC/9fMaSfrQboUm\nM702jx0Xr3L4QLnu2NRDp7H5sKVufLWb5Z7/dezCdHguUqlmazmbwS79p8rMbe7RFb1P9Hov\nh8BuvO4oyuQNkV7GJs7O35W6aSWmgdBlvOnGXNol/dYOz9PamallZW9WhZJZDph3baLSveme\n5iqnY27lsFfWtMFnGRrrt8aj7spQ682Ct9Eutr19+IsNZls3TtPyvrp7GCyp2zrpPoO9Ql26\nzzc7O6/1cgjsxrv1tYj0OuOqXTPVtJNI/SVWXxG7CV6uh182OmMapeYB4A0zu+GcjiMyN0dq\n9ARq7JhbOex1D0JMjc3EJrMbJKd1w6Gu+hYXEamtCp3i/Ye03avMR9WZTdW0pf2bSGNgJ569\n91eI9Af875Ee1SXzRBpes268lt1TwMU65bS86kvm/Q0zu5lIzuCcOhaMWGckRD5Ld2/qPSu7\nmUxdi/dWX1UWhvr9koRT1Y1g9o++1z/sBCz4RC/3cpyHBPEQ6Q94icvGMRmKVHWmJGmbDrO8\nqWX/dCU/s/OekZ4XKfZZ9CWjv3de9HOT0NdT03H1qsziUK/337adka2vUaXJRPiJXu5l/2cW\nD5H+gJu4QqXX6hEVqZth3LvLYqmXxBqvZVdXacHkudnHr7GbTlddN+IfKPpZuttRbR4NNFWZ\nCj8jtVr+ejbUx/L+8e+O9+76leh+zj7R673sb3VhPET6A7MzviBSqYpOmKp7LfyW3YU2q98x\ns3O/R2rs45CzojC85uHDjjsG5yLpm2cSthWhny2pdDbU835xMPyubjWYfWjRuS6VmT3PP9Fr\nvRwDO/FsEu+I9Ad8ke5tPcyaH60nUpdlfStS9mnVaakfDJJ3zOzM029uftlQmvVCs7bUPZPl\nbsecSu8DeZ9lqO+fYy765Sa/aqe5zCdfXequXeru2f7V9mFxTef6YTM//0Sv9XIM7MSrWLX7\nM27iyv5x464Hmh667n3JzoMye1V1Wprx/o6ZXcd9+K2dneTZSb1+sHE6NlV6H8jpYSBSrcfl\nzem/DEMik36oO89uQ1+ytf2DYPabHfMmtTvOPtFrvRwDu+fQfBlXINIf8BLXpTK7m7XRexqK\ndDErR5f+92JTSzPvesvMTnPLu4GZXXpNrp3fRXCrnCr9DzT1MHygupgfPuhfNgh6NB63movU\n1oX+FLenguXdI4x9c+tnr+EnerGXY2D3HF74ZcPxVOo9M7sfRvLh7bnAiHQcT8z9YR+HifRA\npKN4au4P+zhKpFI5XznEdhftDLgk04IZSHGUSLn5/5xa2V20MwA/CiIBCIBIAAIgEoAAiAQg\nACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIg\nEoAAiAQgACIBCIBIAAIg0nEoh+mfBmvdurf9r7iBMJyo49glkpXJbzn9b9DDWeB0HMcfRXJv\nW0MEOAqSfxzyIk3vOK//GBJ+HG8Vafi3iDjD/wbSfBz/SqQhxtGf96shu8fx70VS3KPeBUk9\njuNEGhuAFGTzOA4Xyf4fSEAej+MsIvWN4S+QweM4mUjDTvAKZO44zijSsAGehJQdx4lFsu0Z\nHPshV8dxfpHGfWELknQcnyOS+tjp3rsuA7Pry2em5zv4QJE+bbio9j19nifj0zLzTXykSLY3\nxyZuN2/rJyKdiQ8WaTjAyXlvFxHpJHy8SGra+Zyotz0jzQKfNwnfz/eI1Ic/Hepdz0g2+uIb\n2J7+l4lKymZ4d1ez2rpQqnhE2kaO9VUiqbXBdJ1tG1Ln5CtGozfXXqh0ntUx/JDy8SOvDvFZ\nr+bdDCP3yZoFRiSfLZEyM4TS/l2TqLC2MsWkmbWNHevrRFqyqZ7ldUidk68oidnumFRGdhjD\nBynfEGnWq3k3FyKz2LDFhkh3ldRtnai7fZvb5m5t0hWbXJWztrFjfaVIkQx2WQhr+9Q5+YpS\nqkK/5FMoVTT6PlHEwocpXxdp1qt5N5ciI9IWGyKVqupeb+pi3t3sCHRrb2ZINCoJ20aP9SMi\nXVUW1g6pc/IVJVGNf1ZyNTtPU/hZyvuPFmXWq3k3lyPPAiOST5eeUiUmXVWXxKwype7ymZhr\nZq70bL62V8jHkGantphmIV7b+LG+VKRwUHWyBMNuTF2h6nabuWduvCn8ZspXezXv5hOREclH\nqVwPjau+Gqm+dLGlsh3Pn/2TqYctOLWpai+JmX34bePH+laRwgnSLAtj6px8LVPqs+DRqCwW\nfjPlq72ad/OJyIjk092Ems6hVE8qan0vT3XdzcxFWj+fl656JlIvYtIiUvBZ3Xdu6sZ8LXFT\n8yeoq5lpzcI/JVKs4XxPRHoNZR4n7cCInCsnn+Y2HxFJPzwXei79yyItjUeLl7oxX0tc8yTc\n/EhyP6i0SM9HRiSfyYyyu1jW/QT+UV2yUKQ0aaIi6V0e9kbmRIwf60dF8lI35muFwp/bNUkW\nBEWks+GYcdFfYST6GTMbR2Yy5rMwNyzbfKp1E55s5v5HRXJTt2/oB6t62Uy7fv/NlK/0amHP\nvZERycc7vVWZ6ktlodJrZR+O7eLNo5uaOMPOqXVXZ53apWP9pkh+6mabl0+L5ZFmsx9C9Ns3\nU74SdqEfeyMjko8nUuvcZaxIF3Mxrcyy6DQaptq++NCLSk7t0rEQyclXFPs9kjv1q2Jt+/Cb\nKV/pVbxid2RE8plESu1anXnYube1fUaafXVumju13Slv9MPzbccvG36JpRHq5CuK+WVDk0/P\nSHHn+vDPpny/SNs/U9l5yF9hEulmL5r34ddddj0vNaXMb+7WXqZi0PanGUaof69vvXxFSabt\nZo9iuptFwj+Z8lmvlm+lW5ERyceZ2plfNpgrUKELlZkfN+ZHwEFzr7bbrS8GbX+aRZGcfMXp\nkphepz3UqkhPpvwJkbYiIxKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKA\nAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAei7D+J2tr/bI3qq1oV\n+QdN4bRwrg5kj0j9dmUam9M1/XPRcB44HwfyN5GGHft/6Bu3DoXkH4i4SOav3Qf+LWT8QN4p\nkt3I+f1XkOgD+UciqaE5vA+yeyD/WiRuUe+DvB7IUSLhkzwk9ECOFml6gb9CGg/kHCKxxicB\nOTyQU4nE97x/gsQdyDlF4gnqFUjZgZxZJAbGc5CvAzm9SDi1G5J0IB8j0seOkjeupKjVt/Av\n+SyRPnB1b/pg8qER6Tx8pEjqg25Q7xNploSPyck38rkifdKweVNfEek8fLpITjfPCs9IvwAi\n/Qu4Ix3B5qgoE5WUzfDuPo5F5SwQ3VWsbeRgvyPSddborpa3bew732GsGVLunI2V6H6gJna+\nwshjYERaZWtUZGbQpv27JvFESrzaoG3sYD8jUj1rNCQptm1j3/kOY02Q8umD7Yj8SOxJfOyJ\njEgbbJzUu0rqtk7U3b7N3eZVUBu0jR3sV0TqshA2GlMX2ba+73yHsSZM+bpIQaBCld1rqYo9\nkRFpg41RUaqqe72pi3l3cx9lmyRvvVq/bfRgPyLSVWVhozF1kW3r+853mGpmKVdqOXgYaJiy\nqViT+clEpFW6vJUqMemquiRmlSl1l89EX66666i+8dfKOPPwzkSu7Ox6rPXaxg/2pSKFFd2l\nPmg0pW6+bWPf+Q5TzXbKVwL1c81hgv5kZETyUSrXI+uqr0aqL11sqWyDq1amHtOZqM12tzZy\nhQsP9iMi1bMsTKmbb9vYd77DVLOd8pVAl35qd4k12YyMSD7dTajpHEr1BarW9/JU193MXKT1\n83npqqfEDjekqRaRvLy67/zUbY372fbIEV8QadbwqlcbkutrkRHJR5nHSTOezKzY3dJ6+TS3\n+TGxdf+M6tT+skhLI31Ilpu610WahfibSHbmcXktMiL5TA6U3Syvrm3to7pkoUhp0jiJLXvt\nnFpEmuW1x0/dWUS66qldU+jJPCL9GceBi7nT62fMrB+q0wOpagtjzpjYfoNbm2ye1R8VKUjd\n6yLNarZTvhIoNZPzJvjib29kRPLxbiZVmeq0Fiq9Vvbh2C7ePLqpiRrRLYflHLd2art4sN8U\nyU+dpEjbKV8JFL/p7I2MSD6eSH3JvLMiXczFtDLLou5ouKpr33yqndouHgyRwm3RNO0WaTvl\nK4HsTadxl7+fiIxIPpNIqV2rM6t297a2z0izr877/OWqngXZ/mXDD7E49qPbNvZdDPZsyr1A\npdK/sysDV/ZGRiSfSaSbvWjezbLDUOz00mR+82F+HdQGbX+a+d3neZHGdstWPplyP3IW23lv\nZETycaZ25pcN5gpU6EJl5sf2F8JB8/DU9u+Ctj/NPxLpyZQHkWM7742MSAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKA\nAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiA\nSAACIBKAAIgEIAAiHYoy/0SpfjV/h1fV/8fp+Rg4U4eyLZIa/nloU7DN7a6HdRoicDoO5c8i\nqT6G2eOYzwAakn8o8iIplDoEUn4obxbJbPnHH+lHIc2H8k9E6nXiVL8Tsnso/1ik6SAgDEk9\nlMNE4jFKGPJ5KEeLxOmXgkweyjlEYhD8HXJ4KGcSiWXzv0DqDuWUIjEmXoCkHcppRWJcPAkJ\nO5RzizT0DbYhS4fyISJ9rEzqfdcBtfoW/i2fJtKHDZfxU7whNCKdCUR6K+8Tafb7kA/LzLfx\nqSKpDxo5b+ooIp2Jzxapb3TmMcQz0m/wLSKdeRhxRzqCzQtYmaikbIZ3dzWrrQulike//boR\n7mdEmifirmLFyJ6pm/A+E+FpGsMPJ8Jpshg8jNz45zYeeQyMSKtsiZSZs5j275pEhbWVKSa9\nVFvhfkWkeSKG1PnFOaWbTxtpLtIY3j8906fYE/mR2IqH22gpMiJtsDHy7yqp2zpRd/s2t83d\n2qQrNrkqdX1X9YMiRT5yJBH5VJOvZKlWRaNvC8VUkS+HD07PqkizyIU5aeVUsRYZkTbYGPml\nqrrXm7qYd7d+lDu1N3M2GpW0+iRliNTGEzGkzi/OydX4QYZYl+Xw/ulpvRneZmQVVqxGRqR1\nuryVKjHpqrokZpUpddfMxNxjcqVv/P1V8TGk2aktVD3FKjdnir8h0jwRj0mtx/blxhfpuhze\nOz37mCL3E0xzDXw+MiL5KJXrIXvVZ0z1pYstlW1w1crUYxi8Y22q2ktiJg1d2jdvcN8qUvCp\n54kYU+cVF2lUNhRzVRX9VW0ePnJT2R/50k/tnDveE5ERyae7CTWdQ6m+QNX6Xp7qupuZgLR+\nPi9d9UykXsRkDLd1uF8QaZaIKXVucZmrmVgZcntVy/wGL4vkRG6verUhCW54iPQayjxO2sFT\nBVtaL5/mNh8RSS82FMNl7VdFmn1sr8ZJnVNc5pFMMypzVWtKM8Gb9ntVJDdyP/O4+BEQ6TUm\nM8ru5lL3DzyP6pKFIqVJExVJ7/IY1kkRyctrj5M6p7hIk2SzKpPgP4vkRb7qqV13DXxNUUTy\nccy4JMO3Ctn41UUy5rMwNyzbfKoNE45IXl4tTurcLC6SpfO6YJf+rXMiduFFTpV+sm2Uf7S9\nkRHJx7vFVGWq01qo9FrZJ2K7ePPo5iNqxK0NF1URycvr+KGH1LlZXOCRZo95bVwk50TsIIgc\nv+nsjYxIPp5IrXOXsSJdzBW0Msui0xCYavviY3gYRiQvr+OH3i9SFawrJOa+EY7ofn/nRGwz\nj6xfG3f5+4nIiOQziZTatTqzandva/uMNPvq3DR3aruno0ZPtG/udlhXa+1+FK7PleZJpoyt\nBMV+f7BMLHLTx38hMifaZxLpZq+U9+FHWXY9L/UXX/s0O7UXrwEi9cznuvtEKpw7lnlp7C/i\ngjvDECE4PWvMIvePwvGF9c3InGgfZ2pnftlgrkCFLlRmNmF/IRw092q73aYGiNTzqkhqNtx1\nrtPw1w1DhOD0rPcojNzGdt4bmRMNIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKA\nAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACINJxTP+Kqf932sLp\n+Rg4U8exSyTl01q7FP/K88ngfBzH30SabeJUHgnZPw5pkdy6oz/bz0HGj+OtIrmv8H7I8nH8\nK5Gm+PA2SO5x/HOR0Ol9kNbjOEykMQyIQTaP42iRxvfwd0jjcZxHJHz6M+TvOM4l0vgGXoHM\nHccZRRoOeXRuPg4ydhxnFomfIT0JuTqO84s0tYANSNJxfIxIn3t7euNlQK2+hX/Ih4nUb/0k\n1PgiHxqRTsNHitT/9xm8T6TZJeVTUvKNfLRIn3N3elM3Eek0fL5I5789va97iHQavkWkcY9z\nwh3pCDbHQ5mopGyGd3cV1jZOA79t5FhfJZJaS97V33RNncxcV3LudMGvWQg/pNxpshi9LpQq\nHn7dPWwdRh4DI9IqWyJl5iym/bsmUUHtIzHF5DFrGztWMPy/QKSF/NV+fWmT1MS2xXOUOJGG\nXkfCBylfXWyo3E70DCd0MzIibbAh0l0ldVsn6m7f5ra5U1uostUDpZi1jR0rGP5fIVIsgV0W\n3OpaFY2+1heRbVGqKYm1ypfDhylfFSnp2ja5OV8jeWjoUmRE2mDjpJaq6l5v6mLe3fph49QO\n930Vto0e60tFmk+PMi+v+ZSk2bYYTTLJc53ncwoxS7mKi92aVlqhZrrXtdMJ3REZkdbp8laq\nxKSr6pKYVabUXagSc+XKlZ6z9VfFx5Bmp7afGujT47WNH+tHROqGbGw820PHt3nkapp/XdV1\nFmcMsZlyh0LVYdUjlPqJyIjko1Sux8RVnzHVly62VLbeDUdPmx+24NRe+qndJWgbP1Yw/L9V\npDqahUZli9v83d3pV66qor+qzcNvptwhVe0lMXPMkfGEvhAZkXy6m1DTOZTqe0ut7+WprruZ\nu37r5/PSVc9Eaq96tSG5togUfNZZ1dXMllYzZHFvSG1uj5VFwz8jUn/JdGZ20wl9ITIi+Sjz\nOGlHShVsab18mtv8XCR7+7q0iDTPnsdjfPDZGPe1XZQYG9/0dwzKuVTNcr1TJL3YUExPPc4J\nfSEyIvlMZpTdJavup9GP6pKFIqVJExHpqqch3em5IlIkrw5Nki1u8ymDK5rZ2yxE/1EkfXIf\n04q2c0IR6c84ZlzMJG36QkjXJWM+C3N6bfOptpt562lIY+eGTsT4sb5WpNlnnmUhS5e3+YRf\n7cT26d9upny+y9jWPaEvREYkH+8WU5WpNqJQ6bWyj6F28ebRTQLcUTbVuqfHqV061s+K9Eiz\nx9K2gPhKWVykzZQ75IEa7qd9ITIi+XgitcPTUqtzaJ+A9GWrMsuiU96n2v7KZb6dcGqXjhUM\n/58RqfIWC9ZFCta7E3PLD0d0H2Iz5Q627WPsiftpX4iMSD6TSKldqzOrdve2ts9Is6/OTXOn\ntlT6l16lTvjmLxt+CW94PvxFt3WRcv/7HpPaJnxu6kM8k/Lu6ajRT7O35X4+ExmRfCaRbvby\ndB9+GWbX81JTCp+TndpsKgZtf5phgJq/hX/tXxcpHRe/TbvG/pgxuDMMIZ5J+WVq689AXoqM\nSD7O1M78ssFcgQpdqMxswv64O2ju1U7FoO1P40mjnhEpHOQ6q2n464ah0VMp705w33aHSFuR\nEQlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlA\nAEQCEACRAARAJAABEAlAAEQCEACRAARApGNx/vVS57X1K+H8cKaOZVskZeqcZpyyM8JZOZY/\ni6RilfDP4QQci7xIyosK/wgyfixvFIlT+y8h28fyr0TCrDdDco8Fkb4EknssiPQlkNxjOVgk\nTr8UZPJYEOlLIJPHgkhfApk8lrOJxHh4ERJ3LIj0JZC4Y0GkL4HEHctniPSxo0Spt/1aan6m\n4EAQ6Z2o8UU+NCKdis8U6VNGzftEmv3e/lNS8q18qEhvvNSL86ZuItKpQKT3wjPSj/A9Ip12\nJHFHOoLNC1iZqKRsgqKyLDfYcbzPF2ktdVe/N06+Ou5re3ZJzKqgbrbDGH5IuRN+MXhdKFU8\nnow8BkakVbZEyswISP1i7QyMaIM9x/sOkRbSV3t5rX2RmmQl6TaJF69utsMYPkj56gy0Mm0T\n70K3OzIibbAh0l0ldVsn6u4Va5WvN9hzvG8WqcuCL1Lubs1Xkn5VWdM2harXdhjDhylfFSnp\n2ja5Kl+KjEgbbIhUKj3LuOkrpFO8TlfMeIM9x/tikTobvLxevZzc1hYEMjN0H95wD3eYws9S\nvvKF7M3EbFTyWmREWqfLW6kSk66qS6KdnVfdhSoxpzJXek5trqhO8aquw/7xBqvHG0rOa+tU\nfoFI3ZANRLpObx6BZbEjtSpb2WEKvyvlPcFd7q+REclHqVxfxa76bKu+dLGlsp3Oq/KKuaqK\n3rR4g7XjjSXntXUqv0CkOsiCky99z3nsEMlpMduhDpvtWvJOVXtJVOE+Iv0lMiL5KD0nv+qH\nykRfsW66pNTN3PXbZZEM2WKDteONJee1dSq/QKS2DUUa89VdpW5rKUrNneDutIju8IJI/SXT\nmdn9KTIi+SgzJ7ceVMGWtl3wxJjWlPruhUhxvCw4+TKTpZUUXVTetLUz43J3cJL3ikh6saGY\nnnr+FhmRfPohovTjpcrrfhr9qC7ZikiWxt69lhssH8+UnNfWqfxCkSwmX2nSrKco0bcNZzHN\n3eGPIumT+5hWtP8WGZF8JpH0DLq7aumZRTZ+7ZFM+UxmqQ1q5w0Wj2dKzmvrVH6tSLquMLf9\ntRR194zkMrVY2KGv2JVyfxepyIjk44jUtlWZ6gtWodJrZR9D7eLNY1qUezjrOPZxaaXB4vFM\nyXltncqPFGkeNy6SGoked6Ae7xsLO/Rvd6W8Jw/U+GNkRPLxRGqdOZoV6WIuW5VewXOKiWra\nPsvxBlvHMyXntXUqv1CkKV+bItmm13EArw/3XSnvsW0f48L6HyMjks8kUmrX6sxzz3143o3/\ncKHU+W3Md3bP/rLhV/CGp5Ov+UafUhVdJs25WIrmVDyT8u7pyPxmQigyIvlMIt3s5cl4MhY7\nvcaF26nYJOMXTfEGMAxQ89fNl7sxQt8095stDfenUn6Z2gpERiQfZ2pnftlgrkCFLlTmbDbm\nR8CmaVBMr7HaZM8s4wfwRHLz5W6M8eiSn1dBs8Xh/lTKuxPctxWIjEgAAiASgACIBCAAIgEI\ngEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACI\nBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgA\nAiASgACIBCAAIgEIgEgAAiDSp8KZOxWcjk+FM3cqOB2fytaZ48z+U0j3p/KESJzk90OOPxVE\nOhXk+FNBpFNBjj8VRDoV5PhTQaRTQY4/FUQ6FeT4U3mTSAyI1yBvnwoinQry9qkg0qkgb5/K\nvxbpQ0fKW7qtOv7FceAfgEh7mA14kaDjS1AHHwgi7UC9pduI9E0g0i4QCdZBpF0gEqxzNpFO\nOpIQ6Rg2H07LRCVlExY77tOeQ9FvIMyniRQkQw3027IqHtVtFq0wXIe3w1HGFhvduvrbY2ds\nFhuRdrAlUmZOYhoUO5pk3HMo+g2k+TCRwmQMPiTjtks0qtMsXqGph9M2T/l6t2r/fMfOWCQ2\nIm2zIdJdJXVbJ+ruFTX5tGdfDBqI93T/9hOItJCMSldcVda0TaHqxb2r6H4jXVy1dJTVbo07\nLncyFhuRttkQqVR6BnLTV0+naP6Oew5Fv4F8T/dvP4FI8WQ0Sd7qS70euQ9VLu1smy1WdCL2\nyY8cZa1b046LnYzH5gvZTboElSoxmay6FNqZe9XdYxJzmnP1aPXdPveKehSMp2Qseg3e0NP9\n20q0zRsAABZcSURBVE8gUjwZuTKPHP3TTLa8c7NW0RnYh3gy5dOOi53cHRuRfJTK9eXmqq9F\nqi9dbKlsp3OuvKK+qD6mqXRf9Bq8oaf7t59ApGgyansT2kpUHd6rgoo6DLE35XXQNLL77tiI\n5KP0fP2qHykTPWe/6ZJSNzNhaxdFunQt+vdTEZHcrbFk9DeW1Fzr74uJ2rghzcM/kfItkXbH\nRiQfZebrVpQq2NIuiWTu9/a9U0Qkd2skGbUqzN+Lypu2zpYSNTSbV7iL4otHWWdRpGdjI5LP\n5EDZzfLqfiHpUV2yFZHSpBneO0VEcrdGklEOl6pEz5zzpUSVwRVtqkCk8+I4cNHnN9Gzjmz8\nCjCZsjkVC3NizXan6LZ9S0/3bz9QpCFzsWSM37w1hV7fWUpUEtbPKsa4z6c81qH41G4jNiL5\neDeTqkz1M1Kh0mtlVxDs0s1jWrXTxfHLduUW3bZv6en+7ScQKZKMYAGsXvjqerZOFls481bW\nnkl5ZNUu3H1fbETy8URqnUnco19V0HecSq8aTcUFkZy2b+np/u0nmNpFknHVK6KaxKwdXBeG\n6NhssaIdT9jzKfdEiu++LzYi+UwipXatLrULEP2z8PIvG9xTYotv/mXDZxFJRj78lKHUiwd3\nk+8IefiLh1lFOyb/+ZR7IsV33xcbkXwmkW72znI3yw5DsdNLY746dIrTnm4xaPDbOMkY09Mv\nYjdmsWFpzjQ2m+3nMCT/6ZQPO/ZLRbHd98VGJB9namd+2WCuP4UuVOZMN+YnwKapU5z2dItB\ng9/GScZs/evRJTiP//p7efks2ujplPsiRXffFxuRAARAJAABEAlAAEQCEACRAARAJAABEAlA\nAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARA\nJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQC\nEACRAARAJAABEOnbUOML/EPI+LeBSIdAxr8NRDoEMv5tINIhkPFvA5EOgYx/G4h0CGT820Ck\nQyDj3wYiHQIZ/zb+JhLj4UVI3LeBSIdA4r4NRDoEEvdtIJLHWz6Q6vgXx4EDQSSX2YAXCTq+\nBHXwRSCSg3rLB0KkX+CfiPQ5a+yIBK+BSB6IBK+BSB6IdAybD6dlopKyCYrKYmqv6dig4/7P\nE3xeka7+Dk4mZ9t8wiTOG481Q9DxdGx0cq1PS7ERaQdbImXGmNQv1o5IpSkl/cloEkQaqP3c\nOpmcbfMJkzhvPNZ4QaeevtSn5diItM2GSHeV1G2dqLtXrFU+NKhV0eiLWGHf5m9Zfl3lrCJ1\nqXJ3cNI32xYQJHHeeKzxgjo9faVPK7ERaZuNgV+qqnu9qYtXvOoXS67cMLf5F3dv56QiXVXm\n5cJJ32ybT5DEeeOpxg3q9vSFPq3G5gvZTboElSoxmay6FGaVKXXXxKTUpVw92v4O5BSv6joP\n0/FYHSBv4qQiqdK/SDnpm23zCJM4bzzVuEH/2KdnYyOSj1K5vtxctR2qL11sqWzHO43+4xRz\nVRW9aZZGZfpPph6I1FMHd3snfbNtHmES541rP9r+nzOs9enZ2Ijk092Ems6h7pEyUbW+k6e6\n7mYmGO2ySIZsjHI104BLtx8iOTutDNrFPMWSOG/8mkgbfXoqNiL5KPOkaUWpgi3tkkjGtKYc\nJ3iPRN//zSwAkZydXhApmsT5TvIiPRsbkXxsnvRr2c3y6trWPqpLtiKSpRnWTZvE3JvSpEEk\nb6cXRIomEZHOzyRSe0n0F0L6CTMbvyZKpmwms8SOXzgYoQpzR/t1kZxvqv1cBOlbyFM8iYtT\nu/k52e7fcp+eio1IPo5IbVuVqb7LFCq9VvaJ1y7dPKZVu4ezitMv1aXZw74b+MefYHx5eW/B\ngyyKFKRvIUvxJC6KND8n2/1b7tNTsRHJxxOpdSZxVqSLuUJWegXPKSaqaYccV8OaAyLNdnJ3\ncNI32+bu8pRIQdA/9ump2IjkM4mU2rW61C5A1PYZKf7LhlJntzHf2D2ctbspHmi8XAS/IljN\n0+6pXeynCa/36anYnGefSaSbvRLeh5/P2fW8dFronopNMn7RVARXUESa8Kd4TibbfSKtJHWo\n8YP+sU9PxeY8+zhTO/PLBnP9KXShMjO3xvwE2DQNiunV7olIS/iD1klfKyaSH/SPfXoqNucZ\nQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE+h0412+E5P4O\nnOs3QnJ/B871GyG5vwPn+o2Q3N+Bc/1GSO7vwLl+IyT3d3jmXDMunoSE/Q6I9EZI2O/wkyK9\n5YOojn9xHDglvyjSbMCLBB1fgjr4CWRE+qgRo97SXUT6bX5QpPd0F5F+G0SSjIlIPwsiScZE\npFU2H07LRCVl4xfVQFduYg3OwdEiXf1dp+w0hVJFvbzj3dsxltUx9LBxXFfb6O5in1ZiI9IO\ntkTKjDCpXxw8Str2kdjSw297Cg4WqfZz62TH5mzRpCZxd4xldQw937je3eU+rcVGpG02RLqr\npG7rRN29oqXSxUKVXbFURaTB0RwrUpcJd1cnOyZbpcqX9szdHWNZHUNHNq52d7lPq7ERaZsN\nkUpVda83dfGKhibJp/31n7DB4Rwq0lVlXm6d7CTKzo8X9rx5335GsjqFjmxc6+5Kn9Zj84Xs\nJl2CSpWYTFZdCrPKlLprYqLvNN3FUc/Zan31dIqt3aSHQz8L0bO8sMHhHCpSd6f2Bt8sOzpl\nMR7+aI9kdQr9ZMo3+/REbETyUSrXl5urvhapvnSxpbL1bjhOUVOb7V1bO7W7zBocz6Ei1UEm\nwuyUOtMxMvVwd4xktQ5r96Z8q0/PxD7NWT4J3U2o6RxK9b2l1nfyVNfdzASjXRPJ3pA6//ST\nc3JtEWm+5/KgvdnrVIRLl/wNkWa1T6R8S6TdsU9zlk+CMk+aVpQq2NKuiFTrB2aNvX1dWkSa\n77k8aK95En+WNFOpJZGmenmRno19mrN8Emye7FqByut+SfZRXbJ1kcpeu6u+sjaFnqcgUrjn\n6tW/iM7t0qRBpE9kEqm9JP4XQroumbKZeIkdvupIzQyvsXNDt8EJOEKkMXNtkIlZdprYakNh\nrk/ujvGs+mfhVZH+Evs0Z/kkOCK1bVWm2ohCpdfKPvHapZvHtGr3sKs442KOc93yGpyBU4k0\nz05siI6/GVHL+zn7Pp/yjT7tj41IPp5IrTOJsyJdzBWy0vM3p6hndP28xF63zNXVa3AGTjW1\nc7Jjv0d6xH4DMhcpntV++/MpX+rT07ERyWcSKbVrdaldgKjtM9LSLxvy4QcupdK/xip1uk/3\ny4aj8QZt+MuGJl9a//Z3jGe1b/F8ypf69HRsRPKZRLrZK6E50WOx00uTtX6xfzTSZPEGMA1a\n+9fJTrKRKO85P5rVIfTTKV/s07OxEcnHmdqZXzaY60+hC5WZHdsfd5umTtG9sMUbQDho3ex0\nxXTxfhSIFM3qEPrplC/36cnYiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIB\nCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAA\niAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiPRJqJYzdlI4\nLZ/Egkj4dTzk/5PYKZLivP5zSPgn8YRIyPRvIdmfBCKdFpL9SQiIxAl/D+T1k0Ck00JePwlE\nOi3k9ZN4m0gMg79CBj+Jd4vEaHgZUvdJINJpIXWfxL8TidXzJyFbnwQinRay9UkcI9JHj5G3\ndF4pFcb96CT9HCcQ6cMGzGzAiwQdX4I6+BDOIdIHjRn1ls4i0qdzIpE+ZeAgEsxBpKdBJJhz\nVpHe8iQiAyIdw+aQKBOVlE1QVJa+yXUYll6tSO/aM4u09UmvfgMnk3WhVPGI7tToTbUXJh33\nm4cego6J3+jUcp+WYyPSDrbGfWbcSP1i7SlT94UakTxqPxVOJitTTJrYXonZ5phURhqPoZ2g\nsc4/0ae12Ii0zca4v6ukbutE3b1irfKpSVc1iJQvRHm9d+3nijTmxeKkr026YpOrMrJXqQr9\nMmWyVkWj7xJFLLQbNNb5J/q0FhuRttkQqVRV93pTF6941S89V5X1Mdxaqd61HyvSlBeLk76b\nUahRSWS3RDWtd1Zy5RwxDO0EjXZ+f5/WY/OF7CZdgkqVmExWXQqzypTybi5hrpe50jN5c69x\nild1nQKU7SjSNYz+5961HyvSlBeLk75C1Ut7DTvPHHNjTaGdoHtY6dPTsRHJR6lcX26u2gPV\nly62VLbj+VPKK+aqKnrT2rqdMj/VSvWu/ViRan/wu+lLVXtJzIxtgXJ2SWpUFgvt/9lkpU9P\nx0Ykn+4m1HQOpXpSUes7earrbrpkhuuCSIZsjGH+BLUivWs/VqR2edD2F6/YzE5zU/Onp6uZ\nZ0WjhQd6rU9Px0YkH2WeNO3pjZyruEjGtGa8bo5bvFqR3rUfIlL80Isi6cWGYumR8pon4aZH\nkvsB5UV6NjYi+fRDQpll1rzuJ++P6pKtiGRphnVTL9dNsJr6t9613ymSTvNjJVGFfzVqkiwI\niEhnYxJJz9u7a6V+wszGL4SSKZvJLLFDUUUGmVDv2k8Tyf0qzcuEk77t4R+s6GUz5fp95+dk\ni6U+PR0bkXwckdq2KlN9mSxUeq0eps4u3TymVbuHs4qDSPOjLIrkZXK+Oeyfs+2RZrMfQXgr\na48nvsBb6tPTsRHJxxOpda6YVqSLeXCq9NOvU7Rfdow5Hi9hXq1I79pPE2neqmeWyUd0WWZI\n4nQPqmLt+tBO0J0s9enp2IjkM4mU2rU6s2p3b2v7jBT/ZUOps9uUw+pEn/mgFrxB66Sv06TR\niw23yD7mlw1NPj0jxX3rQ8d+mvBan56OjUg+k0g3Oyu5D7/usut56bSkPRWbZPyiaYoR1oI/\nxXMyeVn5niCZtpn9ChX7CePwNl2J9Fyfno2NSD7O1M78ssFcfwpdqMwcrTE/ATZNg2J69WOE\nteAPWid9OtWL31xPSexX09ZEcoP+sU9PxkYkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEA\nBEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAA\nRAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAk\nAAEQCUAARAIQAJEABEAkAAEQ6afgdL8LMvtTKOcVJCGnPwUivQty+lPERFKMAgFI4U+BSO+C\nFP4UmyIxHl6ExP0U+0ViYDwH+fopEOldkK+f4tdEUh3/Jux35At28mMiqfHl7WG/Il+wF0R6\nV9ivyBfs5cdEMiASiINIYhER6Zf5QZHe8EEQaZvNRZ4yUUnZDO/uY/PrULqmYwOneBIOFenq\nxVQWU64LpYpHdKdGb6rXKrzQw+kZQ298Dr9PwdldiI1IO9gSKTNnP+3fNcnQvB52LE2DpPGL\nZ+H9Ii3vV3u5rR2RqpVEJWZbvVLhhvZPz2p3In2a7x6PjUjbbIh0V0nd1om627f50Lyr6q+t\nqmj0RazwiqfhQJHGFPVvVT6Wky6pTa7KyF6lTl/ptJ1VuKGD07Ptkd+n2e4LsRFpmw2RSlV1\nrzd1Me9uw/zhqrK+lKsxjFM8DceJNKVoeH8ZijejUKOSyG6JMnMptVzhhvZPzzB9XOxr2Kdw\n98XYfCG7SZegUiUmk1WXwqwype7Ok5jrZa70TL6/mj7GJHcjwc+s8w6RbH0Zjv7rUCxU+Mgz\n2zl0zKuYQnunZ0dfgz5Fdt8d+0xn+QwolevLzVWfadWXLrZUtqMV9k+mHsOELvClUVmkeAKO\nEylMUa6qor86paq9JGYevEA5SRerqIPzsvvaFfYpsvvu2Ijk092Ems6hVM8han0nT3XdzUzj\nWj+bl656Smtwua0ixRNw5GJDKJIha8eLV2xmp7mp4OlpVvGqSGHT+O6I9ArKPGnqbKnAgFAk\nc5OPi/RI8kjxDJxHJHN1asytRenH+KZwH01crnnibxorZtkXFOnZ2IjkY/OkX8vuQln3k/dH\ndclCkdKkWRCpSbJI8RScRyRLY+/4Os2PYNnZpQjndrYCkc7LJJKet3fXSv2EmY2rP8mYzcLc\nsKIiZWmseAr+uUjuullsDNp7/+LmntmKXljR75v8TaT47vtiI5KPI1LbVmWqL5OFSq+VXVew\nSzePblanRtwd9cY0e8yLJ+GUIu34nmC+2hwd7NPp2d1d/7kttvu+2Ijk44nUOlfMR7/AoO9D\nlVkUjYtUTat01akW7AznmdrZ74PMuLRJfUSzNTRLFyvc0NPp2d1dt0/x3ffFRiSfSaTUrtWl\ndgGits9I4Xffs6mdMxziI+OHCb78LM1iQ2WsaPRiwy2yj/khQ5NPz0izCjd07KcJT/Qpvvu+\n2IjkM4l0s/eb+/CbObuel46Ltk5zp1RM96nCv2WBd/NukvHLuf6LuvhlJ5m22f2TWOMhdLoS\nabtP8d33xeYs+zhTO/PLBnP9KXShMrPjxvwEOGjulJwJXzD3A3/Q6kym/X2lS3WyNB2bmvX7\nO/vNQgen55U+zXbfF5uzDCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgA\nAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAA\nIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiAS\ngACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEjH8d/79/gHh3hhj28EkY4Dkb4IRDoORPoiEOk4\nEOmLQKTjQKQvApEABEAkAAEQCUAARAIQAJEABECko0g69rVKnNbh37B5G2u0ttO0x77DbMXe\n97G+D0Q6iGR82W42tQ7/hq2TtcaxnfpRv/swW7F3fqzvA5EO4h0iJU79vkGeDHekvYdBpAUQ\n6SD2jbjE/btjvD59R/Ir9974EGkGIh3ETpGGZ5e+9btEeu4wiDQHkQ7iiTvSE+P1L3ek3YdZ\nUw6R4N/yxIh7v0jPHQaRIiDSQXyuSC/cw34ARDqIj53aJS917OtBpIPYL9LaTGo56tMi7TxM\n4u+GSD2IdBT7f9mw9jds/vROT+6RbP0Ggl82AMDLIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKA\nAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiA\nSAACIBKAAP8DcACMYP8DX+YAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select the top five up-regulated pathways \n",
    "topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=5), pathway]\n",
    "# select the top five down-regulated pathways \n",
    "topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=5), pathway]\n",
    "# concatenate them for visualization\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "#Viewing the 5 most significantly up-regulated and down-regulated pathways each with the FGSEA internal plot function\n",
    "plotGseaTable(pathways = KEGG_hallmark[topPathways], stats = stats, fgseaRes = fgseaRes,\n",
    "              gseaParam = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "181989b1",
   "metadata": {},
   "source": [
    "## Gene Set Enrichment Analysis using GSA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f471b1",
   "metadata": {},
   "source": [
    "Gene Set Analysis (GSA), an Enrichment Analysis, is a method that is commonly used to summarize high-dimensional gene expression data sets into sets according to their biological relevance. GSA takes the ranked gene lists from the initial stage of a gene expression analysis and aggregates the genes into sets based on shared biological or functional properties as specified by a reference knowledge base. Such databases often contain phenotype associations, molecular interactions and regulation and are referenced in the analysis of the resultant gene sets to find the relevance of the gene properties to the phenotype of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "443e636c-265e-4829-85f3-1999eeff4385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-4.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Run the following command to take the quiz\n",
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-4.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f0c9233",
   "metadata": {},
   "source": [
    "### Data preparation\n",
    "The GSA method is freely available as standalone package in CRAN repository. We can use the following code to install the package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c3efb930",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install GSA from CRAN\n",
    "suppressMessages({if (!require(\"GSA\"))\n",
    "        suppressWarnings(install.packages(\"GSA\"))\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ee423bc8-1c31-46b5-8955-5d89f6f1a745",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "  library(GSA)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a2ed98",
   "metadata": {},
   "source": [
    "The GSA method requires an expression matrix, a numeric vector containing the class of each sample and a vector of genes the inputs. We can easily get those inputs by loading the data that we processed in the [**Module 01** ](./Module01-GEO_Data_Processing.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3982c128",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading expression data with groups\n",
    "data <- readRDS(\"./data/GSE48350.rds\")\n",
    "expression_data <- data$expression_data\n",
    "norm_expression_data <- data$norm_expression_data\n",
    "groups <- data$groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad173974",
   "metadata": {},
   "source": [
    "We can also use the sample approach available in the [**Module 01** ](./Module01-GEO_Data_Processing.ipynb) to map the probe IDs to gene symbols. The step-by-step coding instruction is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e0d1acf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the probe IDs\n",
    "expression_data$PROBEID <- rownames(expression_data)\n",
    "probeIDs <- rownames(expression_data)\n",
    "# Perform gene mapping\n",
    "suppressMessages({\n",
    "  annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'SYMBOL'))\n",
    "})\n",
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, expression_data, by=\"PROBEID\")\n",
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$SYMBOL),]\n",
    "# Remove duplicated genes symbol\n",
    "new_expression_data <-  new_expression_data[!duplicated(new_expression_data$SYMBOL,fromLast=FALSE),]\n",
    "rownames(new_expression_data) <- new_expression_data$SYMBOL\n",
    "# Drop PROBEID, GENENAME, and SYMBOL columns\n",
    "new_expression_data <- new_expression_data[,-c(1:3)]\n",
    "genenames= rownames(new_expression_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8acf4f4",
   "metadata": {},
   "source": [
    "### GSA Enrichment Analysis using GO Terms\n",
    "Using data obtained from the previous step, we can run the GSA method by calling the function `GSA`. We can reuse `GO_term_hallmark` and `KEGG_hallmark` loaded in FGSEA to perform analysis. The code details are shown below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d60c66b0-038f-4645-8076-10fb60436a69",
   "metadata": {},
   "source": [
    "Used **arguments** of the `GSA` function:\n",
    "- `x` (*matrix*): Data x: p by n matrix of features (expression values), one observation per column (missing values allowed); y: n-vector of outcome measurements.\n",
    "- `y` (*numeric vector*): Vector of response values: 1,2 for two class problem, or 1,2,3 ... for multiclass problem, or real numbers for quantitative or survival problems\n",
    "- `genesets` (*list*): Gene set collection.\n",
    "- `genenames` (*vector*): Vector of genenames in expression dataset.\n",
    "- `resp.type` (*character*): Problem type: \"quantitative\" for a continuous parameter; \"Two class unpaired\" ; \"Survival\" for censored survival outcome; \"Multiclass\" : more than 2 groups, coded 1,2,3...; \"Two class paired\" for paired outcomes, coded -1,1 (first pair), -2,2 (second pair), etc.\n",
    "- `nperms` (*numeric*): Number of permutations used to estimate false discovery rates.\n",
    "- `random.seed` (*integer*): Optional initial seed for random number generator."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0648ecb6-f0ec-41aa-b3e6-e812ba18d85c",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <i class=\"fa fa-exclamation-circle\" aria-hidden=\"true\"></i>\n",
    "    <b>Alert: </b>  Please note that the outputs generated from GSA can be varied depending on R environments and software versions. Especially, when the users run the scripts using User-managed Notebooks instances which have a preinstalled suite of packages.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e026aa03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "# Getting \n",
    "genesets = GO_term_hallmark\n",
    "GSA.obj<-GSA(x = as.matrix(new_expression_data), y = as.numeric(groups$groups), genenames = genenames, genesets = genesets, resp.type = \"Two class unpaired\", nperms = 100, random.seed = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd369a1f-3475-43c9-8583-ba577770bd45",
   "metadata": {},
   "source": [
    "Used **arguments** of the `GSA.listsets` method:\n",
    "\n",
    "- `GSA.obj` (*GSA object*): Object returned by GSA function.\n",
    "- `geneset.names` (*vector*): Optional vector of names for the gene sets.\n",
    "- `FDRcut` (*numeric*): False discovery rate cutpoint for listed sets. A value of 1 will cause all sets to be listed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ebba9b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj = GSA.obj, geneset.names = names(genesets), FDRcut = .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df55e53c",
   "metadata": {},
   "source": [
    "A table of the negative gene sets. “Negative” means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. E.g., for two classes coded 1,2, lower expression correlates with class 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ae97649e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>57 </td><td>GO:0000422</td><td>-0.4838</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>641</td><td>GO:0006123</td><td>-1.3695</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>893</td><td>GO:0006851</td><td>-0.802 </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>925</td><td>GO:0006906</td><td>-0.4592</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>976</td><td>GO:0007007</td><td>-1.2636</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>987</td><td>GO:0007029</td><td>-0.5442</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 57  & GO:0000422 & -0.4838 & 0 & 0\\\\\n",
       "\t 641 & GO:0006123 & -1.3695 & 0 & 0\\\\\n",
       "\t 893 & GO:0006851 & -0.802  & 0 & 0\\\\\n",
       "\t 925 & GO:0006906 & -0.4592 & 0 & 0\\\\\n",
       "\t 976 & GO:0007007 & -1.2636 & 0 & 0\\\\\n",
       "\t 987 & GO:0007029 & -0.5442 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 57  | GO:0000422 | -0.4838 | 0 | 0 |\n",
       "| 641 | GO:0006123 | -1.3695 | 0 | 0 |\n",
       "| 893 | GO:0006851 | -0.802  | 0 | 0 |\n",
       "| 925 | GO:0006906 | -0.4592 | 0 | 0 |\n",
       "| 976 | GO:0007007 | -1.2636 | 0 | 0 |\n",
       "| 987 | GO:0007029 | -0.5442 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score   p-value FDR\n",
       "[1,] 57       GO:0000422    -0.4838 0       0  \n",
       "[2,] 641      GO:0006123    -1.3695 0       0  \n",
       "[3,] 893      GO:0006851    -0.802  0       0  \n",
       "[4,] 925      GO:0006906    -0.4592 0       0  \n",
       "[5,] 976      GO:0007007    -1.2636 0       0  \n",
       "[6,] 987      GO:0007029    -0.5442 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <-res$negative\n",
    "head(neg.table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c237dea8",
   "metadata": {},
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e640ff47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1681</td><td>GO:0014037</td><td>0.7351</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1682</td><td>GO:0014044</td><td>0.9529</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1713</td><td>GO:0014855</td><td>0.3543</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1729</td><td>GO:0014912</td><td>0.6334</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1877</td><td>GO:0017145</td><td>0.5308</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>2009</td><td>GO:0021533</td><td>0.596 </td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 1681 & GO:0014037 & 0.7351 & 0 & 0\\\\\n",
       "\t 1682 & GO:0014044 & 0.9529 & 0 & 0\\\\\n",
       "\t 1713 & GO:0014855 & 0.3543 & 0 & 0\\\\\n",
       "\t 1729 & GO:0014912 & 0.6334 & 0 & 0\\\\\n",
       "\t 1877 & GO:0017145 & 0.5308 & 0 & 0\\\\\n",
       "\t 2009 & GO:0021533 & 0.596  & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 1681 | GO:0014037 | 0.7351 | 0 | 0 |\n",
       "| 1682 | GO:0014044 | 0.9529 | 0 | 0 |\n",
       "| 1713 | GO:0014855 | 0.3543 | 0 | 0 |\n",
       "| 1729 | GO:0014912 | 0.6334 | 0 | 0 |\n",
       "| 1877 | GO:0017145 | 0.5308 | 0 | 0 |\n",
       "| 2009 | GO:0021533 | 0.596  | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score  p-value FDR\n",
       "[1,] 1681     GO:0014037    0.7351 0       0  \n",
       "[2,] 1682     GO:0014044    0.9529 0       0  \n",
       "[3,] 1713     GO:0014855    0.3543 0       0  \n",
       "[4,] 1729     GO:0014912    0.6334 0       0  \n",
       "[5,] 1877     GO:0017145    0.5308 0       0  \n",
       "[6,] 2009     GO:0021533    0.596  0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <-res$positive\n",
    "head(pos.table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bffb3bdc-7ec1-4d1e-9012-fe6cbe70b9e8",
   "metadata": {},
   "source": [
    "Used **arguments** of the `GSA.genescores` method:\n",
    "- `geneset.number` (*numeric*): Number indicating which gene set is to examined.\n",
    "- `genesets` (*list*): The gene set collection.\n",
    "- `GSA.obj` (*GSA obj*ect): Object returned by function GSA.\n",
    "- `genenames` (*vector*): Vector of gene names for gene in expression dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "79687f5d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 114 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SH3GLB1  </td><td>2.729</td></tr>\n",
       "\t<tr><td>IFT88    </td><td>2.491</td></tr>\n",
       "\t<tr><td>MTMR3    </td><td>2.271</td></tr>\n",
       "\t<tr><td>TP53INP1 </td><td>2.134</td></tr>\n",
       "\t<tr><td>ATG16L2  </td><td>1.973</td></tr>\n",
       "\t<tr><td>ATG2B    </td><td>1.944</td></tr>\n",
       "\t<tr><td>ATG14    </td><td>1.935</td></tr>\n",
       "\t<tr><td>ATG7     </td><td>1.597</td></tr>\n",
       "\t<tr><td>STING1   </td><td>1.457</td></tr>\n",
       "\t<tr><td>MAP1LC3B2</td><td>1.385</td></tr>\n",
       "\t<tr><td>ATG9B    </td><td>1.105</td></tr>\n",
       "\t<tr><td>TBC1D12  </td><td>1.061</td></tr>\n",
       "\t<tr><td>FEZ1     </td><td>1    </td></tr>\n",
       "\t<tr><td>EPHB2    </td><td>0.888</td></tr>\n",
       "\t<tr><td>NUPR1    </td><td>0.822</td></tr>\n",
       "\t<tr><td>ATG4C    </td><td>0.784</td></tr>\n",
       "\t<tr><td>WDR45B   </td><td>0.775</td></tr>\n",
       "\t<tr><td>TMEM39A  </td><td>0.751</td></tr>\n",
       "\t<tr><td>PIP4K2A  </td><td>0.659</td></tr>\n",
       "\t<tr><td>ATG2A    </td><td>0.622</td></tr>\n",
       "\t<tr><td>RAB7A    </td><td>0.588</td></tr>\n",
       "\t<tr><td>EFNB1    </td><td>0.571</td></tr>\n",
       "\t<tr><td>RNF186   </td><td>0.551</td></tr>\n",
       "\t<tr><td>UBXN2A   </td><td>0.516</td></tr>\n",
       "\t<tr><td>WIPI2    </td><td>0.464</td></tr>\n",
       "\t<tr><td>BECN1    </td><td>0.392</td></tr>\n",
       "\t<tr><td>IFT20    </td><td>0.389</td></tr>\n",
       "\t<tr><td>TBC1D14  </td><td>0.381</td></tr>\n",
       "\t<tr><td>WIPI1    </td><td>0.352</td></tr>\n",
       "\t<tr><td>SNX18    </td><td>0.35 </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SCFD1    </td><td>-1.858</td></tr>\n",
       "\t<tr><td>LRSAM1   </td><td>-1.899</td></tr>\n",
       "\t<tr><td>UBXN2B   </td><td>-1.922</td></tr>\n",
       "\t<tr><td>RAB3GAP1 </td><td>-1.924</td></tr>\n",
       "\t<tr><td>GABARAPL3</td><td>-1.988</td></tr>\n",
       "\t<tr><td>MAP1LC3B </td><td>-2.017</td></tr>\n",
       "\t<tr><td>MTOR     </td><td>-2.019</td></tr>\n",
       "\t<tr><td>RNF5     </td><td>-2.026</td></tr>\n",
       "\t<tr><td>VMP1     </td><td>-2.122</td></tr>\n",
       "\t<tr><td>SMURF1   </td><td>-2.124</td></tr>\n",
       "\t<tr><td>RAB1A    </td><td>-2.16 </td></tr>\n",
       "\t<tr><td>TRAF6    </td><td>-2.232</td></tr>\n",
       "\t<tr><td>RAB3GAP2 </td><td>-2.265</td></tr>\n",
       "\t<tr><td>PHF23    </td><td>-2.39 </td></tr>\n",
       "\t<tr><td>ATP2A2   </td><td>-2.517</td></tr>\n",
       "\t<tr><td>ATG5     </td><td>-2.518</td></tr>\n",
       "\t<tr><td>RALB     </td><td>-2.528</td></tr>\n",
       "\t<tr><td>STX12    </td><td>-2.608</td></tr>\n",
       "\t<tr><td>MOAP1    </td><td>-2.609</td></tr>\n",
       "\t<tr><td>ATG4D    </td><td>-2.669</td></tr>\n",
       "\t<tr><td>PIP4K2C  </td><td>-2.684</td></tr>\n",
       "\t<tr><td>MAP1LC3A </td><td>-2.716</td></tr>\n",
       "\t<tr><td>PINK1    </td><td>-2.797</td></tr>\n",
       "\t<tr><td>UBQLN1   </td><td>-2.802</td></tr>\n",
       "\t<tr><td>EMC6     </td><td>-2.957</td></tr>\n",
       "\t<tr><td>ATG9A    </td><td>-2.995</td></tr>\n",
       "\t<tr><td>ARFIP2   </td><td>-3.055</td></tr>\n",
       "\t<tr><td>SMC2     </td><td>-3.079</td></tr>\n",
       "\t<tr><td>SEC22B   </td><td>-3.088</td></tr>\n",
       "\t<tr><td>RB1CC1   </td><td>-4.001</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 114 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " Gene & Score\\\\\n",
       "\\hline\n",
       "\t SH3GLB1   & 2.729\\\\\n",
       "\t IFT88     & 2.491\\\\\n",
       "\t MTMR3     & 2.271\\\\\n",
       "\t TP53INP1  & 2.134\\\\\n",
       "\t ATG16L2   & 1.973\\\\\n",
       "\t ATG2B     & 1.944\\\\\n",
       "\t ATG14     & 1.935\\\\\n",
       "\t ATG7      & 1.597\\\\\n",
       "\t STING1    & 1.457\\\\\n",
       "\t MAP1LC3B2 & 1.385\\\\\n",
       "\t ATG9B     & 1.105\\\\\n",
       "\t TBC1D12   & 1.061\\\\\n",
       "\t FEZ1      & 1    \\\\\n",
       "\t EPHB2     & 0.888\\\\\n",
       "\t NUPR1     & 0.822\\\\\n",
       "\t ATG4C     & 0.784\\\\\n",
       "\t WDR45B    & 0.775\\\\\n",
       "\t TMEM39A   & 0.751\\\\\n",
       "\t PIP4K2A   & 0.659\\\\\n",
       "\t ATG2A     & 0.622\\\\\n",
       "\t RAB7A     & 0.588\\\\\n",
       "\t EFNB1     & 0.571\\\\\n",
       "\t RNF186    & 0.551\\\\\n",
       "\t UBXN2A    & 0.516\\\\\n",
       "\t WIPI2     & 0.464\\\\\n",
       "\t BECN1     & 0.392\\\\\n",
       "\t IFT20     & 0.389\\\\\n",
       "\t TBC1D14   & 0.381\\\\\n",
       "\t WIPI1     & 0.352\\\\\n",
       "\t SNX18     & 0.35 \\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t SCFD1     & -1.858\\\\\n",
       "\t LRSAM1    & -1.899\\\\\n",
       "\t UBXN2B    & -1.922\\\\\n",
       "\t RAB3GAP1  & -1.924\\\\\n",
       "\t GABARAPL3 & -1.988\\\\\n",
       "\t MAP1LC3B  & -2.017\\\\\n",
       "\t MTOR      & -2.019\\\\\n",
       "\t RNF5      & -2.026\\\\\n",
       "\t VMP1      & -2.122\\\\\n",
       "\t SMURF1    & -2.124\\\\\n",
       "\t RAB1A     & -2.16 \\\\\n",
       "\t TRAF6     & -2.232\\\\\n",
       "\t RAB3GAP2  & -2.265\\\\\n",
       "\t PHF23     & -2.39 \\\\\n",
       "\t ATP2A2    & -2.517\\\\\n",
       "\t ATG5      & -2.518\\\\\n",
       "\t RALB      & -2.528\\\\\n",
       "\t STX12     & -2.608\\\\\n",
       "\t MOAP1     & -2.609\\\\\n",
       "\t ATG4D     & -2.669\\\\\n",
       "\t PIP4K2C   & -2.684\\\\\n",
       "\t MAP1LC3A  & -2.716\\\\\n",
       "\t PINK1     & -2.797\\\\\n",
       "\t UBQLN1    & -2.802\\\\\n",
       "\t EMC6      & -2.957\\\\\n",
       "\t ATG9A     & -2.995\\\\\n",
       "\t ARFIP2    & -3.055\\\\\n",
       "\t SMC2      & -3.079\\\\\n",
       "\t SEC22B    & -3.088\\\\\n",
       "\t RB1CC1    & -4.001\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 114 × 2 of type chr\n",
       "\n",
       "| Gene | Score |\n",
       "|---|---|\n",
       "| SH3GLB1   | 2.729 |\n",
       "| IFT88     | 2.491 |\n",
       "| MTMR3     | 2.271 |\n",
       "| TP53INP1  | 2.134 |\n",
       "| ATG16L2   | 1.973 |\n",
       "| ATG2B     | 1.944 |\n",
       "| ATG14     | 1.935 |\n",
       "| ATG7      | 1.597 |\n",
       "| STING1    | 1.457 |\n",
       "| MAP1LC3B2 | 1.385 |\n",
       "| ATG9B     | 1.105 |\n",
       "| TBC1D12   | 1.061 |\n",
       "| FEZ1      | 1     |\n",
       "| EPHB2     | 0.888 |\n",
       "| NUPR1     | 0.822 |\n",
       "| ATG4C     | 0.784 |\n",
       "| WDR45B    | 0.775 |\n",
       "| TMEM39A   | 0.751 |\n",
       "| PIP4K2A   | 0.659 |\n",
       "| ATG2A     | 0.622 |\n",
       "| RAB7A     | 0.588 |\n",
       "| EFNB1     | 0.571 |\n",
       "| RNF186    | 0.551 |\n",
       "| UBXN2A    | 0.516 |\n",
       "| WIPI2     | 0.464 |\n",
       "| BECN1     | 0.392 |\n",
       "| IFT20     | 0.389 |\n",
       "| TBC1D14   | 0.381 |\n",
       "| WIPI1     | 0.352 |\n",
       "| SNX18     | 0.35  |\n",
       "| ⋮ | ⋮ |\n",
       "| SCFD1     | -1.858 |\n",
       "| LRSAM1    | -1.899 |\n",
       "| UBXN2B    | -1.922 |\n",
       "| RAB3GAP1  | -1.924 |\n",
       "| GABARAPL3 | -1.988 |\n",
       "| MAP1LC3B  | -2.017 |\n",
       "| MTOR      | -2.019 |\n",
       "| RNF5      | -2.026 |\n",
       "| VMP1      | -2.122 |\n",
       "| SMURF1    | -2.124 |\n",
       "| RAB1A     | -2.16  |\n",
       "| TRAF6     | -2.232 |\n",
       "| RAB3GAP2  | -2.265 |\n",
       "| PHF23     | -2.39  |\n",
       "| ATP2A2    | -2.517 |\n",
       "| ATG5      | -2.518 |\n",
       "| RALB      | -2.528 |\n",
       "| STX12     | -2.608 |\n",
       "| MOAP1     | -2.609 |\n",
       "| ATG4D     | -2.669 |\n",
       "| PIP4K2C   | -2.684 |\n",
       "| MAP1LC3A  | -2.716 |\n",
       "| PINK1     | -2.797 |\n",
       "| UBQLN1    | -2.802 |\n",
       "| EMC6      | -2.957 |\n",
       "| ATG9A     | -2.995 |\n",
       "| ARFIP2    | -3.055 |\n",
       "| SMC2      | -3.079 |\n",
       "| SEC22B    | -3.088 |\n",
       "| RB1CC1    | -4.001 |\n",
       "\n"
      ],
      "text/plain": [
       "      Gene      Score \n",
       " [1,] SH3GLB1   2.729 \n",
       " [2,] IFT88     2.491 \n",
       " [3,] MTMR3     2.271 \n",
       " [4,] TP53INP1  2.134 \n",
       " [5,] ATG16L2   1.973 \n",
       " [6,] ATG2B     1.944 \n",
       " [7,] ATG14     1.935 \n",
       " [8,] ATG7      1.597 \n",
       " [9,] STING1    1.457 \n",
       "[10,] MAP1LC3B2 1.385 \n",
       "[11,] ATG9B     1.105 \n",
       "[12,] TBC1D12   1.061 \n",
       "[13,] FEZ1      1     \n",
       "[14,] EPHB2     0.888 \n",
       "[15,] NUPR1     0.822 \n",
       "[16,] ATG4C     0.784 \n",
       "[17,] WDR45B    0.775 \n",
       "[18,] TMEM39A   0.751 \n",
       "[19,] PIP4K2A   0.659 \n",
       "[20,] ATG2A     0.622 \n",
       "[21,] RAB7A     0.588 \n",
       "[22,] EFNB1     0.571 \n",
       "[23,] RNF186    0.551 \n",
       "[24,] UBXN2A    0.516 \n",
       "[25,] WIPI2     0.464 \n",
       "[26,] BECN1     0.392 \n",
       "[27,] IFT20     0.389 \n",
       "[28,] TBC1D14   0.381 \n",
       "[29,] WIPI1     0.352 \n",
       "[30,] SNX18     0.35  \n",
       "[31,] ⋮         ⋮     \n",
       "[32,] SCFD1     -1.858\n",
       "[33,] LRSAM1    -1.899\n",
       "[34,] UBXN2B    -1.922\n",
       "[35,] RAB3GAP1  -1.924\n",
       "[36,] GABARAPL3 -1.988\n",
       "[37,] MAP1LC3B  -2.017\n",
       "[38,] MTOR      -2.019\n",
       "[39,] RNF5      -2.026\n",
       "[40,] VMP1      -2.122\n",
       "[41,] SMURF1    -2.124\n",
       "[42,] RAB1A     -2.16 \n",
       "[43,] TRAF6     -2.232\n",
       "[44,] RAB3GAP2  -2.265\n",
       "[45,] PHF23     -2.39 \n",
       "[46,] ATP2A2    -2.517\n",
       "[47,] ATG5      -2.518\n",
       "[48,] RALB      -2.528\n",
       "[49,] STX12     -2.608\n",
       "[50,] MOAP1     -2.609\n",
       "[51,] ATG4D     -2.669\n",
       "[52,] PIP4K2C   -2.684\n",
       "[53,] MAP1LC3A  -2.716\n",
       "[54,] PINK1     -2.797\n",
       "[55,] UBQLN1    -2.802\n",
       "[56,] EMC6      -2.957\n",
       "[57,] ATG9A     -2.995\n",
       "[58,] ARFIP2    -3.055\n",
       "[59,] SMC2      -3.079\n",
       "[60,] SEC22B    -3.088\n",
       "[61,] RB1CC1    -4.001"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(geneset.number = 10, genesets = genesets, GSA.obj = GSA.obj, genenames = genenames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d76b9c0-d891-4622-8b3e-1311ab6b85b7",
   "metadata": {},
   "source": [
    "Used **argument** of `GSA.plot` method:\n",
    "- `GSA.obj` (*GSA object*): Object returned by GSA function.\n",
    "- `fac` (*numeric*): value for jittering points in plot (\"factor\" in called to jitter()).\n",
    "- `FDRcut` (*numeric*): False discovery rate cutpoint for sets to be plotted. A value of 1 (the default) will cause all sets to be plotted.plotted\r\n",
    "\r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1cfde1f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAil+ZNTU1h0E9o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD///9W+WhdAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAd3UlEQVR4nO3di5aiOgJG4TQihYqC7/+yLeAFFSGan0DC/tbMOXaV\nVXFa94AhoDkDcGbmfgBADAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUIC\nBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQI\nCRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQ\nICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAk\nQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECA\nkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAA\ngd9DOu4yU8vyo/DxAEH6NaQqNQ/b4SGAwPgLKTfJ4dTcKovE5FMMAczEY0iJOd1vn0wyxRDA\nTDyG9LT1G94UEhICwxYJEPD7Hqkom1u8R0JsPIZ03nbmONJqkiGAefgM6XzMm+NISbYbOY5E\nSAiM15CWNASgREiAgN9dO9slQoSEwHgM6ZslQj8OAcyEJUKAAAdkAQGWCAECbJEAAZYIAQIs\nEQIEWCIECLCyARizuTu3/32/CyEBIzbv3u7jP6R9akxWTDoEINTT0XtJ3o8jXWccBiftCAnT\nqSN4SaFvX637XYuSfIeUm7w6n8vc7N+/7XRtI8BKfxZ9e2uf77qAkBLTzHtXJp1iCGDEWBp9\nrSxx1+62sWGJEGYxuJXp3S71/tDbXXyH9HcLiSVCmMPXEVnyGlK22xfmcLlZ5SwRwiwm6shv\nSPeJBGMSlghhFtN05PU40um032dZM+WQD3ZESJhO20ynnw9rFb7Dygasj2IT9IKQsCr6hFqE\nhJWYYDPUMVdIHEeCP5Mm1CIkxGzazVAHu3aIlK+EWoSEMNwWbVscUPW2GeogJARhM7pK7rHq\ndI7Hx7W/EQKbjuYp6IprfyMENhmtJSSu/Y3fEdIdV1qFg2V3xLW/EYrXaq7/bQuauyO2SAjX\n3PF0ce1vBGn2TdALrv2N4CwtohrX/kZQlhhRjZUNCMXsM3NDCAkhWHJDDULC0i0+ohohYcmC\niKhGSFioRb8lekNIWKCgGmoQEhYmvIhqhITlCGtv7gkhYRnCbahBSJhf4BHVCAnziiCiGiFh\nNgG/JXpDSJhFRA01CAnexRZRjZBg6+XyCE+nebdfuJ/+/flTV2OMqEZIsDNy3ZHhbz/am/t/\nxlQICVa+yOjTfef+nzApQoKN8Q2O1SYpXoQEG3Q0gpBg5S2Ltxu9311LR4QES59j6Vyq8fUj\nwx/3mPvhT42QMIH4w3lFSIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBA\nSIAAIQECfj+Medd8GLPJcj6MGXHxGFKVmoftJEMAM/EYUm6Sw6m5VRaJyacYApiJx5ASc7rf\nPplkiiGAmXgMyZhPf5ANAcyELRIg4Pc9UlE2t3iPhNj4nP7edmbt0mqSIYB5+D2OlDfHkZJs\nx3EkxIWVDYAAIQECLBECBFgiBAiwRAgQ4IAsIMASIUCALRIgwBIhQIAlQoAAS4QAAVY2AAKE\nBAj4DKn6M2ZbXH8J09+Iic8lQkm70K79JYSEmHid/t5fatonzTK795BM149DADPxekC2+VeZ\npCVbJERmhiVC1XZLSIiMx5BSczsIm24JCXHxGNLe/F1vlWZLSIiKz+nv/F5PMTKfQEgIjNcD\nsqfsdqv8IyTEhJUNgAAhAQKEBAjMFRKTDYgKIQEC7NoBAoQECBASIMC1vwEBrv0NCHDtb0CA\nK60CAlz7GxBgiwQIcO1vQIBrfwMCXPsbEGBlAyBASIAAIQEChAQIEBIgQEiAACEBAoQECBAS\nIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBA\nSIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQILCck0zXN\nEMBUlhOS5yEAJUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJ\nECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAg\nJECAkAABQgIECAkQICRAgJAAAUICBAgJEPAa0nGXmVqWH6caApiFx5Cq1DxsJxkCmInHkHKT\nHE7NrbJITD7FEMBMPIaUmNP99skkUwwBzMRjSMZ8+oNsCGAmbJEAAb/vkYqyucV7JMTG5/T3\ntjNrl1aTDAHMw+9xpLw5jpRkO44jIS6sbAAEnEMqsnoCLitFj6dvCGDxXENq3vdcvpZYl7RP\njckK+aMC5uQY0t5sqzqkvfkb/7nmB68zDoOTdoSE0DiGlJiqDWT4AGv7c/VdcpNX53OZm734\nUQFzcgyp2a37JqS6vIvKpOJHBczJMaT0ukU6DYfR/lw3OJYIISqa90hFMryr1v5c/YN/t5BY\nIoSYuM7aZVbnF7U/Z7LdvjCHy80qZ4kQoiI5jmSyg83PXTU3E5YIISY+VzacTvt9ljVTDvlg\nR4SE0LBECBAQTH83ksHJA5chgACIQiotjiNxFSHEyyGkwnSNH0fiKkKIl8sWqRtGOrKNOY9e\nRegpy+8fFTAn1XskC1yzAfHiKkKAgCqkYzb6c2yREC/XkHL7tzVcRQjxcgzp0dHIOa81riKE\naDmf2He49FGWWzM+a8dVhBAvwazd7rI1Otks//5tCCAAgpCK+lwk7aEfQkJgHEPKLrt2pUnP\nR0LCqjmGVNQBNZMI41cRev4lHEdCTFynv3ft+eMjl9fq+SWEhJhwPhIg4Poe6dst0fdDAAHw\nuGj1xyGAAAiua2ePE/sQK8eQqmxrs6ShvS8n9iFazrt2Xy1aHTqxz/FRAXPyGBKnUSBenNgH\nCHgMiS0S4uUxJE7sQ7x8rmzgxD5Ey+sSIU7sQ6xYawcIEBIg4LpEaFfKHsqHIYAAuB+QnaIl\nQkJgXNfaHf6maImQEBjBe6TjLlW3REgIjGay4ZRctkvjn2vuMgSwZJKQiq3tJ5v/PASwaO4h\nVbvL5igtqktN4xfS/20IYOlcQzrWkw15uxpVd945ISEwzqeam3R/WzY3vKL71yGAALgeR8os\nPoXCbQggAFyOCxDgclyAgNfLcf00BBAAj5fj+nEIIAAeryL04xBAAAgJEODEPkCAkAAB55CK\nrN6ry7QnJBESAuMa0rZ9e2QSaUmEhMA4hrQ326oOaf/tZ8jaDwEEwDGkxFTt6gZm7bBqgiVC\nhAQIlgjVDZ1MKntIZ0JCcDTvkYpEecUGQkJwXGftMquPsnQaAlg+yXEkkx1ED6d3CGDxWNkA\nCLieRiF7IB+HAALgOv295ZoNgOIqQibXn9pHSAiM63uksr3wt3gXj5AQGMFkQ5knRryLR0gI\njGbWbs8Zslg3xRap2buTHkkiJARG8h4pye3ORjru2oUQ2dj8BCEhMIJZuz/LWbsq7VwqZXhJ\nESEhMM7Hkex36XKTHNqPrSiLxAxe65iQEBiPKxsSc7rfPg1/cgUhITDOkw2Hre2i1aeJveFZ\nPkJCYCQXP7E7jYItEuLlfGJfUh+JtTqx7/IeqWhn93iPhNg4z9q1WxmrU823nVm7dPDdFSEh\nMKrPR7Ja2XDMm+NISbbjOBLiItsiyT4/9nUIIAAe3yP9OAQQAI+zdufq77FKnOlvRMX9OJL1\nxU+qpF1o1/4SQkJMPF78JK93/6p90my8CAlR8RhS0v5gmaQlISEyriFVeT1dl+QWi+5u7VTb\nLSEhMo4hlcn1EvoWn49UXyf8emtLSIiLY0hb81fXUeXXOYQhj89QKs2WkBAVnysb8vudipFr\nPBASAiP4oLFaZbVE6HTfbJV/hISYOIaUm229bO64HV7N7TIEEACfKxt+HAJYPs3Khq3dSjuu\nIoRYeTwgy1WEEC+vS4QGryJkun4cApiJa0j79NJFalKLa9txzQbEyzGkot54NKu6x0viKkKI\nl/PKhkNzvYaDxbQdWyTES7Cy4VS/37F4W8NVhBAvQUiZKeyWCHEVIUTLedfuVNR7aTa7dlxF\nCPFyn2wwZldvkPjEPqyZ8/R3+25H+zljhITQeDwgu6whAKW5QuI4EqLiEFLdws+reggJUZkr\nJPmjAubEeyRAgJAAAaddu2/PfODEPsTKY0ic2Id4ue7aZc3HuhyTvw937hg5sc/xUQFzcr6K\n0O2DxsavIsRpFIiXxwtEcmIf4uV8gUj7j75ki4R4Oe/aJfUE3OU9z2705zixD/FSXSBy/Br6\nnNiHiGkuEJnZnY3EiX2IFSsbAAFCAgQICRAgJECAkAABQgIECAkQICRAwDmkImsuW1yKHk/f\nEMDiSZYIXb6WSEsiJATGMaS92VZ1SHtjcWbfb0MAAXA+jaJqTy3iclxYNcGJfYQEOIaUXrdI\n9af2CRESAqN5j1QkZi97SGdCQnCcryJkdXktpyGA5ZMcRzKZ9uORCAmhYWUDIEBIgIDzR1+m\n53OZmnTkIgwuQwDL5xhSUc99J/Vsg7QkQkJgHEPamkNzDOmgnbYjJARGsLKhue43KxuwaoKQ\nMlMQElbOedfuVNSX8WbXDuvmPtlg6st+G2N3rdUfhgAC4Dz93V4OP9UubSAkBIYDsoAAIQEC\nDiF9/6nmUz4qYE6EBAiwawcIEBIgoArpaPPZl25DAMvlGlLOeyRA8KnmN6xswJo5XyDycN6a\nstxyPhJWTbD6e3fZGp1YtIpVE4RU1Ne04z0SVs0xpOyya1ea9HwkJKya4poNzUe7OH8axWTL\nJIDpuU5/7+o//ZnmXAodQkJgWNkACBASIOC0+lv6SPqGAALhHNIkORESAkNIgAAhAQKEBAgQ\nEiBASIAAFz8BBAgJEGBlAyBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQ\nEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEC/kPap8Zk\nxaRDAL55DKm9zv62veR+PskQwEx8h5SbvDqfy9zspxgCmInvkBJT1bcrk04xBDAT3yHdPkdp\n+POUCAmB8R3S3y2kZIohgJl4DSnb7QtzuNys8uHZBkJCYLyGdP+MTGOSaoohgJn4PI50Ou33\nWdZMOeSDHRESQsPKBkCAkAABryEdd1nzLinLj1MNAczCY0hVah62kwwBzMRjSLlJDqfmVlkk\nTH8jKh5DSszpfvvEAVlExfvq774/yIYAZsIWCRDw+x6pKJtbvEdCbHxOf287s3YpS4QQE7/H\nkfLmOFKS7TiOhLiwsgEQICRAgCVCgABLhACB5SwRMl0/DgHMhAOygABLhAABtkiAAEuEAAGW\nCAECLBECBFjZAAgQEiBASIDAXCFNeBxps3H5aeAX8YW0af4DeBXdrt3m/g/An1WGNOHOH/uV\nKxVdSBa7dhPu/LFfuVYRntg3tlGYcOeP/crVWuGJfYQEveWc2KcYwg67dpBb5WkUTDZAjRP7\nAIFVbpHE2AqBE/tcbcgI50hP7Pvw2ta/4qkIVzGe2Nc/daZ+zW/6KiKstYpvZUP/wRzxK/w9\nouYLzH6v1jpC0mbUuymq/8Px2PWKMKT37YLytd3/tqhNiJDWK8aQXjZAws3Rx8mFa0Ls2q1W\nlCF16TIanKK7JsRkw1pFHpLqhd07RTfJSAhT3CFpXtwcLcKomEOSvP6pCDbiDUkRABXBUigh\nffuKFhRARbAXSEjfzis7J0BF+EoYIX15pNMxgvEpOuBFhCG5VfBtRJuXf2OdlhrSZvO0Xfhi\n187lFf39poiO0FhmSO+vZ9tXuMPm6Jcdus3bDazTMkMaGeLzkrefX8+/vS2iI1wFGdK5nRCQ\nnXP06+QCHeEm1JAab++jfuAwRUdHuAs6pMa1hNccrPJwmud2TRgxCT+k2vuOnsU0n9vRIudN\nIaISRUjXU1M7NY0eeHI95rrpvYnViiGk7iv5WtNwSO4rF+gIz2II6U1b08cp8vGXvtUnw/Tc\nxmpFGVKtb378ZYruYy5jb7DoCK+iDanxVJP1fMTYGyw6wptgQ7J/m7PZbPo2T59zGQmJjvAu\n1JC+OkGpN6WXXLp3Gfzlm49/wIoFGtIX51U879yNrSjfPPk0cu8fsGaRh/RhwuG64Rn9Be9Z\n0RF6BRqS5cqFoSNJP5wz8ZLV8Mwela1KqCGNvVAtQvm08zb0I68//nEn8Kv3cAhfsCEN+SKQ\nr2rqma14+j2dd2Pvd0fU4gvph502y5oGOuq9KyGtSBQhfT7o+tUvGfnh1/M0hn8bHa1LFCFd\nU3JfizowCfF2vpPNI8JqBB/SfSpb88rdfFqlN/Ln24NQPAYEKPSQml2oTe85sr/4cOzWoiNV\nyQhT4CFtXlYuuA78Mp/Qexi2/36uIyNsoYf0cmDU8RXduybo04LXx7ddhkQcQgzpZZ/rdRag\nZ1GQ5bCf5hl6VuldfzERoRVgSK/boPc7WJ531Pebh77a3dH7YVkEYhZeSDbHOt/mCaxe8yMd\nXX/zx3XhWLM4Q3o6Rmt1/493ettx3GzYHOFNeCHZ7qrdX+u2u3ajHT2v+aYmdAQY0heTB7bn\nHTX3Gvpi5ySmlxGoCbUQQ/rCF6/y3qnv9ncMHZp9uwNWKfKQ7FPq7+itkU+/jU3TykUfkmVK\nlhuakSXf1LRaKwjpKaWPW5TXn+hN4qtJjpeREbVVhNR9dfe/tF/W/Gw+3O/bLDgraTVWEtK5\nU0bfWoj7tzpz5j2/4tshf/khBGk9IQ2ldFv08/SF95//esCffgohWlNI3ZRevvzyjkjTEbt2\nK7KukB4pbbp7cwNnSYx+1W44xG5tIT3mHTb3o0TTdYTVWF9I53Y78WnRDx3hFysNqdvQ6yGk\nDz8z6SNC6FYX0v3svNs/7QqhIwxaU0ivVwbqBDX6o5M8IMRjLSH1LvlpTy+yiISOMGI5IZku\n6W8eWkraHood/Q3Sx4MILSekaYYYWY+9+eZewEcRh2RzUsP7IruBOwEfRRqS5YlB3RVDH3+A\njjAuwpDsz67bPN34+DkULg8GKxFZSN+dovq02lt1DhJWKaKQvj7P+/lSW91/fbgT8EkcIf10\nsQSrNXV0BCvhh/TzBUc+vSUaPzMJeBV2SC5X7RmY7h6/VgrwLNyQXC99NXgEdujKDcC7MEMS\nXD9u5Oe53AK+ElxIooswji+v43oL+EJQIQkvZGqz5JuOYC2YkLRXA+ZsPmgFEZL8ktq2v42N\nEiwtPaRprktv/xtJCVaWHNJkn+3w1W8lJVhYakhTfkDKT5fwBoYsM6RptwLfr8ojJYxYZkiT\nDjF61mzPHUgJw9YXksXZ572nJZESBhBS37dVn0aB1VhdSFaXQ/lwqhIp4RNC6rnD56vgkRL6\nrS0k1zV2pIReKwvJx7pxrBEhff87SAlv1hWSKAFSwqtVhaR7/ZMSnhHSAn4XwremkMTnNJES\nHlYUkvyFT0q4I6SF/UqEaT0hTfKiZ6OE1mpCmuoVT0qoEZL7byYlrCakZZ1yi+gQkuS3k9La\nrSSkyV/opLRy6wjJx6uclFaNkEIbBYu0ipB4hWNqHkMyz6YYoh8dYXIeQ9qvJiTeLq2Pz127\nU7Kdeog+/jtiI7g+Xt8jnUw+9RDv5uiIklbH72TD3pymHuINIcGD6GftZnhFs2u3QrGHNMsr\nmsmG9SEkQMBrSMdd1sx8Z/lxqiFeeOyIzdCqeQypSjtHkYYnwoMLacLPF0QQPIaUm+TQTtqV\nRTI8Ea4KydOrm4rgMaSkM/d9MskUQ7zw8vpmY4Sz57V2n/4gG+LF9K9wKkIr5i3S1K9xKsKd\n3/dIRdnc8vYeaUJUhC6f09/bzqxdWk0yhC9UhGd+jyPlzXGkJNv9ehzp37+fB9dhY4Q3Qa1s\nqDOaOyUqQp+QQvpXm7UkKsIHAS0R+nf38wNwQkX4LJwlQv/+2ZY0yQueijBkOUuExi7oYB3S\nBKcDsTHCiGAOyP47W4YkP0GVijAumCVC/2zfJGlDoiJYCWaL1C1peBzdrh0VwVZAS4T+3VIa\nG0j08qci2GOJUD82RvhKYEuE/KAifCuklQ1+UBF+QEjPqAg/IaQONkb41Vwh+fw0CjtUBAeE\n1KAiuGHXjoogQEhUBIGVh8TGCBoBndgnR0WQCefEPjEqgtJyTuxTDGGLiiAWzmkUOlQEuWBO\n7FNhY4QprGuLREWYSEAn9rmiIkxnLSf2UREmtdAT+8Q2G/VvBJ59n8OcZ93NNfb8574PmenR\n8WS4IaSlIaQgEdLSEFKQCGlpCClIhLQ0hBQkQloaQgoSIS0NIQWJkJaGkIJESEtDSEEipKUh\npCAR0tIQUpAIaWkIKUix/O8AZkVIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQE\nCBASIEBIgAAhAQKEBAj4CSlPTJJX/V94+t5e+3gsx92nL3fzw+7R/XxZ91+G7TwDb3ebctzO\nMyAf1w8vIbUfAZP2fuHpeyflC8Z63Ly5mfh+8uwe3Ukd0tCwnWfg7W5Tjtt5BuTjeuIjpKNJ\nTudTYo49X3j63uXfysdjOe7J/FX1/xP/CYdWPrrM27CdZ+DtblOO23kG5OP64iOk3BSXfx7M\nrucL3e/tzVYakuW4WTumdmMoe3T7xx0mH7bzDLzdbcpxO8+AfFxffLx6MlN/SGbn/1o7X+h+\nz+TaV7PtuC3fIVk+ur3Zexu28wy8//1MOW6rHlo+ri8+Xj3m9f/yO1/ofu8kfjXbjtuozFY4\ntAXLR5eZ4u/y7tvLsJ1n4O1uk47baJ4B+bi+LCmks/gv8KuQ9s0+hUfWITVklY+9oOcLqXkG\nCGlojABCKhPfOxOWj86Yw+X/rHPZDt5iQ2qfAUIaGmP5IVWJ5x27r3c8VRPCSw3p+gwQ0oDk\n9S+n84Wk92n0Pu7W/4GLLx7dWfgXMzRs59/vj2Dica/PgHxcX3w84HYmpnydsCkf81P370n/\nAq3HLdNtKRxX/Ogasr+YoWE7A70/gknHvT8D8nF98RHSrnkfX5i85wsv35OGZDtu4XvC7ptH\nl5j6eL/ulTU0bO36DLzdbdJxH8+AfFxffIRkvbJBHJLluOUsHdk+urx+TVW5bE5xeGXD/Rnw\nurKh8wywsmFI+pjBbZ+nzhfSp9ld7b6x3bh/Rr4uVPjoqqS5qft/6KFhz49n4Pmr047bfQbk\n43ri5dVTNSt62/HMyxc6N8/qkOzGNZ2n0SfLv5X6Zipc3TA07PnxDDx/ddpxu8+AfFxPgpsd\nAZaIkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRA\ngJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQ\nguf9owbRgycheIS0BDwJwSOkJeBJCB4hLQFPwoJdEsm7H/FdmbT5d2qqc5GZ+2eC32Jq/7lP\nTSL8FHRYIaQFM2ZnLrb3L2xNeflneflK8w1j8vNbSJl5/hl4QUgLdtnmnM6nxBxuXziY3eWf\nO1NcvnWo/2jOryEVZludq+3lLvCJkBbMNDkUJnt8pd63S83jDufXkLLLbl+9E5id4RMhLdh1\nGqEzm/B32bcrmx26c1nstj0hmZsZHu+a8fe9YI+Qbm0cL/t2uTme67dLt1wIaQn4+16w95DO\nSVr/p940pfui7A1pvse7Zvy1L5hptj2F+Xt8KTf7ZsKh6eUlpGP7HolphjkQ0oLdZu06aVza\naWYT6sZOj/dIqdnXU3Wmnsm7/Mx5z2SDZ4S0YMY074SemkjbQ0T59Z3QsQ1p39yv6ap985SU\nszzi9SKkBbuEkV3eCz197XDdPv1dKjs2M+NNPrvksgN4X9lg/ujIM0JaMCYOwsFTtWCEFA6e\nqgUjpHDwVC0YIYWDpwoQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQ\nICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAgf9b6NN66t2d\n3wAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "suppressWarnings(GSA.plot(GSA.obj, fac=1, FDRcut = 0.5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7107ef92",
   "metadata": {},
   "source": [
    "### GSA Enrichment analysis using KEGG pathways\n",
    "We can use the same procedure to per enrichment analysis with the KEGG pathway. All the codes are similar but `genesets` are assigned from `KEGG_hallmark`. The code is shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "361aff0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "genesets = KEGG_hallmark\n",
    "GSA.obj<-GSA(x = as.matrix(new_expression_data), y = as.numeric(groups$groups), genenames=genenames, genesets=genesets, resp.type=\"Two class unpaired\", nperms=100, random.seed=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "00afb710",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj = GSA.obj, geneset.names = names(genesets), FDRcut = .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15cdae3",
   "metadata": {},
   "source": [
    "A table of the negative gene sets. \"Negative\" means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. E.g., for two classes coded 1,2, lower expression correlates with class 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c9e7aad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>214</td><td>hsa04260</td><td>-0.6732</td><td>0   </td><td>0     </td></tr>\n",
       "\t<tr><td>239</td><td>hsa04723</td><td>-0.703 </td><td>0   </td><td>0     </td></tr>\n",
       "\t<tr><td>15 </td><td>hsa00190</td><td>-1.0477</td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>89 </td><td>hsa00970</td><td>-0.8305</td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>252</td><td>hsa04714</td><td>-0.523 </td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>318</td><td>hsa05010</td><td>-0.497 </td><td>0.01</td><td>0.2485</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 214 & hsa04260 & -0.6732 & 0    & 0     \\\\\n",
       "\t 239 & hsa04723 & -0.703  & 0    & 0     \\\\\n",
       "\t 15  & hsa00190 & -1.0477 & 0.01 & 0.2485\\\\\n",
       "\t 89  & hsa00970 & -0.8305 & 0.01 & 0.2485\\\\\n",
       "\t 252 & hsa04714 & -0.523  & 0.01 & 0.2485\\\\\n",
       "\t 318 & hsa05010 & -0.497  & 0.01 & 0.2485\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 214 | hsa04260 | -0.6732 | 0    | 0      |\n",
       "| 239 | hsa04723 | -0.703  | 0    | 0      |\n",
       "| 15  | hsa00190 | -1.0477 | 0.01 | 0.2485 |\n",
       "| 89  | hsa00970 | -0.8305 | 0.01 | 0.2485 |\n",
       "| 252 | hsa04714 | -0.523  | 0.01 | 0.2485 |\n",
       "| 318 | hsa05010 | -0.497  | 0.01 | 0.2485 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score   p-value FDR   \n",
       "[1,] 214      hsa04260      -0.6732 0       0     \n",
       "[2,] 239      hsa04723      -0.703  0       0     \n",
       "[3,] 15       hsa00190      -1.0477 0.01    0.2485\n",
       "[4,] 89       hsa00970      -0.8305 0.01    0.2485\n",
       "[5,] 252      hsa04714      -0.523  0.01    0.2485\n",
       "[6,] 318      hsa05010      -0.497  0.01    0.2485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <- res$negative\n",
    "head(neg.table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e226ae92",
   "metadata": {},
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y. See \"negative\" above for more info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "bcd7ff3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 3 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>7  </td><td>hsa00053</td><td>0.7726</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>164</td><td>hsa04520</td><td>0.474 </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>169</td><td>hsa04810</td><td>0.2231</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 7   & hsa00053 & 0.7726 & 0 & 0\\\\\n",
       "\t 164 & hsa04520 & 0.474  & 0 & 0\\\\\n",
       "\t 169 & hsa04810 & 0.2231 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 7   | hsa00053 | 0.7726 | 0 | 0 |\n",
       "| 164 | hsa04520 | 0.474  | 0 | 0 |\n",
       "| 169 | hsa04810 | 0.2231 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score  p-value FDR\n",
       "[1,] 7        hsa00053      0.7726 0       0  \n",
       "[2,] 164      hsa04520      0.474  0       0  \n",
       "[3,] 169      hsa04810      0.2231 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <- res$positive\n",
    "head(pos.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0d676c13",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 47 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ALDH9A1</td><td>2.707 </td></tr>\n",
       "\t<tr><td>ACSS1  </td><td>2.612 </td></tr>\n",
       "\t<tr><td>ADH1B  </td><td>1.513 </td></tr>\n",
       "\t<tr><td>GRHPR  </td><td>0.854 </td></tr>\n",
       "\t<tr><td>ALDH2  </td><td>0.768 </td></tr>\n",
       "\t<tr><td>ALDH3A2</td><td>0.726 </td></tr>\n",
       "\t<tr><td>ALDH7A1</td><td>0.64  </td></tr>\n",
       "\t<tr><td>ACACB  </td><td>0.455 </td></tr>\n",
       "\t<tr><td>ADH6   </td><td>0.346 </td></tr>\n",
       "\t<tr><td>LDHAL6B</td><td>0.22  </td></tr>\n",
       "\t<tr><td>LDHD   </td><td>0.22  </td></tr>\n",
       "\t<tr><td>ADH4   </td><td>0.186 </td></tr>\n",
       "\t<tr><td>ADH1C  </td><td>0.146 </td></tr>\n",
       "\t<tr><td>ADH5   </td><td>0.136 </td></tr>\n",
       "\t<tr><td>ACSS2  </td><td>0.053 </td></tr>\n",
       "\t<tr><td>PKLR   </td><td>0.045 </td></tr>\n",
       "\t<tr><td>PCK1   </td><td>-0.022</td></tr>\n",
       "\t<tr><td>ADH1A  </td><td>-0.123</td></tr>\n",
       "\t<tr><td>PCK2   </td><td>-0.16 </td></tr>\n",
       "\t<tr><td>PC     </td><td>-0.213</td></tr>\n",
       "\t<tr><td>PDHA2  </td><td>-0.239</td></tr>\n",
       "\t<tr><td>ADH7   </td><td>-0.378</td></tr>\n",
       "\t<tr><td>ACOT12 </td><td>-0.533</td></tr>\n",
       "\t<tr><td>AKR1A1 </td><td>-0.559</td></tr>\n",
       "\t<tr><td>ALDH1B1</td><td>-0.623</td></tr>\n",
       "\t<tr><td>ACAT1  </td><td>-0.689</td></tr>\n",
       "\t<tr><td>LDHC   </td><td>-0.882</td></tr>\n",
       "\t<tr><td>LDHAL6A</td><td>-0.939</td></tr>\n",
       "\t<tr><td>PDHB   </td><td>-1.449</td></tr>\n",
       "\t<tr><td>DLD    </td><td>-1.453</td></tr>\n",
       "\t<tr><td>LDHB   </td><td>-1.481</td></tr>\n",
       "\t<tr><td>ME2    </td><td>-1.546</td></tr>\n",
       "\t<tr><td>ACAT2  </td><td>-1.793</td></tr>\n",
       "\t<tr><td>PKM    </td><td>-1.877</td></tr>\n",
       "\t<tr><td>HAGH   </td><td>-1.899</td></tr>\n",
       "\t<tr><td>DLAT   </td><td>-1.904</td></tr>\n",
       "\t<tr><td>ME3    </td><td>-2.031</td></tr>\n",
       "\t<tr><td>MDH1   </td><td>-2.045</td></tr>\n",
       "\t<tr><td>ACYP2  </td><td>-2.244</td></tr>\n",
       "\t<tr><td>GLO1   </td><td>-2.309</td></tr>\n",
       "\t<tr><td>FH     </td><td>-2.453</td></tr>\n",
       "\t<tr><td>MDH2   </td><td>-2.547</td></tr>\n",
       "\t<tr><td>LDHA   </td><td>-2.576</td></tr>\n",
       "\t<tr><td>ACYP1  </td><td>-2.677</td></tr>\n",
       "\t<tr><td>ME1    </td><td>-2.688</td></tr>\n",
       "\t<tr><td>PDHA1  </td><td>-3.078</td></tr>\n",
       "\t<tr><td>ACACA  </td><td>-3.324</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 47 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " Gene & Score\\\\\n",
       "\\hline\n",
       "\t ALDH9A1 & 2.707 \\\\\n",
       "\t ACSS1   & 2.612 \\\\\n",
       "\t ADH1B   & 1.513 \\\\\n",
       "\t GRHPR   & 0.854 \\\\\n",
       "\t ALDH2   & 0.768 \\\\\n",
       "\t ALDH3A2 & 0.726 \\\\\n",
       "\t ALDH7A1 & 0.64  \\\\\n",
       "\t ACACB   & 0.455 \\\\\n",
       "\t ADH6    & 0.346 \\\\\n",
       "\t LDHAL6B & 0.22  \\\\\n",
       "\t LDHD    & 0.22  \\\\\n",
       "\t ADH4    & 0.186 \\\\\n",
       "\t ADH1C   & 0.146 \\\\\n",
       "\t ADH5    & 0.136 \\\\\n",
       "\t ACSS2   & 0.053 \\\\\n",
       "\t PKLR    & 0.045 \\\\\n",
       "\t PCK1    & -0.022\\\\\n",
       "\t ADH1A   & -0.123\\\\\n",
       "\t PCK2    & -0.16 \\\\\n",
       "\t PC      & -0.213\\\\\n",
       "\t PDHA2   & -0.239\\\\\n",
       "\t ADH7    & -0.378\\\\\n",
       "\t ACOT12  & -0.533\\\\\n",
       "\t AKR1A1  & -0.559\\\\\n",
       "\t ALDH1B1 & -0.623\\\\\n",
       "\t ACAT1   & -0.689\\\\\n",
       "\t LDHC    & -0.882\\\\\n",
       "\t LDHAL6A & -0.939\\\\\n",
       "\t PDHB    & -1.449\\\\\n",
       "\t DLD     & -1.453\\\\\n",
       "\t LDHB    & -1.481\\\\\n",
       "\t ME2     & -1.546\\\\\n",
       "\t ACAT2   & -1.793\\\\\n",
       "\t PKM     & -1.877\\\\\n",
       "\t HAGH    & -1.899\\\\\n",
       "\t DLAT    & -1.904\\\\\n",
       "\t ME3     & -2.031\\\\\n",
       "\t MDH1    & -2.045\\\\\n",
       "\t ACYP2   & -2.244\\\\\n",
       "\t GLO1    & -2.309\\\\\n",
       "\t FH      & -2.453\\\\\n",
       "\t MDH2    & -2.547\\\\\n",
       "\t LDHA    & -2.576\\\\\n",
       "\t ACYP1   & -2.677\\\\\n",
       "\t ME1     & -2.688\\\\\n",
       "\t PDHA1   & -3.078\\\\\n",
       "\t ACACA   & -3.324\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 47 × 2 of type chr\n",
       "\n",
       "| Gene | Score |\n",
       "|---|---|\n",
       "| ALDH9A1 | 2.707  |\n",
       "| ACSS1   | 2.612  |\n",
       "| ADH1B   | 1.513  |\n",
       "| GRHPR   | 0.854  |\n",
       "| ALDH2   | 0.768  |\n",
       "| ALDH3A2 | 0.726  |\n",
       "| ALDH7A1 | 0.64   |\n",
       "| ACACB   | 0.455  |\n",
       "| ADH6    | 0.346  |\n",
       "| LDHAL6B | 0.22   |\n",
       "| LDHD    | 0.22   |\n",
       "| ADH4    | 0.186  |\n",
       "| ADH1C   | 0.146  |\n",
       "| ADH5    | 0.136  |\n",
       "| ACSS2   | 0.053  |\n",
       "| PKLR    | 0.045  |\n",
       "| PCK1    | -0.022 |\n",
       "| ADH1A   | -0.123 |\n",
       "| PCK2    | -0.16  |\n",
       "| PC      | -0.213 |\n",
       "| PDHA2   | -0.239 |\n",
       "| ADH7    | -0.378 |\n",
       "| ACOT12  | -0.533 |\n",
       "| AKR1A1  | -0.559 |\n",
       "| ALDH1B1 | -0.623 |\n",
       "| ACAT1   | -0.689 |\n",
       "| LDHC    | -0.882 |\n",
       "| LDHAL6A | -0.939 |\n",
       "| PDHB    | -1.449 |\n",
       "| DLD     | -1.453 |\n",
       "| LDHB    | -1.481 |\n",
       "| ME2     | -1.546 |\n",
       "| ACAT2   | -1.793 |\n",
       "| PKM     | -1.877 |\n",
       "| HAGH    | -1.899 |\n",
       "| DLAT    | -1.904 |\n",
       "| ME3     | -2.031 |\n",
       "| MDH1    | -2.045 |\n",
       "| ACYP2   | -2.244 |\n",
       "| GLO1    | -2.309 |\n",
       "| FH      | -2.453 |\n",
       "| MDH2    | -2.547 |\n",
       "| LDHA    | -2.576 |\n",
       "| ACYP1   | -2.677 |\n",
       "| ME1     | -2.688 |\n",
       "| PDHA1   | -3.078 |\n",
       "| ACACA   | -3.324 |\n",
       "\n"
      ],
      "text/plain": [
       "      Gene    Score \n",
       " [1,] ALDH9A1 2.707 \n",
       " [2,] ACSS1   2.612 \n",
       " [3,] ADH1B   1.513 \n",
       " [4,] GRHPR   0.854 \n",
       " [5,] ALDH2   0.768 \n",
       " [6,] ALDH3A2 0.726 \n",
       " [7,] ALDH7A1 0.64  \n",
       " [8,] ACACB   0.455 \n",
       " [9,] ADH6    0.346 \n",
       "[10,] LDHAL6B 0.22  \n",
       "[11,] LDHD    0.22  \n",
       "[12,] ADH4    0.186 \n",
       "[13,] ADH1C   0.146 \n",
       "[14,] ADH5    0.136 \n",
       "[15,] ACSS2   0.053 \n",
       "[16,] PKLR    0.045 \n",
       "[17,] PCK1    -0.022\n",
       "[18,] ADH1A   -0.123\n",
       "[19,] PCK2    -0.16 \n",
       "[20,] PC      -0.213\n",
       "[21,] PDHA2   -0.239\n",
       "[22,] ADH7    -0.378\n",
       "[23,] ACOT12  -0.533\n",
       "[24,] AKR1A1  -0.559\n",
       "[25,] ALDH1B1 -0.623\n",
       "[26,] ACAT1   -0.689\n",
       "[27,] LDHC    -0.882\n",
       "[28,] LDHAL6A -0.939\n",
       "[29,] PDHB    -1.449\n",
       "[30,] DLD     -1.453\n",
       "[31,] LDHB    -1.481\n",
       "[32,] ME2     -1.546\n",
       "[33,] ACAT2   -1.793\n",
       "[34,] PKM     -1.877\n",
       "[35,] HAGH    -1.899\n",
       "[36,] DLAT    -1.904\n",
       "[37,] ME3     -2.031\n",
       "[38,] MDH1    -2.045\n",
       "[39,] ACYP2   -2.244\n",
       "[40,] GLO1    -2.309\n",
       "[41,] FH      -2.453\n",
       "[42,] MDH2    -2.547\n",
       "[43,] LDHA    -2.576\n",
       "[44,] ACYP1   -2.677\n",
       "[45,] ME1     -2.688\n",
       "[46,] PDHA1   -3.078\n",
       "[47,] ACACA   -3.324"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(geneset.number = 10, genesets = genesets, GSA.obj = GSA.obj, genenames = genenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "2284bc1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAil+ZNTU1h0E9o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD///9W+WhdAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAZgElEQVR4nO3diVbiSACG0ZoIiIgs7/+yI9Aq7sH8VBbuPXN6aJdJ\nIX4TqFRC2QOdlb4HAFMgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCPh7SE8Pi3KwWD4FxwOj9NeQdrPyZh4d\nEozPX0NaluZxc7y1XTdlmRsQjNFfQ2rK5vX2pjSZwcBY/TWkUr77C9wgeyQI6PAaab093vIa\nCf4+/T0/m7Wb7ZJDgvHpcBxpeTyO1CweHEfi5pkmgAAhQYAlQhBgiRAEWCIEAQ7IQkCFJUIF\nRuYPQVz+LUcX7JFMDDIyFUO6YImQkBiZiiFdsERISIxMzZDaLxESEiNTNaQhbQKShAQB9UNa\nzUpZrK+6CaitYkinqfZ/Mw4/L2wQEiNTO6RlWe72++2yrD5/utPRLehT7ZCacpz33pXZNTYB\nPakd0svO5pclQn/cBPSkdkj3LyFZIsSUVA1p8bBal8fnm7ulJUJMStWQXicSSmksEWJKah5H\n2mxWq8XiOOWw/PlqXEJiaO7ufvy0lQ3Qwt3xn+8JCX539/rHN4QEvxtsSI4jMSofntp9esUk\nJGjjXTqfXzF5ageX+uKJnpDgUkKChJ6f2rW+9reQGIbvJur6nGy44NrfQmIIflnNcK7ude3a\nXvtbSPTvgoyqhuRKq4zJJRn1cM2Gr/4S2wSEXLQ72tsjwRcuzci1v+GTyzNy7W/46A8ZufY3\nvPeX3dHeygY498eMhARn/pqRkODVn3dHeyHBP10yEhKcdMpISHDQbXe0FxIEMhISBDISEjcv\nkZGQuHGR3dFeSNy0VEZC4pbFMhIStyu3O9oLiVsVzUhI3KhsRkLiJoV3R3shcYPyGQmJm3ON\njITErblKRkLitlxnd7QXErfkahkJiRtyvYyExM244u5oLyRuxHUzEhK34coZCYlbcO3d0V5I\nTF+FjITE0MR/7WtkJCQG5vP7hbf8vm++q8ruaC8khuXu9MfFv/7f9FcrIyExLKeQLt4t3b3+\n8e6j1TISEgNz9/L7f1EEX35HxYyExNA8Z3R5SF88tau5O9oLiSFq89Tu7uD8rx8+Gx/Uz4TE\nAP3Qwd0/P39/ekC/EhLD9HHGoVVAb99am5AYpLuzeYf9dwF9+dE+MhISw/Q63/Djk7xvv7M+\nITFEZxN336R0wT6qhqohPT0sysFi+XStTTARr6+Q7r6sY2AZVQ1pNytv5lfZBNPxGsXnJ3iX\nvGKqpWJIy9I8bo63tuumLK+xCabo7n1K3y5PrTScr1UMqSmb19ub0lxjE4zMxz3Nd1/29tm+\nV3l/p2JIpXz3l9gmGJf3h4p+Xc/w/WGkvjOyR6JH79fUfVph92kR6renV/SeUe3XSOvt8ZbX\nSBz8EtL+q+d9A2jmSzWnv+dns3az3VU2wai0f2r3usTh6mP6o7rHkZbH40jN4sFxJA7aTTa8\nvjR6l9qworKygUF6m1e4+/rJ318v7nAlQmKoPi33Pg/pDyf/XZUlQgzYh5bu3u2QbjUkS4S4\nzNmzu8+TDbf71M4SITr4dFbfzU42OCDL737O4/M5skPJyRIhhuTjE7avOrkb4hM8eyQG5OMU\nwm+dDGfKwRIhBuRDGL92cpMhWSLErz4tGvp9lzSIjiwRYlg+zMz92skNTjYMaxOMwlA6+ZWQ\nIKBmSLv7Uubrf/8R099MSc0lQs1pod3pPyIkpqTq9PfquaZVc1xm9zmkcu6Pm4CeVD0ge/zX\ntplt7ZGYmB6WCO3mcyExMRVDmpWXg7CzuZCYloohrcr9v1vbMhcSk1Jz+nv5Ws/6l/kEITEy\nVQ/IbhYvt7b3QmJKrGyAACFBgJAgoK+QTDYwKUKCAE/tIEBIECAkCHDtbwhw7W8IcO1vCHCl\nVQhw7W8IsEeCANf+hgDX/oYA1/6GACsbIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQI\nEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQE\nAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBgOCGVc9fZBFzLcEKqvAlIEhIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCKga0tPDohwslk/X2gT0omJIu1l5M7/KJqAnFUNaluZxc7y1XTdleY1NQE8qhtSUzevtTWmu\nsQnoScWQSvnuL7FNQE/skSCg7muk9fZ4y2skpqbm9Pf8bNZutrvKJqAfdY8jLY/HkZrFg+NI\nTIuVDRAgJAioH9JqVspifdVNQG3VjyP9m3H4cdJOSIxN7ZCWZbnb77fLsrrGJqAntUNqynHe\ne1dm19gE9KR2SC9LgywRYlJqh3T/EpIlQkxJ1ZAWD6t1eXy+uVtaIsSkdA5pvTjsaRbbFt/3\nz/FmY4kQU9I1pPkpjdK0KGmzWa0Wi+OUw/LHjoTE2HQMaVXmu0NIq3IfG9JeSIxOx5AOs9nv\nZuMyhMTIdAzp+LSudUiuIsRUdQxp9m+PtPn5AOuRqwgxXZnXSOvm5yU/R79cRaicu3xU0Keu\ns3aLVnuYI9dsYLoix5HK4rHN97mKEJPlKkIQ4CpCEBCY/j5qftzDnLiKEJMVCmnb7jiSqwgx\nUR1CWr+bsP79ONJ1RwV96rJHOj/AOvtlH3P1UUGfUq+RsoTEyPR1XTvHkZiUVEhPiwv/I0Ji\nSrqGtLzK8jghMTIdQ3rr6Jdrp/59EzACnU/se9zPy3Y7L2btuGWBWbuH573Rps3ybyf2MVmB\nkNaHc5FavEZyYh/T1TGkxfNTu22Z7Z9ahPTLiX0dRwV96hjS+hDQcTHq71cRchoF09V1+vvh\ndB3iX96m5fR9TuxjspzYBwFdXyO12BO9cGIf01Vz0aoT+5iswHXt2nNiH1PVMaTdYh5d0vDF\nJmAEOj+1s2gVhAQRfZ3Y1/smIElIECAkCBASBAgJAoQEAUKCgK5LhB62saF8swkYge4HZK/R\nkpAYma5r7R7vr9GSkBiZwGukp4dZuiUhMTKZyYZN87xf+v19zbtsAoYsEtJ63vadzf+8CRi0\n7iHtHp53R7P17rmmCy+k33oTMHRdQ3o6TDYsT1c1yZ1KISRGpvOp5mW2ejnb/OcrA/11EzAC\nXY8jLaLvQvHVJmAEKl6O64+bgBHwHrIQUPVyXH/aBIyAy3FBgKsIQYCQIMCJfRAgJAjoHNJ6\ncXhWt8iekCQkRqZrSPPTy6PSREsSEiPTMaRVme8OIa1avIfsHzcBI9AxpKbsTqsbzNpx0wJL\nhIQEgSVCh4Y2ZRYb0l5IjE7mNdK6SV6xQUiMTtdZu8W/dQ256zV82gQMX+Q4Ulk8hobz5SZg\n8KxsgICup1Fc9J1PD6dngovlL+deCImR6Tr9PW9/zYbd7Gyt+M+vqYTEyHS/ilD5bffyYlma\nx9N1u7brpvx4sQchMTJdXyNtTxf+bvMUrymb19ubny/dJSRGJjDZsF02pc1TvHeLH35eCSEk\nRiYza7dqc4asPRLTldgjHZ/d/X4k6fk10vp0soXXSExN5DVSs2x1NtL8bNZu9uOrKiExMoFZ\nu/vWF+R6Wh6PIzWLB8eRmJbOx5HCi4M+bwJGoOrKhj9tAkag82TD47z1otXd/ds0uelvJiVy\n8ZN2p1HsmtNCu9N/REhMSecT+5rDLqbViX3Lw9fsVs2xOSExKZ1n7U4HWducat6cvnHbzLZC\nYmJS74/UYmXDy5fs5nMhMTGxPdLv7x/79l5Ks7mQmJaKr5HeLiK5LXMhMSkVZ+32y9d61r8s\nchUSI9P9ONIFFz/ZLF5ube+FxJS4+AkECAkCuoa0Wx6m65plq0V3riLEVHUMadv8u4R+i/dH\nchUhpqtjSPNyf9gX7ZZl8d2Xv/rlKkLl3OWjgj5VXNngmg1MV+CNxg52FywR+vyXxKigTx1D\nWpb5Yd7gaf7zxUyO7JGYroorG1xFiOnKrGyYt3qbMVcRYrKqHpB1FSGmysoGCOga0mr2/JJn\nVmatr213+SZg+DqGtD5MYx8vahItSUiMTOeVDY/H6zU8XvpuzI4jMSmBlQ2bw1T2pat6bjWk\nu7u+R8BVBEJalPXlIbXfxKTcHf9hejo/tdusD4sULn5q134TU3L3+gcT032yoZSHww6p/Zsy\nX7iJKRHSZHWe/j4t9mnxPmN7J/Z5ajddFQ/IOrFvb7JhsiqG9MuJfYlNQE86hHSYqLvkpFan\nUTBdFUNyYh/TVfGpnT0S01X3NZIT+5ioTk/tLrzwjxP7mKyaITmxj8nq+tRucXxbl6fm/psv\n/hshMTKdryL08kZjv19F6I+bgBGoeIHIP24CRqDzBSLbv/XlHzcBI9D5qV1zmDdYN4cl4DlC\nYmRSF4j8/Rr6f94EDF/mApGL6NlIQmJ0XNcOAoQEAUKCACFBgJAgQEgQICQIEBIEdA5pvThe\ntngbGs9Xm4DBiywRev5YEy1JSIxMx5BWZb47hLQq0TP7hMTIdD6NYnc6Fcn5SNy0wIl9QoKO\nIc3+7ZEO79oXJCRGJvMaad2UVWxIeyExOp2vItTq3SU6bQKGL3IcqSxavT3SXzcBg2dlAwQI\nCQI6v/XlbL/fzsrsl2sQd9kEDF/HkNaHue/mMNsQLUlIjEzHkObl8XgM6TE7bSckRiawsuF4\n3W8rG7hpgZAWZS0kblznp3ab9eGy357acdu6TzaUw2W/S4lea1VIjEzn6e/Tu8HOsksbhMTI\nOCALAUKCgKpvxnzNUUGfhAQBntpBgJAgIBXSU/S9L4XEyHQNaek1EgTe1fxF55UNV5u5gOvr\nfIHIx/28bLdz5yNx0wKrvx+e90Ybi1a5aYGQ1odr2nmNxE3rGNLi+andtsz2T0LipiWu2XB8\naxfvRsEt6zr9/XD42305nkuRIyRGxsoGCBASBHRa/R0dyVebgJHoHNJVchISIyMkCBASBAgJ\nAoQEAUKCABc/gQAhQYCVDRAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUFA/ZBWs1IW66tuAmqrGNLpGpLz0+Ukf37PWSEx\nMrVDWpblbr/fLsvqGpuAntQOqSm7w+1dmV1jE9CT2iG9XCP852uFC4mRqR3S/UtIzTU2AT2p\nGtLiYbUuj883d8ufZxuExMhUDen1/V9KaXbX2AT0pOZxpM1mtVosjlMOyx87EhJjY2UDBAgJ\nAqqG9PSwOL5KWiyfrrUJ6EXFkHazs3ecnV9lE9CTiiEtS/O4Od7arhvT30xKxZCasnm9vXFA\nlkmpvvr7q7/ENgE9sUeCgLqvkdbb4y2vkZiamtPf87NZu5klQkxJ3eNIy+NxpGbx4DgS02Jl\nAwQICQIsEYIAS4QgYDhLhMq5P24CeuKALARYIgQB9kgQYIkQBFgiBAGWCEGAlQ0QICQIEBIE\n9BWS40hMipAgwFM7CBASBAgJApzYBwFO7IOA4ZzYl9gE9MRpFBDgxD4IsEeCACf2QYAT+yDA\niX0QYGUDBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIEBIECAkChAQBQoIAIUGAkCBASBAwnJDKuetsAq5lOCFV3gQkCQkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEARVD\nKu9dYxPQk4ohrYTEZNV8ardp5tfeBPSj6mukTVleexPQi7qTDauyufYmoA9m7SBASBAgJAio\nGtLTw+I4871YPl1rE9CLiiHtZmdHkX6eCBcSI1MxpGVpHk+Tdtt18/NEuJAYmYohNWdz35vS\nXGMT0JOqa+2++0tsE9ATeyQIqPsaab093vIaiampOf09P5u1m+2usgnoR93jSMvjcaRm8eA4\nEtMyspUN//13/Y3D5cYV0n//KYlBGtMSoWNFSmKIRrRE6L9TREJigIazROi3Czr8d9ol2SMx\nROM5IHvcG3mNxDCNaInQv10SDNB49kjmGRgwS4QgwBIhCLBECALGtbIBBkpIECAkCOgrJKea\nMylCggBP7SBASBAgJAgY04l9MFgjOrEPhms4J/YlNgE9GdNpFDBYIzqxD4bLHgkCnNgHAU7s\ng4CBntgHI3N5Dn3uLoa3qxreiC40vDswuBFdaUBCOje8EV1oeHdgcCMSUgXDG9GFhncHBjci\nIVUwvBFdaHh3YHAjElIFwxvRhYZ3BwY3IiFVMLwRXWh4d2BwIxJSBcMb0YWGdwcGNyIhVTC8\nEV1oeHdgcCMSUgXDG9GFhncHBjciIVUwvBFdaHh3YHAjElIFwxvRhYZ3BwY3IiFVMLwRXWh4\nd2BwI5pgSDAZQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAuqE\ntGxKs9x9/YF3n1vVCrvliFazD182GO3uwJ+vCR8e0dkj++nLehnQ2QObGVCVn/HpLWBmX37g\n3ec2lR7ztiNaHm82Ayyp3R3YVAzppxGdPbKfvqyXAZ09sKEB1fgZP5Vms9805emLD7z73PO/\n64TUckSbcr87/L/0vsqgLtH6DiyGMKKzR/bTl/UyoLMHNjWgGr+4y7J+/vOxPHzxgfPPrcq8\n1rOQdiNanEZTazd5gZZ3YPX2BX2O6OyR/fRlvQzo7IFNDajG78iiHN4k8+z/jmcfOP9cWdb6\nnW07opMBhtTyDqzKaggjOntkP/90exnQyWFMqQHV+B0pH//HfvaB889tqv3Oth3R0a7Mqwzq\nEi3vwKKs759fSfc9orNH9tOX9TOgo+MDmxrQkELaV/kBXziiw//W11UGdYnWIR3V+B/Bb7+3\nAwzp+MAKqdKI9tum2gv29lregVIen//Hu6zxBG98IZ0eWCFVGtGuGd4Tu4ufm1aYbR5dSP8e\n2DGF1Hwc69kHmi9/3AMa0bzGIY+LXXAH9nV+qj+N6OzfnwfX14D+PbCpAdX4xT1NjGw/zp9s\n36aYXj9XKaTWI9rO5tsqI7rQBT/SfZ2f6k8jOhvD58H1M6DXBzY1oBq/uA/HV+vrsvziAx8+\nVymktiNaD3DC7qjlHWjK4dh9jV/bH0d08O+R/fRl/Qzo7YFNDajGL27rlQ3VQmo5ou1QO2p7\nB5aH34/dssa0488rG14f2WGsbDh7YMe0smE/e5uEPf08zz4wezdBWymkliO6LzUXfV6k3R3Y\nNcebVQ4k/TSi/dsj+/6jPQ3o/IENDajK78juuMD2tL3y4QNnN/f1Qmo3onL28x6Ylj/Sw81Z\nndUNP41o//bIvv9oTwM6f2BDAxrg7wiMj5AgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkEZviG8oeHs8CKMnpCHwIIyekIbAgzB6QhoCD8KAPSeyPH/H7V2ZHf89K7v9\nelFe36L7JabTn6tZaeq8kTlvhDRgpTwc3sN+/vqBedk+/7l9/sjD6e3tl/tPIS3K+++hCiEN\n2PM+Z7PfNOXx5QOP5eH5z4eyfv7U4+GvZf8xpHWZ7/a7+fOXUJOQBqwcc1iXxdtHDs/tZuXt\nC/YfQ1o8P+07PAlc7KlJSAP2bxrhbDbh/vm53fb4hG6/XT/MvwipvOhhvLfMz3vA3kJ6aePp\n+bndsjztDy+XXnIR0hD4eQ/Y55D2zezwz2HXNFutt1+G1N94b5kf+4CV475nXe7fPrQsq+OE\nw7GXDyE9nV4jmWbog5AG7GXW7iyN53aOswmHxjZvr5FmZXWYqiuHmbzn79mvTDZUJqQBK+X4\nSuhdE7PTIaLlv1dCT6eQVsevO3Z1evHUbHsZ8e0S0oA9h7F4fi307mOP//ZP98+VPR1nxo/5\nPDTPTwBfVzaUex1VJqQBM3EwHh6qARPSeHioBkxI4+GhGjAhjYeHCgKEBAFCggAhQYCQIEBI\nECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCgoD/ASdtRbeB/S6NAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "suppressWarnings(GSA.plot(GSA.obj, fac=1, FDRcut = 0.5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49fd3929-7d2e-401d-93c6-24572bec35f6",
   "metadata": {},
   "source": [
    "The next submodule will go indepth about meta analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e870e8eb-ee47-439b-8154-0a720927bd79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.2 (2023-10-31 ucrt)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 19045)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=English_United States.utf8 \n",
       "[2] LC_CTYPE=English_United States.utf8   \n",
       "[3] LC_MONETARY=English_United States.utf8\n",
       "[4] LC_NUMERIC=C                          \n",
       "[5] LC_TIME=English_United States.utf8    \n",
       "\n",
       "time zone: America/Chicago\n",
       "tzcode source: internal\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] GSA_1.03.2            fgsea_1.28.0          BiocManager_1.30.22  \n",
       " [4] lubridate_1.9.3       forcats_1.0.0         stringr_1.5.1        \n",
       " [7] dplyr_1.1.4           purrr_1.0.2           readr_2.1.4          \n",
       "[10] tidyr_1.3.0           tibble_3.2.1          ggplot2_3.4.4        \n",
       "[13] tidyverse_2.0.0       hgu133plus2.db_3.13.0 org.Hs.eg.db_3.18.0  \n",
       "[16] AnnotationDbi_1.64.1  IRanges_2.36.0        S4Vectors_0.40.2     \n",
       "[19] Biobase_2.62.0        BiocGenerics_0.48.1  \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] tidyselect_1.2.0        IRdisplay_1.1           farver_2.1.1           \n",
       " [4] blob_1.2.4              Biostrings_2.70.1       bitops_1.0-7           \n",
       " [7] fastmap_1.1.1           RCurl_1.98-1.13         digest_0.6.33          \n",
       "[10] timechange_0.2.0        lifecycle_1.0.4         KEGGREST_1.42.0        \n",
       "[13] RSQLite_2.3.4           magrittr_2.0.3          compiler_4.3.2         \n",
       "[16] rlang_1.1.2             tools_4.3.2             utf8_1.2.4             \n",
       "[19] data.table_1.14.10      labeling_0.4.3          bit_4.0.5              \n",
       "[22] repr_1.1.6              BiocParallel_1.36.0     pbdZMQ_0.3-10          \n",
       "[25] withr_2.5.2             grid_4.3.2              fansi_1.0.6            \n",
       "[28] colorspace_2.1-0        scales_1.3.0            cli_3.6.2              \n",
       "[31] crayon_1.5.2            generics_0.1.3          httr_1.4.7             \n",
       "[34] tzdb_0.4.0              DBI_1.2.0               cachem_1.0.8           \n",
       "[37] zlibbioc_1.48.0         parallel_4.3.2          XVector_0.42.0         \n",
       "[40] base64enc_0.1-3         vctrs_0.6.5             Matrix_1.6-1.1         \n",
       "[43] jsonlite_1.8.8          hms_1.1.3               bit64_4.0.5            \n",
       "[46] snow_0.4-4              glue_1.6.2              codetools_0.2-19       \n",
       "[49] cowplot_1.1.2           stringi_1.8.3           gtable_0.3.4           \n",
       "[52] GenomeInfoDb_1.38.2     munsell_0.5.0           pillar_1.9.0           \n",
       "[55] htmltools_0.5.7         IRkernel_1.3.2          GenomeInfoDbData_1.2.11\n",
       "[58] R6_2.5.1                vroom_1.6.5             evaluate_0.23          \n",
       "[61] lattice_0.21-9          png_0.1-8               memoise_2.0.1          \n",
       "[64] Rcpp_1.0.11             uuid_1.1-1              fastmatch_1.1-4        \n",
       "[67] pkgconfig_2.0.3        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca1ba4ed-3e1c-4d5a-8554-40703527a4d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff8e552-9979-4d45-9fbd-debe85a17f62",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "environment": {
   "kernel": "ir",
   "name": "r-cpu.4-2.m107",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-2:m107"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  },
  "toc-showcode": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
